---
title: "Combined analysis of RNA-seq, ChIP-seq, and Cut&Run data of Ikzf1/3 KO samples"
author: Sepideh (Momeneh) Foroutan
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    code_folding: hide
    fig_width: 9
    fig_height: 6
params:
  orig_date: "Date first generated: 11 Dec, 2019"
  update_date: !r paste("Last updated on:", format(Sys.time(), '%d %B, %Y'))
---


`r params$orig_date`

`r params$update_date`

This script reads in several processed RNA-seq, ChIP-seq, and Cut&Run data and performs several combined exploratory analyses of the data; these include in-house in-vitro and ex-vivo data, as well as a number of publicly available data sets. Some of the analyses performed in this script have been published in [Nature Immunology](https://www.nature.com/articles/s41590-023-01718-4) in 2024. 

# Set up
Set paths and required libraries.

```{r}
library(tidyr)
library(limma)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(stringr)

chipPath <- "../data/ChIPseq/"
rnaPath <- "../data/RNA-seq/"
sigPath <- "/Users/mfor0011/Documents/data/signatures/ProcessedSigs/"
hpaPath <- "/Users/mfor0011/Documents/data/HPA/"

mainDir <- getwd()
outPath <- "../output/"
figPath <- "../figure/"
dataPath <- "../data/"
scriptPath <- "../script/"


equal_breaks <- function(n = nBreak, s = scalingFactor, ...){
  function(x){
    # rescaling
    d <- s * diff(range(x)) / (1+2*s)
    round( seq(min(x)+d, max(x)-d, length=n), 2)
  }
}

nBreak = 3
scalingFactor = 0.05
t <- tSize <- 1.2

currentcol <- c(brewer.pal(9, "Set1")[-6])

ifelse(!dir.exists(file.path(mainDir, outPath)), dir.create(file.path(mainDir, outPath)), FALSE)
ifelse(!dir.exists(file.path(mainDir, figPath)), dir.create(file.path(mainDir, figPath)), FALSE)
ifelse(!dir.exists(file.path(mainDir, dataPath)), dir.create(file.path(mainDir, dataPath)), FALSE)
ifelse(!dir.exists(file.path(mainDir, scriptPath)), dir.create(file.path(mainDir, scriptPath)), FALSE)
```

# Read data
Read the filtered RPKM data for RNA-seq, the results of DE analaysis performed using limma, as well as the results of ChIP-seq peak calling. Additionally, I read selected signatures and DEGs from previous works to perform single sample scoring using singscore. Note that I did not perform the pre-processing of the RNA-seq data and ChIP-seq data

## Read RPKM data
```{r}
##-------

rpkm <- read.table(
  paste0(
    rnaPath,
    "RPKM_DKO_IKZF1_IKZF3.txt"
  ),
  header = T, 
  sep = "\t"
)
## remove dup columns
rpkm <- rpkm[! duplicated(rpkm$external_gene_name), ]
ex <- as.matrix(rpkm[, 4:ncol(rpkm)])
row.names(ex) <- rpkm$external_gene_name

ex <- log2(ex + 1)


##---- read ex-vivo data: from RPKM_Ikzf1KO_ImmM1M2_RNAseq.xlsx file

rpkm_exvivo <- read.table(
  paste0(
    rnaPath,
    "RPKM_IKZF1_Exvivo.txt"
  ),
  header = T, 
  sep = "\t"
)
## remove dup columns
# rpkm_exvivo <- rpkm_exvivo[! duplicated(rpkm_exvivo$X), ]
ex_ex <- as.matrix(rpkm_exvivo[, 2:ncol(rpkm_exvivo)])
row.names(ex_ex) <- rpkm_exvivo$X

ex_ex <- log2(ex_ex + 1)

```


```{r}
annot_rna <- data.frame(SampleID = colnames(ex),
                        Samples = stringr::str_remove( colnames(ex), "RPKM_"))

annot_rna$Group <- stringr::str_remove(annot_rna$Samples, "_S[0-9]+")
row.names(annot_rna) <- annot_rna$SampleID

all(row.names(row.names(annot_rna)) == colnames(ex))

annot_rna$Group <-
  factor(annot_rna$Group,
         levels = c("Cre", "IKZF1_KO", "IKZF3_KO", "IKZF3_IKZF1_KO"))



design <- model.matrix(~ 0 + Group, data = annot_rna)
colnames(design) <- gsub("Group", "", colnames(design))
cont.mat <-
  makeContrasts(
    DKO = IKZF3_IKZF1_KO - Cre,
    KO1 = IKZF1_KO - Cre,
    KO3 = IKZF3_KO - Cre,
    levels = design
  )


## for ex-vivo
annot_rna_exvivo <- data.frame(Group_exvivo = c("Cre", "Cre", "Cre", "Ikzf1", "Ikzf1", "Ikzf1"))
Group_exvivo <- c("Cre", "Ikzf1")
design_exvivo <- model.matrix(~ 0 + Group_exvivo, data = annot_rna_exvivo)
colnames(design_exvivo) <- gsub("Group_exvivo", "", colnames(design_exvivo))
cont.mat_exvivo <-
  makeContrasts(
    Exvivo = Ikzf1 - Cre,
    levels = design_exvivo
  )

```

### QC
#### RLE plot
sample S3 and S4 from IKZF3 KO seem to be different than the other two S1 and S2 samples. Similarly, S6 from DKO group seems to be different than the other three samples.
```{r}
source(paste0(scriptPath, "RLE_ggplot.R"))
# pdf(paste0(figPath, "QC_RLE_logRPKM.pdf"), height = 6, width = 8)
RLE_ggplot (
  exprData = ex,
  annot = annot_rna[, "Group", drop = F],
  largeData = F,
  textSize = 1,
  ylim = c(-2, 2),
  mainTitle = "",
  cols = c(
    brewer.pal(8, "Dark2")[-5],
    brewer.pal(10, "Paired"),
    brewer.pal(12, "Set3")
  )
)
# dev.off()
```

#### PCA
```{r}
source(paste0(scriptPath, "PCA_plot.R"))

cols <- brewer.pal(8, "Dark2")[1:4]

s <- svd(apply(ex ,
  1,
  function(x)
  scale(x, scale = FALSE , center = TRUE)))  

# pdf(paste0(figPath, "QC_PCA_logRPKM.pdf"), height = 8, width = 8)
PCA_plot (
  ex,
  clin = annot_rna,
  group_to_test = "Group",
  svdResult = s,
  plot_cex = 1,
  legend_cex = 1,
  labeled = FALSE,
  group_to_shape = NULL,
  cols = cols)
# dev.off()
```

I remove the two outlier samples to see how PCA changes. We need to take into account this differences when doing DE analysis. 
```{r}
badSamples <- c("RPKM_IKZF3_KO_S3", "RPKM_IKZF3_KO_S4" )
ex2 <- ex[, ! colnames(ex) %in% badSamples]

s <- svd(apply(ex2 ,
  1,
  function(x)
  scale(x, scale = FALSE , center = TRUE)))  

# pdf(paste0(figPath, "QC_PCA_logRPKM_rm2Samples.pdf"), height = 8, width = 8)
PCA_plot (
  ex2,
  clin = annot_rna,
  group_to_test = "Group",
  svdResult = s,
  plot_cex = 1,
  legend_cex = 1,
  labeled = FALSE,
  group_to_shape = NULL,
  cols = cols)
# dev.off()
```


## Set colours
```{r}
targetCols <- brewer.pal(8, "Set2")[c(1, 2, 3, 8)]
names(targetCols) <- c("Ikzf1/3 target", "Ikzf3 target", "Ikzf1 target", "Non-Target gene")

dirCols <- brewer.pal(9, "Set1")[c(9, 1)]
names(dirCols) <- c("Consistent", "Different")

# degCols <- currentcol[1:7]
# names(degCols) <- unique(plotData$DEGs_group)

degCols <- c(currentcol[1:7], "gray")
names(degCols) <- c(
  "DEGs only in Ikzf1 KO",
  "DEGs in DKO & Ikzf1 KO",
  "DEGs in all KOs",
  "DEGs in DKO & Ikzf3 KO",
  "DEGs in Ikzf1 & Ikzf3 KOs",
  "DEGs only in DKO",
  "DEGs only in Ikzf3 KO", 
  "Non-DEGs"
)
```

## Read DE stats
Three in vitro, and one ex-vivo (for Ikzf1); we later also read in the 
```{r}
DE_DKO <-
  read.table(
    paste0(dataPath, "RNA-seq/DEstat_DKO_IKZF1_IKZF3_Cre.txt"),
    header = T,
    sep = "\t",
    stringsAsFactors = F,
    check.names = F
  )

DE_KO1 <-
  read.table(
    paste0(dataPath, "RNA-seq/DEstat_IKZF1_Cre.txt"),
    header = T,
    sep = "\t",
    stringsAsFactors = F,
    check.names = F
  )

DE_KO3 <-
  read.table(
    paste0(dataPath, "RNA-seq/DEstat_IKZF3_Cre.txt"),
    header = T,
    sep = "\t",
    stringsAsFactors = F,
    check.names = F
  )


## Ex vivo: from DGE_GeneSetTesting_Ikzf1KORNAseq.xlsx
DE_KO1_Exvivo <-
  read.table(
    paste0(dataPath, "RNA-seq/DEstat_IKZF1_Cre_Exvivo.txt"),
    header = T,
    sep = "\t",
    stringsAsFactors = F,
    check.names = F
  )


```

Klrg1 (0.94 logFC); CD11b (Itgam is DEG in all KD) and CD27 (is DEG in KO1) and CD122 (IL2RB not available in the data - removed probably due to low expression).

Define DEGs using p-values and logFCs
In Aug 2021, I considered DEGs only based on adjusted p-values
```{r}
DEGs_DKO <- DE_DKO %>% 
  filter(adj.P.Val < 0.05) %>% 
  # filter(abs(logFC) > 1 & adj.P.Val < 0.05) %>% 
  data.frame()
## 1768 |logFC| > 1 & adjpval <0.05
## 5119 only adjpval <0.05

dim(DEGs_DKO)

DEGs_KO1 <- DE_KO1 %>% 
  # filter(abs(logFC) > 1 & adj.P.Val < 0.05) %>% 
  filter(adj.P.Val < 0.05) %>% 
  data.frame()
## 1057 |logFC| > 1 & adjpval <0.05
## 3680  only adjpval <0.05

dim(DEGs_KO1)

DEGs_KO3 <- DE_KO3 %>% 
  # filter(abs(logFC) > 1 & adj.P.Val < 0.05) %>% 
  filter(adj.P.Val < 0.05) %>% 
  data.frame()
## 763 |logFC| > 1 & adjpval <0.05
## 2233  only adjpval <0.05

dim(DEGs_KO3)


# DEGs_KO1_Exvivo <- DE_KO1_Exvivo %>% 
#   # filter(abs(logFC) > 1 & adj.P.Val < 0.05) %>% 
#   filter(adj.P.Val < 0.05) %>% 
#   data.frame()

DEGs_KO1_Exvivo <- DE_KO1_Exvivo[DE_KO1_Exvivo$adj.P.Val < 0.05, ]

dim(DEGs_KO1_Exvivo)
# write.csv(DEGs_KO1_Exvivo, paste0(outPath, "DEGs_Ikzf1_Exvivo.csv"), row.names = F)
```


```{r}

gs <-
  c(
    "Ube2f",
    "Rnf7",
    "Cish",
    "Arih2",
    "Cul5",
    "Dcun1d3",
    "Rasa2",
    "Spop"   ,
    "Tm2d1",
    "Ptpn1" ,
    "Cpsf7",
    "Aip" ,
    "Tsc2",
    "Pten" ,
    "Ptpn9" ,
    "Ahr",
    "Ptpn4",
    # "Kmt2d",
    "Sbno2",
    "Arrdc3",
    "Cand1",
    "Inpp5d" ,
    "Sh2d1a", 
    
    ## also add some other genes
    "Trp53",
    "Pmaip1",
    # "Socs1",
    "Socs2",
    "Socs3",
    "Socs4",
    # "Socs5",
    "Socs6",
    "Socs7", 
    "Serpinb1a", 
        
    ## from these papers:
    #     Fig 3 from this paper: https://celldiv.biomedcentral.com/articles/10.1186/s13008-016-0016-3
    #    Fig 3 from this paper: https://www.sciencedirect.com/science/article/pii/S1044579X20300857
    
    "Wsb1",
    "Wsb2",
    "Spsb1",
    "Spsb2",
    "Spsb3",
    # "SPSB4",
    paste0("Asb", 1:18),
    # "RAB40A",
    # "RAB40B",
    "Rab40c",
    "Lrrc41",
    # MUF1
    "Eloa"
  )

gs <- gs[gs%in% row.names(ex)]
png(
  # paste0(figPath, "CRL5_SOCS_relatedGenes_Invitro_logRPKM.png"),
  paste0(figPath, "CRL5_SOCS_relatedGenes_Invitro_logRPKM_rmBadSamples3.png"),
  height = 11,
  width = 6,
  res = 300,
  unit = "in"
)
# Heatmap(matrix = ex[gs,], name = "logRPKM")
Heatmap(matrix = ex[gs, -c(3, 4, 5)], name = "logRPKM")
dev.off()

png(
  # paste0(figPath, "CRL5_SOCS_relatedGenes_Invitro_logRPKM_standardize.png"),
  paste0(figPath, "CRL5_SOCS_relatedGenes_Invitro_logRPKM_rmBadSamples3_standardize.png"),
  height = 11,
  # width = 6,
  width = 5,
  res = 300,
  unit = "in"
)

coolmap(ex[gs, -c(3, 4, 5)], margins=c(14,6))
# coolmap(ex[gs, ], margins=c(14,6))
dev.off()
```


```{r}
dd <- DEStatDataPrs[DEStatDataPrs$Symbol %in% gs, ]
rmGenes <- c("Ahr", "Arrdc3", "Cpsf7", "Inpp5d", "Pten", "Ptpn4", "Ptpn9", "Sbno2", "Serpinb1a", "Sbno2", "Sh2d1a", "Tsc2")
dd2 <- dd[ dd$DEGs_group != "Non-DEGs" & !dd$Symbol %in% rmGenes, 1:13]


png(
  # paste0(figPath, "CRL5_SOCS_relatedGenes_Invitro_logRPKM.png"),
  paste0(figPath, "CRL5_SOCS_SelGenes_Invitro_logRPKM_rmBadSamples3.png"),
  height = 8,
  width = 5,
  res = 300,
  unit = "in"
)
# Heatmap(matrix = ex[gs,], name = "logRPKM")
Heatmap(matrix = ex[dd2$Symbol, -c(3, 4, 5)], name = "logRPKM")
dev.off()

png(
  paste0(figPath, "CRL5_SOCS_SelGenes_Invitro_logRPKM_standardize.png"),
  # paste0(figPath, "CRL5_SOCS_SelGenes_Invitro_logRPKM_rmBadSamples3_standardize.png"),
  height = 8,
  width = 6,
  # width = 5,
  res = 300,
  unit = "in"
)

coolmap(ex[dd2$Symbol,], margins=c(14,6))
# coolmap(ex[dd2$Symbol, -c(3, 4, 5)], margins=c(14,6))
dev.off()

```


Merge three In vitro DE stats together:
```{r}
colnames(DE_DKO)[-1] <- paste(colnames(DE_DKO)[-1], "DKO", sep = "_")
colnames(DE_KO1)[-1] <- paste(colnames(DE_KO1)[-1], "KO1", sep = "_")
colnames(DE_KO3)[-1] <- paste(colnames(DE_KO3)[-1], "KO3", sep = "_")


DEstats <- Reduce(merge, list(DE_DKO, DE_KO1, DE_KO3))

## remove some of the duplicated columns:
rmcols <- c("external_gene_name_KO1",
            "gene_biotype_KO1",
            "AveExpr_KO1",
            "external_gene_name_KO3",
            "gene_biotype_KO3", 
            "AveExpr_KO3")

DEstats <- DEstats[, ! colnames(DEstats) %in% rmcols]

colnames(DEstats)[2:3] <- c("Symbol", "Biotype")

```

###Check DE stat for genes
PRDM1 – KLRG1 – ZEB2 – MCAM – Ki67 – Bhlhe40 – Zbz32 () – STAT4
Waterfall plots for the common ikzf2 and 

```{r}
# checkg <- c("Mki67", "Bcl3", "Prdm1", "Klrg1", "Zeb2", "Mcam", "Zbtb32", "Bhlhe40", "Stat4")
# 
# DEstats_g <- DEstats[DEstats$Symbol %in% checkg, 
#                      grepl("Symbol|adj|logFC", colnames(DEstats))]

# write.csv(DEstats_g, paste0(outPath, "DE_stat_selectedGenes.csv"), row.names = F)
```


## Read signatures
### Imm/M1/M2
```{r}
## 51 genes
m1_im <- read.csv(paste0(dataPath, "ImmM1M2/DE_NK_mature_vs_immature.csv"), 
                  stringsAsFactors = F)
## 1041
m2_im <- read.csv(paste0(dataPath, "ImmM1M2/DE_NK_terminal_vs_immature.csv"), 
                  stringsAsFactors = F)
## 37
m2_m1 <- read.csv(paste0(dataPath, "ImmM1M2/DE_NK_terminal_vs_mature.csv"), 
                  stringsAsFactors = F)

## also read the RPKM data for immM1M2 - 12482 genes
mm <- read.csv(paste0(dataPath, "ImmM1M2/IMM_M1_M2 normalized_intensities(log2 RPKM).csv"), 
               stringsAsFactors = F)

## remove 62 NA symbols:
mm <- mm[complete.cases(mm$Symbol), ]   ## 12420 genes
```

### Processed signarures
```{r}
# sigsTum <- readRDS(paste0(sigPath, "SigLists_Tumour.RDS"))
sigsImmune <- readRDS(paste0(sigPath, "SigLists_Immune.RDS"))
```

### Selected MSigDB signatures
prolif, OXPHOS, glycolysis, IL15/2/21 signalling, apoptosis, MAPK 
```{r}

martList <- readRDS(paste0(dataPath, "MartList_Ensembl_Human_Mouse.RDS"))

martMouse <- martList$martMouse
martHuman <- martList$martHuman


mSig <- msigdf::msigdf.mouse

checkSigName_unidir <- c(
  unique(grep("WHITFIELD_CELL_CYCLE_", mSig$geneset, ignore.case = T, value = T)),
  "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY",
  "HALLMARK_OXIDATIVE_PHOSPHORYLATION",
  "HALLMARK_PEROXISOME",
  "HALLMARK_TGF_BETA_SIGNALING",
  "HALLMARK_GLYCOLYSIS",
  "HALLMARK_HYPOXIA", 
  "PID_NFAT_TFPATHWAY",
  "HALLMARK_INTERFERON_ALPHA_RESPONSE",
  "HALLMARK_INTERFERON_GAMMA_RESPONSE",
  "HALLMARK_IL6_JAK_STAT3_SIGNALING",
  "HALLMARK_DNA_REPAIR",
  "HALLMARK_IL2_STAT5_SIGNALING",
  "HALLMARK_APOPTOSIS", 
  "DUTTA_APOPTOSIS_VIA_NFKB",
  "WU_APOPTOSIS_BY_CDKN1A_VIA_TP53", 
  "WU_APOPTOSIS_BY_CDKN1A_NOT_VIA_TP53", 
  "TURJANSKI_MAPK1_AND_MAPK2_TARGETS",
  "TURJANSKI_MAPK8_AND_MAPK9_TARGETS",
  "TURJANSKI_MAPK7_TARGETS",                                             
  "TURJANSKI_MAPK11_TARGETS",                                            
  "TURJANSKI_MAPK14_TARGETS"    
)

checkSigID_unidir <- mSig[mSig$geneset %in% checkSigName_unidir, ]

library(org.Mm.eg.db)

checkSigID_unidir$Symbol <-
  annotate::getSYMBOL(as.character(checkSigID_unidir$entrez), data = 'org.Mm.eg')


##------ Collins' signatures

CollinsPath <- "/Users/mfor0011/Documents/projects/Monash/NK_Bulk/output/Collins/"

collinsSig <-
  read.table(
    paste0(
      CollinsPath,
      "Collins_Signatures_ILCs_CD56bright_CD56dim_GPCRs.txt"
    ),
    header = T,
    sep = "\t"
  )

## subset to up sets
collinsSig <- collinsSig[collinsSig$Direction == "Up", ]


collinsIDs <- biomaRt::getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = collinsSig$Genes,
  mart = martHuman,
  attributesL = c("mgi_symbol"),
  martL = martMouse
)

collinsIDs <- collinsIDs[! duplicated(collinsIDs$HGNC.symbol), ]
collinsIDs <- merge(collinsIDs, collinsSig, by.x = "HGNC.symbol", by.y = "Genes")


##----- res and exh signatures

outPathNK <- "/Users/mfor0011/Documents/projects/Monash/Foroutan_ResExh_CRC/output/NK/"


nkSigs <-
  read.table(
    paste0(outPathNK, "ResExhMarkers_NK_CLs_LCM.txt"),
    header = T,
    sep = "\t",
    stringsAsFactors = F,
    check.names = F
  )

markerIDs <- biomaRt::getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = nkSigs$gene,
  mart = martHuman,
  attributesL = c("mgi_symbol"),
  martL = martMouse
)

markerIDs <- markerIDs[! duplicated(markerIDs$HGNC.symbol), ]
markerIDs <- merge(markerIDs, nkSigs, by.x = "HGNC.symbol", by.y = "gene")

##------- get selected Immune signatures

getImmSigs <- c(
  # "TRM_TGFb_Unstim_Up", 
  #               "TRM_TGFbIL2_Unstim_Up", 
                "Proliferation_imsig"
  )

sigsImmuneIDs <- lapply(sigsImmune[getImmSigs],
                     function(x) {
                       biomaRt::getLDS(
                         attributes = c("hgnc_symbol"),
                         filters = "hgnc_symbol",
                         values = x,
                         mart = martHuman,
                         attributesL = c("mgi_symbol"),
                         martL = martMouse
                       )
                     })

sigsImmuneIDs_mouse <- lapply(sigsImmuneIDs, function(x){
  gg <- unique(x[, "MGI.symbol"])
  return(gg)
})

##---------- combine unidir signatures
 
AllSigs_unidir <- c(
  sigsImmuneIDs_mouse,
  list(
    Hypoxia_Hallmark = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_HYPOXIA"],
    ReactiveOx = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_REACTIVE_OXIGEN_SPECIES_PATHWAY"],
    OXPHOS = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_OXIDATIVE_PHOSPHORYLATION"],
    Peroxisome = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_PEROXISOME"],
    TGFb_Signaling = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_TGF_BETA_SIGNALING"],
    Glycolysis = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_GLYCOLYSIS"],
    
    IFNa_Response = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_INTERFERON_ALPHA_RESPONSE"],
    IFNg_Response = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_INTERFERON_GAMMA_RESPONSE"],
    PID_NFAT_TF = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "PID_NFAT_TFPATHWAY"],
    IL6_JAK_STAT3 = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_IL6_JAK_STAT3_SIGNALING"],
    IL2_STAT5 = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_IL2_STAT5_SIGNALING"],
    DNA_Repair = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_DNA_REPAIR"],
    Apoptosis = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "HALLMARK_APOPTOSIS"],
    Apoptosis_NFKB = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "DUTTA_APOPTOSIS_VIA_NFKB"],
    Apoptosis_CDKN1A_viaTP53 = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WU_APOPTOSIS_BY_CDKN1A_VIA_TP53"],
    Apoptosis_CDKN1A_noTP53 = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WU_APOPTOSIS_BY_CDKN1A_NOT_VIA_TP53"],
    
    MAPK1.2_Targets = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "TURJANSKI_MAPK1_AND_MAPK2_TARGETS"],
    MAPK8.9_Targets = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "TURJANSKI_MAPK8_AND_MAPK9_TARGETS"],
    MAPK7_Targets = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "TURJANSKI_MAPK7_TARGETS"],
    MAPK11_Targets = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "TURJANSKI_MAPK11_TARGETS"],
    MAPK14_Targets = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "TURJANSKI_MAPK14_TARGETS"],
    
    ## cell cycle
    CellCycle_literature = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WHITFIELD_CELL_CYCLE_LITERATURE"],
    G1_S = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WHITFIELD_CELL_CYCLE_G1_S"],
    S = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WHITFIELD_CELL_CYCLE_S"],
    G2 = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WHITFIELD_CELL_CYCLE_G2"],
    G2_M = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WHITFIELD_CELL_CYCLE_G2_M"],
    M_G1 = checkSigID_unidir$Symbol[checkSigID_unidir$geneset == "WHITFIELD_CELL_CYCLE_M_G1"],
    # 
    # NK_Exh = unique(markerIDs$MGI.symbol[markerIDs$Marker == "Exh"]),
    # NK_Res = unique(markerIDs$MGI.symbol[markerIDs$Marker == "Res"]),
    
    CD56bright = unique(collinsIDs$MGI.symbol[collinsIDs$Signature == "CD56bright"]),
    CD56dim = unique(collinsIDs$MGI.symbol[collinsIDs$Signature == "CD56dim"]),
    CD56dimCD57neg = unique(collinsIDs$MGI.symbol[collinsIDs$Signature == "CD56dimCD57neg"]),
    CD56dimCD57pos = unique(collinsIDs$MGI.symbol[collinsIDs$Signature == "CD56dimCD57pos"]),
    ILC1 = unique(collinsIDs$MGI.symbol[collinsIDs$Signature == "ILC1"]),
    ILC3 = unique(collinsIDs$MGI.symbol[collinsIDs$Signature == "ILC3"])
  )
)


AllSigsGeneSet_unidir <- lapply(AllSigs_unidir, GSEABase::GeneSet)
# AllSigsGeneSet_unidir <- lapply(AllSigs_unidir, GSEABase::GeneSetCollection)

AllSigsGeneSetName_unidir <- lapply(names(AllSigsGeneSet_unidir), function(x) {
  currentSig <- AllSigsGeneSet_unidir[[x]]
  GSEABase::setName(currentSig) <- x
  return(currentSig)
})

names(AllSigsGeneSetName_unidir) <- names(AllSigsGeneSet_unidir)

AllSigsCollection_unidir <- GSEABase::GeneSetCollection(AllSigsGeneSetName_unidir)



##============= bidirectional signatures

checkSigName_bidir <- c(
  "GSE22886_UNSTIM_VS_IL2_STIM_NKCELL_UP",
  "GSE22886_UNSTIM_VS_IL2_STIM_NKCELL_DN",
  "GSE22886_UNSTIM_VS_IL15_STIM_NKCELL_UP",
  "GSE22886_UNSTIM_VS_IL15_STIM_NKCELL_DN",
  
  "GSE7764_IL15_TREATED_VS_CTRL_NK_CELL_24H_UP",
  "GSE7764_IL15_TREATED_VS_CTRL_NK_CELL_24H_DN",
  "GSE7764_IL15_NK_CELL_24H_VS_SPLENOCYTE_UP",
  "GSE7764_IL15_NK_CELL_24H_VS_SPLENOCYTE_DN",
  "GSE22886_IL2_VS_IL15_STIM_NKCELL_UP",
  "GSE22886_IL2_VS_IL15_STIM_NKCELL_DN",
  "GSE12198_CTRL_VS_HIGH_IL2_STIM_NK_CELL_UP",                
  "GSE12198_CTRL_VS_HIGH_IL2_STIM_NK_CELL_DN",                
  "GSE12198_CTRL_VS_LOW_IL2_STIM_NK_CELL_UP",                 
  "GSE12198_CTRL_VS_LOW_IL2_STIM_NK_CELL_DN",  
  "GSE12198_LOW_IL2_STIM_NK_CELL_VS_HIGH_IL2_STIM_NK_CELL_UP",
  "GSE12198_LOW_IL2_STIM_NK_CELL_VS_HIGH_IL2_STIM_NK_CELL_DN",
  "GSE22919_RESTING_VS_IL2_IL12_IL15_STIM_NK_CELL_UP",
  "GSE22919_RESTING_VS_IL2_IL12_IL15_STIM_NK_CELL_DN",
  
  "HAMAI_APOPTOSIS_VIA_TRAIL_UP", 
  "HAMAI_APOPTOSIS_VIA_TRAIL_DN", 
  "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_SMAD4_UP", 
  "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_SMAD4_DN", 
  "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_MAPK1_UP", 
  "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_MAPK1_DN", 
  
  "YOSHIMURA_MAPK8_TARGETS_UP",
  "YOSHIMURA_MAPK8_TARGETS_DN"
)


checkSigID_bidir <- mSig[mSig$geneset %in% checkSigName_bidir, ]

checkSigID_bidir$Symbol <-
  annotate::getSYMBOL(as.character(checkSigID_bidir$entrez), data = 'org.Mm.eg')



AllSigs_bidir <- list(
  M1_Imm = list(
    M1_Imm_Up = m1_im$Symbol[m1_im$logFC > 0],
    M1_Imm_Dn = m1_im$Symbol[m1_im$logFC < 0]),
  M2_Imm = list(
    M2_Imm_Up = m2_im$Symbol[m2_im$logFC > 0],
    M2_Imm_Dn = m2_im$Symbol[m2_im$logFC < 0]),
  M2_M1 = list(
    M2_M1_Up = m2_m1$Symbol[m2_m1$logFC > 0],
    M2_M1_Dn = m2_m1$Symbol[m2_m1$logFC < 0]),
  Unstim_IL2 = list(
      Unstim_IL2_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22886_UNSTIM_VS_IL2_STIM_NKCELL_UP"], 
      Unstim_IL2_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22886_UNSTIM_VS_IL2_STIM_NKCELL_DN"]), 
  Unstim_IL15 = list(
    Unstim_IL15_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22886_UNSTIM_VS_IL15_STIM_NKCELL_UP"],
    Unstim_IL15_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22886_UNSTIM_VS_IL15_STIM_NKCELL_DN"]),
  IL15_Ctrl = list(
    IL15_Ctrl_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE7764_IL15_TREATED_VS_CTRL_NK_CELL_24H_UP"],
    IL15_Ctrl_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE7764_IL15_TREATED_VS_CTRL_NK_CELL_24H_DN"]),
  IL15_Splenocyte = list(
    IL15_spl_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE7764_IL15_NK_CELL_24H_VS_SPLENOCYTE_UP"],
    IL15_spl_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE7764_IL15_NK_CELL_24H_VS_SPLENOCYTE_DN"]),
  IL2.vs.IL15 = list(
    IL2.vs.IL15_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22886_IL2_VS_IL15_STIM_NKCELL_UP"],
    IL2.vs.IL15_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22886_IL2_VS_IL15_STIM_NKCELL_DN"]),
  Ctrl_IL2High = list(
    Ctrl_IL2High_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE12198_CTRL_VS_HIGH_IL2_STIM_NK_CELL_UP"],
    Ctrl_IL2High_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE12198_CTRL_VS_HIGH_IL2_STIM_NK_CELL_DN"]),
  Ctrl_IL2Low = list(
    Ctrl_IL2Low_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE12198_CTRL_VS_LOW_IL2_STIM_NK_CELL_UP"],
    Ctrl_IL2Low_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE12198_CTRL_VS_LOW_IL2_STIM_NK_CELL_DN"]),
  IL2Low_IL2High = list(
    IL2Low_IL2High_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE12198_LOW_IL2_STIM_NK_CELL_VS_HIGH_IL2_STIM_NK_CELL_UP"],
    IL2Low_IL2High_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE12198_LOW_IL2_STIM_NK_CELL_VS_HIGH_IL2_STIM_NK_CELL_DN"]),
  Rest_IL2.12.15 = list(
    Rest_IL2.12.15_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22919_RESTING_VS_IL2_IL12_IL15_STIM_NK_CELL_UP"],
    Rest_IL2.12.15_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "GSE22919_RESTING_VS_IL2_IL12_IL15_STIM_NK_CELL_DN"]),
  
  Apoptosis_TRAIL = list(
    Apoptosis_TRAIL_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "HAMAI_APOPTOSIS_VIA_TRAIL_UP"],
    Apoptosis_TRAIL_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "HAMAI_APOPTOSIS_VIA_TRAIL_DN"]),
  Apoptosis_TGFB1_SMAD4 = list(
    Apoptosis_TGFB1_SMAD4_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_SMAD4_UP"],
    Apoptosis_TGFB1_SMAD4_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_SMAD4_DN"]),
  Apoptosis_TGFB1_MAPK1 = list(
    Apoptosis_TGFB1_MAPK1_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_MAPK1_UP"],
    Apoptosis_TGFB1_MAPK1_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "RAMJAUN_APOPTOSIS_BY_TGFB1_VIA_MAPK1_DN"]),
  MAPK8_Targets = list(
    MAPK8_Targets_Up = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "YOSHIMURA_MAPK8_TARGETS_UP"],
    MAPK8_Targets_Dn = checkSigID_bidir$Symbol[checkSigID_bidir$geneset == "YOSHIMURA_MAPK8_TARGETS_DN"])
)


```

# Score Ikzf in vitro data
```{r}
library(singscore)
rankedIkzf <- rankGenes(ex)

##------ unidir signatures
uniDirScore_Ikzf <- multiScore(rankedIkzf, 
                         upSetColc = AllSigsCollection_unidir, 
                         knownDirection = T, 
                         centerScore = T)
uniDirScore_Ikzf <-
  data.frame(t(uniDirScore_Ikzf$Scores),
             check.rows = F,
             check.names = F)


##------ bidir signatures
biDirScore_Ikzf <- sapply(names(AllSigs_bidir), function(x) {
  currentSig <- AllSigs_bidir[[x]]
  
  currentUp <- currentSig[grep("Up", names(currentSig), value = T)][[1]]
  currentUp <- currentUp[! duplicated(currentUp)]
  
  currentDn <- currentSig[grep("Dn", names(currentSig), value = T)][[1]]
  currentDn <- currentDn[! duplicated(currentDn)]
  
  ss <- singscore::simpleScore(
    rankData = rankedIkzf,
    upSet = currentUp,
    downSet = currentDn,
    knownDirection = T,
    centerScore = T
  )$TotalScore
  return(ss)
})


allScores <- data.frame(cbind(uniDirScore_Ikzf, biDirScore_Ikzf))

allScores$Group <-
  c(rep("IKZF3_KO", 4),
    rep("IKZF1_IKZF3_KO", 3),
    rep("IKZF1_KO", 4),
    rep("Cre", 4))

# saveRDS(allScores, paste0(outPath, "Scores_IKZF.RDS"))
# write.csv(allScores, paste0(outPath, "Scores_IKZF.csv"), row.names = T)

# allScores <- readRDS(paste0(outPath, "Scores_IKZF.RDS"))
```


```{r}
badSamples <- c("RPKM_IKZF3_KO_S3", "RPKM_IKZF3_KO_S4")
allScores_sub <- allScores[! rownames(allScores) %in% badSamples, ]
allScores_sub_Long <-  allScores_sub %>% 
  pivot_longer(names_to = "Signature", values_to = "Scores", cols = 1:(ncol(allScores_sub) - 1))

allScores_sub_Long$Group <- factor(allScores_sub_Long$Group, levels = c(
  "Cre", "IKZF1_KO", "IKZF3_KO", "IKZF1_IKZF3_KO"
))

png(
  paste0(figPath, "Boxplot_Scores_All_rm2Samples.png"),
  height = 700,
  width = 1100
)
ggplot(allScores_sub_Long, 
       aes(x = Group, y = Scores, color = Group)) +
  geom_boxplot() +
  geom_jitter(position=position_jitter(0.2)) +
  facet_wrap(~ Signature, scales = "free") +
  scale_color_manual(values = brewer.pal(4, "Dark2")) +
  theme_bw() +
  theme(
    axis.title.x = element_blank(), 
    axis.text.x = element_blank()
        ) 
dev.off()
```


## Heatmap of scores
```{r}

pdf(
  paste0(figPath, "Heatmap_Scores_All_rm2Samples.pdf"),
  height = 10,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores[!grepl("S3|S4", rownames(allScores)),
                                    -ncol(allScores)]),
                        cluster_columns = F,
                        col = viridis::magma(100), 
                        name = "Score")
dev.off()
```

```{r}

pdf(
  paste0(figPath, "Heatmap_Scores_Maturity_rm2Samples.pdf"),
  height = 6,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores[
   !grepl("S3|S4", rownames(allScores)),
   c(
  "M1_Imm",
  "M2_Imm",
  "M2_M1",
  "CD56bright",
  "CD56dim",
  "CD56dimCD57neg",
  "CD56dimCD57pos",
  "ILC1",
  "ILC3"
)]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()


pdf(
  paste0(figPath, "Heatmap_Scores_Apoptosis_Prolif_rm2Samples.pdf"),
  height = 5,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores[
  !grepl("S3|S4", rownames(allScores)), 
  grepl("Apoptosis|Proliferation", colnames(allScores))]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()

pdf(
  paste0(figPath, "Heatmap_Scores_MAPKs_rm2Samples.pdf"),
  height = 5,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores[
   !grepl("S3|S4", rownames(allScores)),
   grepl("MAPK", colnames(allScores))]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()

pdf(
  paste0(figPath, "Heatmap_Scores_ILs_rm2Samples.pdf"),
  height = 5,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores[
   !grepl("S3|S4", rownames(allScores)),
   grepl("IL2|IL15", colnames(allScores))]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()


```




# Score Ikzf ex-vivo data
```{r}
library(singscore)
rankedIkzf_ex <- rankGenes(ex_ex)

##------ unidir signatures
uniDirScore_Ikzf_ex <- multiScore(rankedIkzf_ex, 
                         upSetColc = AllSigsCollection_unidir, 
                         knownDirection = T, 
                         centerScore = T)
uniDirScore_Ikzf_ex <-
  data.frame(t(uniDirScore_Ikzf_ex$Scores),
             check.rows = F,
             check.names = F)


##------ bidir signatures
biDirScore_Ikzf_ex <- sapply(names(AllSigs_bidir), function(x) {
  currentSig <- AllSigs_bidir[[x]]
  
  currentUp <- currentSig[grep("Up", names(currentSig), value = T)][[1]]
  currentUp <- currentUp[! duplicated(currentUp)]
  
  currentDn <- currentSig[grep("Dn", names(currentSig), value = T)][[1]]
  currentDn <- currentDn[! duplicated(currentDn)]
  
  ss <- singscore::simpleScore(
    rankData = rankedIkzf_ex,
    upSet = currentUp,
    downSet = currentDn,
    knownDirection = T,
    centerScore = T
  )$TotalScore
  return(ss)
})


allScores_ex <- data.frame(cbind(uniDirScore_Ikzf_ex, biDirScore_Ikzf_ex))


allScores_ex$Group <-
  c(rep("Cre", 3),
    rep("IKZF1_KO", 3))

saveRDS(allScores_ex, paste0(outPath, "Scores_All_IKZF_ExVivo.RDS"))
write.csv(allScores_ex, paste0(outPath, "Scores_All_IKZF_ExVivo.csv"), row.names = T)

```


## Heatmap of scores
```{r}
pdf(
  paste0(figPath, "Heatmap_Scores_uniDir_Ikzf1_Exvivo.pdf"),
  height = 10,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores_ex[, -ncol(allScores_ex)]),
                        cluster_columns = F,
                        col = viridis::magma(100), 
                        name = "Score")
dev.off()
```

```{r}

pdf(
  paste0(figPath, "Heatmap_Scores_Maturity_Exvivo.pdf"),
  height = 5,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores_ex[,
   c(
  "M1_Imm",
  "M2_Imm",
  "M2_M1",
  "CD56bright",
  "CD56dim",
  "CD56dimCD57neg",
  "CD56dimCD57pos",
  "ILC1",
  "ILC3"
)]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()


pdf(
  paste0(figPath, "Heatmap_Scores_Apoptosis_Prolif_Exvivo.pdf"),
  height = 5,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores_ex[, 
  grepl("Apoptosis|Proliferation", colnames(allScores_ex))]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()

pdf(
  paste0(figPath, "Heatmap_Scores_MAPKs_Exvivo.pdf"),
  height = 5,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores_ex[,
   grepl("MAPK", colnames(allScores_ex))]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()

pdf(
  paste0(figPath, "Heatmap_Scores_ILs_Exvivo.pdf"),
  height = 5,
  width = 6
)
ComplexHeatmap::Heatmap(t(allScores_ex[,
   grepl("IL2|IL15", colnames(allScores_ex))]),
cluster_columns = F, 
col = viridis::magma(100),
name = "Scores")
dev.off()
```


```{r}
allScores_ex_long <-
  pivot_longer(
    allScores_ex,
    cols = 1:(ncol(allScores_ex) - 1),
    names_to = "Signatures",
    values_to = "Score"
  )

pdf(
  paste0(figPath, "Boxplot_Scores_All_Exvivo.pdf"),
  height = 10,
  width = 15
)
ggplot(allScores_ex_long, aes(x = Group, y = Score, color = Group)) +
  geom_boxplot() +
  facet_wrap( ~ Signatures, scales = "free", ncol = 7) +
  scale_color_manual(values = currentcol[3:4]) +
  theme_bw() +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank())
dev.off()
```


# ChIP-seq and Cut and Run
For I kzf1, we had ChIP-seq data with three replicates, whiel for the Ikzf3, we had Cut&Run data with one sample. So, here, I read in the results of MACS peak calling for each of the Ikzf1 replicates along with the csaw results for the same data, as well as the results of the MACS and csaw for Ikzf3 data. 

Original number of genes in CSAW: 259
Common genes in different replicates using MACS: 589 genes [11 up direction]
Common calls from CSAW and MACS2: there are 133 genes.
CSAW only calls: 123
MACS only calls: 456

## Ikzf1/kzf3
```{r}
csaw <- read.csv(
  paste0(
    chipPath,
    "180510 DBsites IKZF1-WTvsKO annotatedToClosestGene_voom_consolidateWindowSizes.csv"
  ),
  stringsAsFactors = F
)

macs1 <- read.csv(
  paste0(
    chipPath,
    "180510 MACS2 peakcalls IKZF1-WT AUTOext_R1.csv"
  ),
  stringsAsFactors = F
)

macs2 <- read.csv(
  paste0(
    chipPath,
    "180510 MACS2 peakcalls IKZF1-WT AUTOext_R2.csv"
  ),
  stringsAsFactors = F
)

macs4 <- read.csv(
  paste0(
    chipPath,
    "180510 MACS2 peakcalls IKZF1-WT AUTOext_R4.csv"
  ),
  stringsAsFactors = F
)

##----- read MACS2 outcome from CnR for Ikzf3
macsIkzf3 <- read.csv(
  paste0(outPath, "annotated_Ikzf3_peak_set_narrow.csv"),
  stringsAsFactors = F,
  check.names = F
)

# macsIkzf3_nf <- read.csv(
#   paste0(outPath, "annotated_Ikzf3_peak_set_narrow_nf.csv"),
#   stringsAsFactors = F,
#   check.names = F
# )

##----- read CSAW for Ikzf3
csaw3 <- read.table(
  paste0(outPath, "Ikzf3_CnR_CSAW_fraglen195_combFDR_ConsWin50-200_normFactors.txt"),
  header = T, 
  sep = "\t"
)

## 768 genes, all up
csaw3 <- csaw3[! duplicated(csaw3$symbol), ]


##----- read edgeR results for Ikzf3
# edg3 <- read.table(
#   paste0(outPath, "Ikzf3_CnR_FeatureCount_edgeR_DB_stats.txt"),
#   header = T, 
#   sep = "\t"
# )
# 
# edg3 <- head(edg3, 1400)

```


## Ikzf2 
Quick examination of the overlapping genes between our results and a publicly available data of IKZF2 data.

### Read gene lists from the paper
The gene lists were taken from the paper by Park *et al* "IKZF2 Drives Leukemia Stem Cell Self-Renewal and Inhibits Myeloid Differentiation", *Cell Stem Cell*, 2019. 

```{r}
# DEGs in Ikzf2 KO vs Ctrl
DEGs_Ctrl_Ikzf2 <-
  read.csv(
    paste0(dataPath, "Ikzf2_CutnRun_GSE120623_RAW/DEGs_IKZF2_KO.csv"),
    stringsAsFactors = F
  )

DEGs_Ikzf2 <- biomaRt::getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = DEGs_Ctrl_Ikzf2$Genes,
  mart = martHuman,
  attributesL = c("mgi_symbol"),
  martL = martMouse
)

DEGs_Ikzf2 <- DEGs_Ikzf2[! duplicated(DEGs_Ikzf2$HGNC.symbol), ]
DEGs_Ikzf2 <-
  merge(DEGs_Ikzf2, DEGs_Ctrl_Ikzf2, by.x = "HGNC.symbol", by.y = "Genes")


DA_atacseq <-
    read.csv(
    paste0(dataPath, "Ikzf2_CutnRun_GSE120623_RAW/ATAC_seq_DiffAccess.csv"),
    stringsAsFactors = F
  )
```

**Genes upregulated in the RNA-seq (Log2FC>0.75, pval<0.05) and with increased accessibility (Log2FC>1, pval<0.05) and have a Cut and Run Peaks**
```{r}
rna_atac_cr <- as.character(quote(
  c(
    Mpo,
    Prom1,
    Elane,
    C3,
    Ralgps2,
    Chil3,
    Chil3,
    Hsd11b1,
    Mrgpra2b,
    Card10,
    Cited2,
    Rtkn2,
    Gphn,
    Cd177,
    Anxa3,
    Fpr2,
    Cracr2b,
    Rybp,
    Nampt,
    Nrg1,
    Slfn2,
    Rnf144b,
    Kctd12,
    Ypel5,
    Ypel5,
    Cdk5rap1,
    Pag1,
    Mgst2,
    S100a8,
    S100a9,
    Naaa,
    C5ar2,
    Dock1,
    Ngp,
  )
)) [-1]

intersect(rna_atac_cr, DEGs_DKO$external_gene_name)
 # [1] "C3"      "Ralgps2" "Card10"  "Cd177"   "Cracr2b" "Rybp"    "Rnf144b" "Pag1"   
 # [9] "Mgst2"   "C5ar2"   "Dock1" 
```


**Genes upregulated in the RNA-seq (Log2FC>0.75, pval<0.05) and with increased accessibility (Log2FC>1, pval<0.05) and have a Cut and Run Peaks and predicted to have CEBP and IKZF2 motifs**
```{r}
rna_atac_cr_motifs.Ikzf2.cebp <- as.character(quote(
  c(
    C3,
    CDK5RAP1,
    CITED2,
    FPR2,
    KCTD12,
    MGST2,
    MPO,
    NAMPT,
    PAG1,
    RTKN2,
    RYBP,
    S100A9,
    SLFN2
  )
))[-1]



rna_atac_cr_motifs.Ikzf2.cebp <- biomaRt::getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = rna_atac_cr_motifs.Ikzf2.cebp,
  mart = martHuman,
  attributesL = c("mgi_symbol"),
  martL = martMouse
)

rna_atac_cr_motifs.Ikzf2.cebp <- rna_atac_cr_motifs.Ikzf2.cebp$MGI.symbol
# rna_atac_cr_motifs.Ikzf2.cebp <- rna_atac_cr_motifs.Ikzf2.cebp[! duplicated(rna_atac_cr_motifs.Ikzf2.cebp$HGNC.symbol), ]

intersect(rna_atac_cr_motifs.Ikzf2.cebp, DEGs_DKO$external_gene_name)
 # "Mgst2" "C3"    "Pag1" "Rybp
```


**Raw data of motif enrichment plot, conditional on Ikzf2 motif and filtered for significant motifs**
```{r}
motif.Ikzf2.signif <- as.character(quote(
  c(
    Nfil3,
    Cebpd,
    Hmg20b,
    Srf,
    Mybl2,
    Creb1,
    Jdp2,
    Cux1,
    Cenpb,
    Hsf1,
    Xbp1,
    Mbd1,
    Mef2c,
    Trp53,
    E4f1,
    Ddit3,
    Cebpe,
    Srebf2,
    Tfdp1,
    Irf5,
    Elk3,
    Mecom,
    Etv6,
    Elf4,
    Elf2,
    Atf1,
    Hhex,
    Atf6,
    Hoxa9
  )
))[-1]

intersect(motif.Ikzf2.signif, DEGs_DKO$external_gene_name)
 # "Nfil3" (down-regulate) "Cebpd" (up-regulate) "Hhex" (down-regulate) 

# Nfil3/E4bp4 is required for the development and maturation of NK cells in vivo

#  [1] "Nfil3"  "Cebpd"  "Hmg20b" "Mybl2"  "Creb1"  "Mef2c"  "Tfdp1"  "Elk3"   "Etv6"  
# [10] "Elf4"   "Elf2"   "Hhex"   "Atf6"  
```


**Genes upregulated in the RNA-seq (Log2FC>0.75, pval<0.05) and with increased accessibility (Log2FC>1, pval<0.05)**
```{r}
rna_atac <- as.character(quote(
  c(
    NAMPT,
    CD177,
    YPEL5,
    GGTA1,
    CLEC4A2,
    ANXA3,
    NAAA,
    RTKN2,
    CRACR2B,
    CITED2,
    CARD10,
    CDK5RAP1,
    RNF144B,
    HOMER1,
    KCTD12,
    C5AR2,
    S100A9,
    NRG1,
    RALGPS2,
    MRGPRA2B,
    YPEL5,
    NGP,
    SLFN2,
    SLCO4C1,
    HSD11B1,
    DOCK1,
    CHIL3,
    S100A8,
    RYBP,
    TLR4,
    C3,
    PAG1,
    GPHN,
    FPR2,
    MGST2,
    ELANE,
    KLF3,
    PTPRR,
    PROM1,
    MPO
  )
))[-1]

rna_atac <- biomaRt::getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = rna_atac,
  mart = martHuman,
  attributesL = c("mgi_symbol"),
  martL = martMouse
)

rna_atac <- rna_atac$MGI.symbol

intersect(rna_atac, DA_atacseq$symbol)
 # "Dock1"   "S100a8"  "Nampt"   "Elane"   "Cited2"  "Hsd11b1" "Rtkn2"  
intersect(rna_atac, DEGs_DKO$external_gene_name)
 # [1] "Rnf144b" "C5ar2"   "C3"      "Homer1"  "Cd177"   "Rybp"    "Pag1"    "Mgst2"  
 # [9] "Card10"  "Ralgps2" "Klf3"    "Dock1"   "Cracr2b"
```


### Read CutnRun peaks
Read in the [raw Cut and Run data](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120623).
I filter for those with high counts and subset to the overlapping ranges, and then annotate peaks, and export them.
```{r}
library(rtracklayer)
# library(magrittr)
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(ggupset) 


txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene

fs <- grep("*.txt$", list.files(paste0(dataPath, "Ikzf2_CutnRun_GSE120623_RAW")), value = T)

##------- read pairs of technical reps for each biological samples
## Then filter for those with high counts
## Subset to the overlapping ranges, and then annotate peaks, and export them

for(i in c("HELIOS1",
           "HELIOS2",
           "HELIOS3")) {
  fs0 <- fs[grepl(i, fs)]
  
  pk1 <- read.delim(paste0(dataPath,
                           "Ikzf2_CutnRun_GSE120623_RAW/", fs0[1]),
                    header = T)
  ## filter to peaks with high counts
  pk1 <- pk1[pk1$count > quantile(pk1$count, probs = 0.97),]
  
  
  pk2 <- read.delim(paste0(dataPath,
                           "Ikzf2_CutnRun_GSE120623_RAW/", fs0[2]),
                    header = T)
  
  pk2 <- pk2[pk2$count > quantile(pk2$count, probs = 0.97),]
  
  colnames(pk1) <- colnames(pk2) <-
    c("chr",
      "start",
      "end",
      "name",
      "count")
  
  pk1 <- makeGRangesFromDataFrame(
    df = pk1,
    keep.extra.columns = TRUE,
    ignore.strand = FALSE,
    seqinfo = NULL,
    seqnames.field = c(
      "seqnames",
      "seqname",
      "chromosome",
      "chrom",
      "chr",
      "chromosome_name",
      "seqid"
    ),
    start.field = "start",
    end.field = c("end", "stop"),
    strand.field = "strand",
    starts.in.df.are.0based = FALSE
  )
  
  
  pk2 <- makeGRangesFromDataFrame(
    df = pk2,
    keep.extra.columns = TRUE,
    ignore.strand = FALSE,
    seqinfo = NULL,
    seqnames.field = c(
      "seqnames",
      "seqname",
      "chromosome",
      "chrom",
      "chr",
      "chromosome_name",
      "seqid"
    ),
    start.field = "start",
    end.field = c("end", "stop"),
    strand.field = "strand",
    starts.in.df.are.0based = FALSE
  )
  
  pk0 <- subsetByOverlaps(pk1, pk2)
  
  annotatedPeaks <- annotatePeak(
    pk0,
    tssRegion = c(-3000, 1000),
    TxDb = txdb,
    annoDb = "org.Mm.eg.db"
  )
  
  pk0_granges <- as.GRanges(annotatedPeaks)
  
  export(pk0_granges,
         paste0(outPath, "annotated_Ikzf2_", i, ".gtf"),
         format = "GTF")
  
  readr::write_csv(as.data.frame(pk0_granges),
                   paste0(outPath, "annotated_Ikzf2_", i, ".csv"))
  
}


```

Now we read what we exported above to compare the biological replicates.
```{r}

##------------

pk1 <- import(paste0(outPath, "annotated_Ikzf2_HELIOS1.gtf"),
         format = "GTF")

pk2 <- import(paste0(outPath, "annotated_Ikzf2_HELIOS2.gtf"),
         format = "GTF")

pk3 <- import(paste0(outPath, "annotated_Ikzf2_HELIOS3.gtf"),
         format = "GTF")

comIkzf2 <- Reduce(intersect, list(pk1$SYMBOL, pk2$SYMBOL, pk3$SYMBOL))
## 130 genes common across all

#   [1] "Arid5a"        "Mreg"          "Pecr"          "A630001G21Rik" "Actr3"        
#   [6] "Cxcr4"         "Srgap2"        "Ndufs2"        "Alyref2"       "2310043L19Rik"
#  [11] "Hhat"          "Ifngr1"        "Srgn"          "Cirbp"         "Dot1l"        
#  [16] "Gng7"          "Mdm1"          "Camk2b"        "Fbxo48"        "G3bp1"        
#  [21] "Snord118"      "Per1"          "Wrap53"        "Polr2a"        "Tex14"        
#  [26] "Pnpo"          "Cwc25"         "Psmd3"         "Stat3"         "Rdm1"         
#  [31] "Arl4d"         "Taco1"         "Cd300a"        "Pum2"          "Hbp1"         
#  [36] "Trappc6b"      "Hspa2"         "Lin52"         "Lman2"         "Vcl"          
#  [41] "Parp2"         "Mtmr6"         "Ints9"         "Dok2"          "Eny2"         
#  [46] "Trappc9"       "Ankrd54"       "Gtpbp1"        "Lncppara"      "Ap2m1"        
#  [51] "4930517M08Rik" "Ppp2r1a"       "Zfp945"        "Rab40c"        "Fgd2"         
#  [56] "Rnf5"          "Gpr108"        "Vmn1r238"      "Mex3c"         "Arhgap19"     
#  [61] "Sufu"          "Csf2ra"        "Dclre1c"       "Rsu1"          "Clp1"         
#  [66] "Hsd17b12"      "Ncoa3"         "Ube2v1"        "A530013C23Rik" "Prcc"         
#  [71] "Vmn1r2"        "Chchd7"        "Plekhf2"       "Otud6b"        "Ripk2"        
#  [76] "Casp8ap2"      "Aco1"          "Kif24"         "Dnaic1"        "Tln1"         
#  [81] "Hint2"         "Zfp462"        "Ptpn3"         "Ifna13"        "Bsdc1"        
#  [86] "Runx3"         "Camta1"        "Grk4"          "Cpeb2"         "Tpcn1"        
#  [91] "Mir721"        "AA545190"      "Paip2b"        "Aplf"          "Brpf1"        
#  [96] "Foxj2"         "Ano2"          "Gm8882"        "Etv6"          "Vmn1r77"      
# [101] "Rps19"         "Sertad3"       "Hnrnpl"        "Psmd8"         "Cyth2"        
# [106] "Uvrag"         "C2cd3"         "Nfatc2ip"      "Dctpp1"        "Cenpu"        
# [111] "Pbx4"          "Gatad2a"       "Mast3"         "Asf1b"         "Rfx1"         
# [116] "Lyl1"          "Gcdh"          "Katnb1"        "Meak7"         "Cdh15"        
# [121] "A630001O12Rik" "Hinfp"         "Gm5122"        "Parp16"        "Snx1"         
# [126] "Pik3cb"        "Nbeal2"        "4933407K13Rik" "G530011O06Rik" "Erdr1"       

unionIkzf2 <- unique(c(pk1$SYMBOL, pk2$SYMBOL, pk3$SYMBOL)) ## 1781

write.csv(
  data.frame(unionIkzf2),
  paste0(outPath, "Ikzf2_targetGenes_UnionList_1781Genes.csv"),
  row.names = F
)
write.csv(
  data.frame(comIkzf2),
  paste0(outPath, "Ikzf2_targetGenes_CommonList_130Genes.csv"),
  row.names = F
)

```

### Intersect: Ikzf2 Targets and other genes
```{r}
## intersect between the common genes in Ikzf2 and DEGs in DK:
comIkzf2_DEGsKO <- intersect(comIkzf2, DEGs_DKO$external_gene_name)
#  [1] "Cxcr4"    "Ndufs2"   "Hhat"     "Srgn"     "Mdm1"    
#  [6] "Fbxo48"   "Per1"     "Polr2a"   "Cwc25"    "Stat3"   
# [11] "Arl4d"    "Taco1"    "Trappc6b" "Hspa2"    "Lin52"   
# [16] "Vcl"      "Parp2"    "Dok2"     "Arhgap19" "Rsu1"    
# [21] "Clp1"     "Ncoa3"    "Prcc"     "Plekhf2"  "Aco1"    
# [26] "Tln1"     "Zfp462"   "Ptpn3"    "Bsdc1"    "Runx3"   
# [31] "Tpcn1"    "Paip2b"   "Aplf"     "Brpf1"    "Foxj2"   
# [36] "Etv6"     "Cyth2"    "C2cd3"    "Nfatc2ip" "Dctpp1"  
# [41] "Cenpu"    "Pbx4"     "Gatad2a"  "Mast3"    "Asf1b"   
# [46] "Gcdh"     "Pik3cb" 

## intersect between the union genes in Ikzf2 and DEGs in DK:
length(intersect(unionIkzf2, DEGs_DKO$external_gene_name)[order(intersect(unionIkzf2, DEGs_DKO$external_gene_name))] ) ## 615 genes


# Genes that are common between the Atac-seq and CutnRun
intersect(DA_atacseq$symbol, DEGs_DKO$external_gene_name)[order(intersect(DA_atacseq$symbol, DEGs_DKO$external_gene_name))]
#  [1] "Acpp"     "Aff3"     "Akap1"    "Aldh1l1"  "Apbb2"   
#  [6] "Asph"     "Atf7ip"   "Bcr"      "Btrc"     "C2cd5"   
# [11] "C77080"   "Cacybp"   "Cers6"    "Chac1"    "Chd2"    
# [16] "Coro1c"   "Crlf2"    "Cstad"    "Dapk1"    "Dcaf6"   
# [21] "Dock1"    "Efcab11"  "Efnb2"    "Fam118b"  "Filip1"  
# [26] "Foxn3"    "Foxp1"    "Ftx"      "Fyb"      "Gas7"    
# [31] "Ggta1"    "Glrp1"    "Gm15326"  "Gstz1"    "Iqgap1"  
# [36] "Iqgap2"   "Irx3"     "Jak1"     "Jazf1"    "Jun"     
# [41] "Klf13"    "Lemd2"    "Lrrc40"   "Mall"     "Mapre3"  
# [46] "Mbd2"     "Mcur1"    "Mef2c"    "Mical3"   "Micall2" 
# [51] "Mtus1"    "N4bp1"    "Nbea"     "Nfatc2"   "Nmnat1"  
# [56] "Nucb2"    "Ogdhl"    "Pbx4"     "Pfkp"     "Pgm1"    
# [61] "Plcb4"    "Plxnd1"   "Prim2"    "Prkx"     "Ramp3"   
# [66] "Rara"     "Rgs3"     "Rragb"    "Sdc1"     "Serpine2"
# [71] "Setd3"    "Sgms2"    "Slc25a13" "Smim3"    "Snx13"   
# [76] "Sorcs2"   "Spef2"    "Srgn"     "Syne1"    "Tbc1d8"  
# [81] "Tle3"     "Tnfrsf9"  "Trim7"    "Tspan18"  "Usp31"   
# [86] "Vav1"     "Vav2"     "Wdr26"    "Zcchc2"  


# Genes that are common between the Atac-seq, CutnRun and RNA-seq (DKO):
xx <- intersect( intersect(DA_atacseq$symbol, DEGs_DKO$external_gene_name), intersect(unionIkzf2, DEGs_DKO$external_gene_name)[order(intersect(unionIkzf2, DEGs_DKO$external_gene_name))])
## 20 genes: 
#  [1] "Jak1"   "C77080" "Asph"   "Gas7"   "Acpp"   "Mbd2"  
#  [7] "Iqgap1" "Cacybp" "Chd2"   "Atf7ip" "Pbx4"   "Jazf1" 
# [13] "Dapk1"  "Mtus1"  "Zcchc2" "Chac1"  "Mef2c"  "Mcur1" 
# [19] "Lemd2"  "Srgn"  

# "C77080" (uncharacterised)
# "Asph" (ASPH [from cancer cells] suppresses the natural killer (NK) cell-surveillance activity) 
# "Gas7" (Growth arrest-specific 7)   
# "Acpp" (up in DNAM-1 + cells) - Nick's paper https://www.cell.com/cell-reports/pdfExtended/S2211-1247(15)00259-4
# "Pbx4" (not much known - embryonic development and cellular differentiation)
# "Jazf1" (Plays a role in glucose homeostasis by improving glucose metabolism)
# "Dapk1" (positive mediator of gamma-interferon induced programmed cell death)
# "Mtus1" (inhibit ERK2 activation and cell proliferation)
# "Chac1" (pro-apoptotic component )  
# "Srgn" (granzymes and perforin mediated apop)

x_atac <- DA_atacseq[DA_atacseq$symbol  %in% xx, c("symbol", "logFC")]
x_expr <- DEGs_DKO[DEGs_DKO$external_gene_name  %in%  xx, c("external_gene_name", "logFC")]

x <- merge(x_atac, x_expr, by.x = "symbol", by.y = "external_gene_name")
plot(x$logFC.x, x$logFC.y)

```

#### Check DE stats of sleected genes
```{r}
DEstats_comIkzf2_DEGsKO <- DEstats[DEstats$Symbol %in% comIkzf2_DEGsKO, 
                                    grepl("Symbol|adj|logFC", colnames(DEstats))]


write.csv(
  DEstats_comIkzf2_DEGsKO,
  # paste0(outPath, "DEstats_comIkzf2_DEGsKO_14CommonGenes.csv"),
  paste0(outPath, "DEstats_comIkzf2_DEGsKO_47CommonGenes.csv"),
  row.names = F
)
```



## IKzf1/3 common/union peaks

For Ikzf1:
```{r}
## common between three Reps in MACS

## 2714 genes
macsUnion <- unique(c(
  macs1$external_gene_name,
  macs2$external_gene_name,
  macs4$external_gene_name
))

## 589 we decided go with union of the peaks
# macsCom <- Reduce(
#   intersect,
#   list(
#     macs1$external_gene_name,
#     macs2$external_gene_name,
#     macs4$external_gene_name
#   )
# )

## common between MACS and CSAW
# comCalls <- intersect(csaw$external_gene_name, macsCom)
comCalls <- intersect(csaw$external_gene_name, macsUnion)  ## 195 common
unionCalls <- unique(c(csaw$external_gene_name, macsUnion))  ## 2775 common

## Method specific calls
## 61 
csawOnly <- unique(
  csaw$external_gene_name[! csaw$external_gene_name %in% macsUnion])

# macsOnly <- unique(
#   macsCom[! macsCom  %in% csaw$external_gene_name])

## 2519
macsOnly <- unique(
  macsUnion[! macsUnion  %in% csaw$external_gene_name])
```



For Ikzf3:
```{r}
## common between MACS and CSAW: 462
comCalls3 <- intersect(csaw3$symbol, macsIkzf3$SYMBOL)

# 1582
unionCalls3 <- unique(c(csaw3$symbol, macsIkzf3$SYMBOL)) 

## Method specific calls: 306
csawOnly3 <- unique(
  csaw3$symbol[! csaw3$symbol %in% macsIkzf3$SYMBOL])

## 814
macsOnly3 <- unique(
  macsIkzf3$SYMBOL[! macsIkzf3$SYMBOL  %in% csaw3$symbol])
```

## Common peak ranges
```{r}
library(GenomicRanges)
##---- for Ikzf1
macs1_granges <- makeGRangesFromDataFrame(macs1, keep.extra.columns = T)

macs2_granges <- makeGRangesFromDataFrame(macs2, keep.extra.columns = T)

macs4_granges <- makeGRangesFromDataFrame(macs4, keep.extra.columns = T)

##---- for Ikzf3
macs3_granges <- makeGRangesFromDataFrame(macsIkzf3, keep.extra.columns = T)

##---- overlapping ranges of peaks
macs1_macs3 <- subsetByOverlaps(macs3_granges, macs1_granges)

macs2_macs3 <- subsetByOverlaps(macs3_granges, macs2_granges)

macs4_macs3 <- subsetByOverlaps(macs3_granges, macs4_granges)


##----- for csaw:
csaw1_granges <- makeGRangesFromDataFrame(csaw, keep.extra.columns = T)

csaw3_granges <- readRDS(paste0(outPath, "Ikzf3_CnR_CSAW_fraglen195_combFDR_ConsWin50-200_normFactors_GRanges600.RDS")) 

csaw1_csaw3 <- subsetByOverlaps(csaw1_granges, csaw3_granges)
```



## Common target genes between Ikzf1 and Ikzf3 KO
67 common targets between Ikzf3 and the targets common between the three replicates of Ikzf1 analysed by MACS2

326 common targets between Ikzf3 and the targets from any of the three replicates of Ikzf1

29 common targets between Ikzf3 and the targets from Ikzf1 analysed by csaw
```{r}
## 292 common between the below two genes (MACS)
length(unique(intersect(macsIkzf3$SYMBOL, macsUnion)))    ## 326
# length(intersect(macsIkzf3_nf$SYMBOL, macsUnion)) ## 248
# length(intersect(edg3$SYMBOL, macsUnion))         ## 318

# Common between both csaw: 
intersect(csaw3$symbol,
                 csaw$external_gene_name)
#  [1] "Clasp1"   "Lta4h"    "Sfi1"     "Polr2a"   "Elp5"    
#  [6] "Rdm1"     "Nemf"     "Btn1a1"   "Malt1"    "Hnrnpul2"
# [11] "Ttc9c"    "Zfp91"    "Nfkb2"    "Zcchc7"   "Vsig10l" 
# [16] "Irf2"

# length(intersect(macsIkzf3$SYMBOL, macsCom))      ## 67
# length(intersect(macsIkzf3_nf$SYMBOL, macsCom))   ## 68
# length(intersect(edg3$SYMBOL, macsCom))           ## 62

# length(intersect(macsIkzf3$SYMBOL, macsUnion))      ## 326
# length(intersect(macsIkzf3_nf$SYMBOL, macsUnion))   ## 348
# length(intersect(edg3$SYMBOL, macsUnion))           ## 318

##--- common between both genes according to MACS results
# comBothMACS <- intersect(macsIkzf3$SYMBOL, macsCom)
comBothMACS <- intersect(macsIkzf3$SYMBOL, macsUnion) ## 326
comBothCSAW <- intersect(csaw3$symbol, csaw$external_gene_name)

##--- common between both genes according to either methods (MACS or CSAW)
comBoth_eitherMethods <- intersect(unionCalls, unionCalls3)  # 402


##--- common between both MACS (com for Ikzf1) and both CSAW
# intersect(
#   intersect(macsIkzf3$SYMBOL, macsCom),
#   intersect(csaw3$symbol, csaw$external_gene_name)
# )
# [1] "Clasp1" "Lta4h"  "Polr2a" "Elp5"   "Rdm1"   "Ttc9c" 
# [7] "Irf2"

##--- common between both MACS (union for Ikzf1) and both CSAW:
## additional: Malt1 and Vsig10l
intersect(intersect(macsIkzf3$SYMBOL, macsUnion),
         intersect(csaw3$symbol, csaw$external_gene_name))
# [1] "Clasp1"  "Lta4h"   "Polr2a"  "Elp5"    "Rdm1"   
# [6] "Malt1"   "Ttc9c"   "Vsig10l" "Irf2" 


length(intersect(macsIkzf3$SYMBOL, 
                 csaw$external_gene_name)) ## 29

# length(intersect(macsIkzf3_nf$SYMBOL, 
#                  csaw$external_gene_name)) ## 35

# length(intersect(edg3$SYMBOL, 
#                  csaw$external_gene_name)) ## 37

length(intersect(csaw3$symbol, 
                 csaw$external_gene_name)) ## 16


```

```{r}
# comBothMACS <- intersect(macsIkzf3$SYMBOL, macsCom)
comBothMACS <- intersect(macsIkzf3$SYMBOL, macsUnion) ##326
comBothCSAW <- intersect(csaw3$symbol, csaw$external_gene_name) ## 16

comBoth_eitherMethods <- intersect(unionCalls, unionCalls3)  # 402

## Take the union of the csaw and macs2 for Ikzf3 or Ikzf1
# targets_Ikzf3 <- unique(c(csaw3$symbol, macsIkzf3$SYMBOL ))  ## 1582
# targets_Ikzf1 <- unique(c(csaw$external_gene_name, macsUnion ))  ## 2775

targets_Ikzf3 <- unionCalls3 ## 1582
targets_Ikzf1 <- unionCalls  ## 2775

  
# comAll <- intersect(
#   intersect(macsIkzf3$SYMBOL, macsCom),
#   intersect(csaw3$symbol, csaw$external_gene_name)
# )

## 9 genes
# [1] "Clasp1"  "Lta4h"   "Polr2a"  "Elp5"    "Rdm1"    "Malt1"   "Ttc9c"   "Vsig10l"
# [9] "Irf2"  
comAll <- intersect(
  intersect(macsIkzf3$SYMBOL, macsUnion),
  intersect(csaw3$symbol, csaw$external_gene_name)
)


## 462
com3 <- intersect(macsIkzf3$SYMBOL, 
                 csaw3$symbol)

## 331
Ikzf3Only <- com3[
  ! com3 %in% macsUnion &
  ! com3 %in% csaw$external_gene_name
]


## 133
# com1 <- intersect(macsCom, 
#                  csaw$external_gene_name)

## 195
com1 <- intersect(macsUnion, 
                 csaw$external_gene_name)
## 111
## 167 (for union)
Ikzf1Only <- com1[
  ! com1 %in% macsIkzf3$SYMBOL &
  ! com1 %in% csaw3$symbol
]


##--- Ikzf1 or Ikzf3 specific targets based on union of MACS and CSAW in both methods
Ikzf3Only_union <- unionCalls3[ ! unionCalls3 %in% unionCalls]
Ikzf1Only_union <- unionCalls[ ! unionCalls %in% unionCalls3]
```

### Save targets
```{r}

saveRDS(list(targets_Ikzf1 = targets_Ikzf1, 
             targets_Ikzf3 = targets_Ikzf3, 
             Ikzf1Only_union = Ikzf1Only_union,
             Ikzf3Only_union = Ikzf3Only_union ), 
        paste0(outPath, "Ikzf1.3_Targets.RDS"))
# length(intersect(macsIkzf3$SYMBOL, macsCom))
# intersect(macsIkzf3$SYMBOL, macsCom)
```

# Prepare data
Here, I attempt to generate a comprehensive list of DEGs/Targets in one table, where I add several columns to group genes; these include whether or not they are DEGs in any comparisons, they are target gene of Ikzf1 or Ikzf3 or both, whether or not they are TFs, are they AP1 genes, are they DEGs in ex-vivo Ikzf1/3 data, etc. 

## Add DEG and Target info
Add columns depending on whether or not a given gene is DEGs in the three comparisons:
```{r}
DEstats$DEGs_group[!DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id &
                     !DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id &
                     !DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "Non-DEGs"

DEstats$DEGs_group[DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id &
                     !DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id &
                     !DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "DEGs only in DKO"

DEstats$DEGs_group[!DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id  &
                     DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id  &
                     !DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "DEGs only in Ikzf1 KO"

DEstats$DEGs_group[!DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id  &
                     !DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id  &
                     DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "DEGs only in Ikzf3 KO"


DEstats$DEGs_group[!DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id &
                     DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id &
                     DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "DEGs in Ikzf1 & Ikzf3 KOs"

DEstats$DEGs_group[DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id  &
                     !DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id  &
                     DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "DEGs in DKO & Ikzf3 KO"

DEstats$DEGs_group[DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id &
                     DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id  &
                     !DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "DEGs in DKO & Ikzf1 KO"

DEstats$DEGs_group[DEstats$ensembl_gene_id %in% DEGs_DKO$ensembl_gene_id  &
                     DEstats$ensembl_gene_id %in% DEGs_KO1$ensembl_gene_id  &
                     DEstats$ensembl_gene_id %in% DEGs_KO3$ensembl_gene_id] <-
  "DEGs in all KOs"
```

Also add a column for the target genes from ChIP-seq data:
```{r}

DEstats$Target_Ikzf1 <- FALSE
DEstats$Target_Ikzf1[DEstats$Symbol %in% targets_Ikzf1] <- TRUE

DEstats$Target_Ikzf3 <- FALSE
DEstats$Target_Ikzf3[DEstats$Symbol %in% targets_Ikzf3] <- TRUE

# DEstats$Target_Ikzf2 <- FALSE
# DEstats$Target_Ikzf2[DEstats$Symbol %in% unionIkzf2] <- TRUE


DEstats$Target <- "Non-Target gene"

DEstats$Target[
  DEstats$Target_Ikzf1 &
  # !DEstats$Target_Ikzf2 &
  !DEstats$Target_Ikzf3 
                 ] <- "Ikzf1 target"

# DEstats$Target[
#   !DEstats$Target_Ikzf1 &
#   DEstats$Target_Ikzf2 &
#   !DEstats$Target_Ikzf3 
#                  ] <- "Ikzf2 target"

DEstats$Target[
  !DEstats$Target_Ikzf1 &
  # !DEstats$Target_Ikzf2 &
  DEstats$Target_Ikzf3 
                 ] <- "Ikzf3 target"
# 
# DEstats$Target[
#   !DEstats$Target_Ikzf1 &
#   DEstats$Target_Ikzf2 &
#   DEstats$Target_Ikzf3 
#                  ] <- "Ikzf2/3 target"

DEstats$Target[
  DEstats$Target_Ikzf1 &
  # !DEstats$Target_Ikzf2 &
  DEstats$Target_Ikzf3 
                 ] <- "Ikzf1/3 target"
# 
# DEstats$Target[
#   DEstats$Target_Ikzf1 &
#   DEstats$Target_Ikzf2 &
#   !DEstats$Target_Ikzf3 
#                  ] <- "Ikzf1/2 target"
# 
# DEstats$Target[
#   DEstats$Target_Ikzf1 &
#   DEstats$Target_Ikzf2 &
#   DEstats$Target_Ikzf3 
#                  ] <- "Ikzf1/2/3 target"

```

```{r}
DEstats$Target_Ikzf1_details <- "Non-Target gene"
DEstats$Target_Ikzf1_details[DEstats$Symbol %in% comCalls] <-
  "Target gene: both methods"
DEstats$Target_Ikzf1_details[DEstats$Symbol %in% csawOnly] <-
  "Target gene: only CSAW"
DEstats$Target_Ikzf1_details[DEstats$Symbol %in% macsOnly] <-
  "Target gene: only MACS2"


DEstats$Target_Ikzf3_details <- "Non-Target gene"
DEstats$Target_Ikzf3_details[DEstats$Symbol %in% comCalls3] <-
  "Target gene: both methods"
DEstats$Target_Ikzf3_details[DEstats$Symbol %in% csawOnly3] <-
  "Target gene: only CSAW"
DEstats$Target_Ikzf3_details[DEstats$Symbol %in% macsOnly3] <-
  "Target gene: only MACS2"

DEstats$Target_details <- "Non-Target gene"

DEstats$Target_details[DEstats$Symbol %in% Ikzf3Only_union] <-
  "Target gene: only Ikzf3, either methods"
DEstats$Target_details[DEstats$Symbol %in% Ikzf3Only] <-
  "Target gene: only Ikzf3, both methods"

DEstats$Target_details[DEstats$Symbol %in% Ikzf1Only_union] <-
  "Target gene: only Ikzf1, either methods"
DEstats$Target_details[DEstats$Symbol %in% Ikzf1Only] <-
  "Target gene: only Ikzf1, both methods"


DEstats$Target_details[DEstats$Symbol %in% comBoth_eitherMethods] <-
  "Target gene: both genes, either methods"
DEstats$Target_details[DEstats$Symbol %in% comBothMACS] <-
  "Target gene: both genes, MACS2"
DEstats$Target_details[DEstats$Symbol %in% comBothCSAW] <-
  "Target gene: both genes, CSAW"
DEstats$Target_details[DEstats$Symbol %in% comAll] <-
  "Target gene: both genes, both methods"

DEstats_export <- DEstats[, c(
    "Symbol",
    "DEGs_group",
    "Target",
    "logFC_DKO",
    "logFC_KO1",
    "logFC_KO3",
    "adj.P.Val_DKO",
    "adj.P.Val_KO1",
    "adj.P.Val_KO3",
    "Target_details",
    "Target_Ikzf1_details",  
    "Target_Ikzf3_details",
    "Biotype"
)]

DEstats_export$Symbol_fixed <- paste0("_", DEstats_export$Symbol)
# write.csv(DEstats_export, paste0(outPath, "DEStat_DEGs_Targets_Ikzfs.csv"), row.names = F)
write.csv(
  DEstats_export,
  paste0(outPath, "DEStat_DEGs_Targets_Ikzf1_Ikzf3_selColumns.csv"),
  row.names = F
)
write.csv(
  DEstats,
  paste0(outPath, "DEStat_DEGs_Targets_Ikzf1_Ikzf3_allColumns.csv"),
  row.names = F
)

dim(DEstats_export[! DEstats_export$DEGs_group == "Non-DEGs" & ! DEstats_export$Target =="Non-Target gene", ]) ## 1689


dim(DEstats_export[! DEstats_export$DEGs_group == "Non-DEGs" | ! DEstats_export$Target =="Non-Target gene", ]) ## 7971 --> 5439

# DEstats_export <-
#   read.csv(
#     paste0(outPath, "DEStat_DEGs_Targets_Ikzf1_Ikzf3_selColumns.csv"),
#     stringsAsFactors = F
#   )

```


## Add protein classes (TF) to DEStats
Use data from human protein atlas to add protein class annotaions to each gene in different DE stat data sets from in vitro and ex-vivo (Ikzf1) data.
```{r}
hpaPath <- "/Users/mfor0011/Documents/data/HPA/"
connectomePath <- "/Users/mfor0011/Documents/data/connectome/"

prs <- read.delim(paste0(hpaPath, "proteinatlas.tsv"))
dim(prs)
```


Subset to selected columns and rows:
```{r}
getColumns <-
  c(
    "Gene",
    "Ensembl",
    "Uniprot",
    "Gene.description",
    "Protein.class",
    grep("evidence", colnames(prs), value = T),
    "Antibody"
  )

getRows <- grepl("secreted proteins", prs$Protein.class, ignore.case = T) |
  grepl("Transcription factors", prs$Protein.class, ignore.case = T) | 
  grepl("membrane proteins", prs$Protein.class, ignore.case = T) 

## 8500 rows have one of the three main classes
summary(getRows) 

## prs selected columns and rows:
prsSelected <- prs[getRows, getColumns]
dim(prsSelected)

prsSelected <- prsSelected[! duplicated(prsSelected$Gene), ]
```

We need to map human gene names to mouse names:

```{r}
library(biomaRt)

martHuman <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
martMouse <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")

human_ids <- as.character(prsSelected$Gene)

# human / mouse
## Retrieves information from two linked datasets
prsMappingIDs <- getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = human_ids,
  mart = martHuman,
  attributesL = c("external_gene_name"),
  martL = martMouse
)

## There are 1031 gene IDs that are duplicated in mouse column (most of whicg which are also duplicated in human column - human column has more NAs: 1600) - delet 1031 duplicated mouse IDs
prsMouseIDs <- prsMappingIDs[! duplicated(prsMappingIDs$Gene.name), ]
## Then remove the remaining 909 duplicated human IDs:
prsMouseIDs <- prsMappingIDs[! duplicated(prsMappingIDs$HGNC.symbol), ]

dim(prsMouseIDs)
## 7654
```

Merge human-mouse mapping data with the HPA data:
```{r}
prsSelectedIDs <- merge(prsSelected, prsMouseIDs, by.x = "Gene", by.y = "HGNC.symbol", all.x = T)

## make a new columns for the protein class that has only what we are interested in:


prsSelectedIDs$ProteinClass[
  grepl("Transcription factors", prsSelectedIDs$Protein.class, ignore.case = T)] <- "TF"

prsSelectedIDs$ProteinClass[
  grepl("membrane proteins", prsSelectedIDs$Protein.class, ignore.case = T)] <- "Membrane"

prsSelectedIDs$ProteinClass[
  grepl("secreted proteins", prsSelectedIDs$Protein.class, ignore.case = T)] <- "Secreted"


prsSelectedIDs$ProteinClass[
  grepl("secreted proteins", prsSelectedIDs$Protein.class, ignore.case = T) & 
  grepl("membrane proteins", prsSelectedIDs$Protein.class, ignore.case = T)] <- "Secreted & Membrane"

prsSelectedIDs$ProteinClass[
  grepl("Transcription factors", prsSelectedIDs$Protein.class, ignore.case = T) & 
  grepl("membrane proteins", prsSelectedIDs$Protein.class, ignore.case = T)] <- "Membrane and TF"


table(prsSelectedIDs$ProteinClass)

#            Membrane     Membrane and TF            Secreted 
#                5294                  35                1519 
# Secreted & Membrane                  TF 
#                 189                1461 

## remove those with NA in mouse symbols
prsSelectedIDs <- prsSelectedIDs[ complete.cases(prsSelectedIDs$Gene.name), ] ## 7387
prsSelectedIDs <- prsSelectedIDs[ !duplicated(prsSelectedIDs$Gene.name), ] ## 7069

table(prsSelectedIDs$ProteinClass)

#            Membrane     Membrane and TF            Secreted 
#                4530                  28                1226 
# Secreted & Membrane                  TF 
#                 168                1117 
```

Next merge these with the DEStat data:
```{r}
DEStatDataPrs <-
  merge(
    DEstats_export,
    prsSelectedIDs ,
    by.x = "Symbol",
    by.y = "Gene.name",
    all.x = T
  )

# 11736

# write.csv(DEStatDataPrs, paste0(outPath, "DEStat_DEGs_Targets_Ikzf1_Ikzf3_selColumns_ProteinClass.csv"), row.names = F)
# 
DEStatDataPrs <-
  read.csv(
    paste0(
      outPath,
      "DEStat_DEGs_Targets_Ikzf1_Ikzf3_selColumns_ProteinClass.csv"
    ),
    stringsAsFactors = F
  )


DEStatDataPrs_TF_DEGs_or_Targe <-
  DEStatDataPrs[grepl("TF", DEStatDataPrs$ProteinClass) &
                  (DEStatDataPrs$Target != "Non-Target gene" |
                     DEStatDataPrs$DEGs_group != "Non-DEGs"),]  ## 444

DEStatDataPrs_TF_DEGs_and_Targe <-
  DEStatDataPrs[grepl("TF", DEStatDataPrs$ProteinClass) &
                  DEStatDataPrs$Target != "Non-Target gene" &
                  DEStatDataPrs$DEGs_group != "Non-DEGs",]  ## 144
```

Add protein classes to KO1_Exvivo DE data too. 
```{r}
DE_KO1_Exvivo_pr <- merge(DE_KO1_Exvivo,
        DEStatDataPrs[, c("Target", "Symbol", "ProteinClass")],
        by.x = "external_gene_name" ,
        by.y = "Symbol", all.x =T)

## get TFs
DE_KO1_Exvivo_prTF <- DE_KO1_Exvivo_pr[DE_KO1_Exvivo_pr$ProteinClass == "TF", ]
DE_KO1_Exvivo_prTF <- DE_KO1_Exvivo_prTF[complete.cases(DE_KO1_Exvivo_prTF$ProteinClass), ]

## get TFs that are DEGs
DEGs_KO1_Exvivo_prTF <- DE_KO1_Exvivo_prTF[DE_KO1_Exvivo_prTF$adj.P.Val < 0.05, ]

DEGs_KO1_Exvivo_prTF <- DEGs_KO1_Exvivo_prTF[order(DEGs_KO1_Exvivo_prTF$adj.P.Val),]
```

## Add AP1 targets
Read in the list of AP-1 targets, map human-mouse IDs and add them to the table
```{r}
ap <- read.csv("../data/AP1_targets.csv", stringsAsFactors = F, header = F)

## map these to mouse IDs:
library(biomaRt)

martHuman <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host="https://dec2021.archive.ensembl.org")
martMouse <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl", host="https://dec2021.archive.ensembl.org")

human_ids <- as.character(ap$V1)

# human / mouse
## Retrieves information from two linked datasets
ap_ids <- getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = human_ids,
  mart = martHuman,
  # attributesL = c("mgi_symbol"),
  attributesL = c("external_gene_name"),
  martL = martMouse
)

ap_ids <- prsMappingIDs

## There are 1031 gene IDs that are duplicated in mouse column (most of whicg which are also duplicated in human column - human column has more NAs: 1600) - delet 1031 duplicated mouse IDs
prsMouseIDs <- prsMappingIDs[! duplicated(prsMappingIDs$Gene.name), ]
## Then remove the remaining 909 duplicated human IDs:
prsMouseIDs <- prsMappingIDs[! duplicated(prsMappingIDs$HGNC.symbol), ]


convertHumanGeneList <- function(x) {
  require("biomaRt")
  human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
  mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
  genesV2 = getLDS(
    attributes = c("hgnc_symbol"),
    filters = "hgnc_symbol",
    values = x ,
    mart = human,
    attributesL = c("mgi_symbol"),
    martL = mouse,
    uniqueRows = T
  )
  humanx <- unique(genesV2[, 2])
  # Print the first 6 genes found to the screen
  print(head(humanx))
  return(humanx)
}

# gsID <- convertHumanGeneList(ap$V1)

## I was getting errors so I searched and found this solution:: https://support.bioconductor.org/p/129636/
# ERROR: biomaRt has encountered an unexpected server error.
# Consider trying one of the Ensembl mirrors (for more details look at ?useEnsembl)

library(dplyr)
mouse_human_genes = read.csv("http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt",sep="\t")

convert_mouse_to_human <- function(gene_list){

  output = c()

  for(gene in gene_list){
    class_key = (mouse_human_genes %>% filter(Symbol == gene & Common.Organism.Name=="mouse, laboratory"))[['DB.Class.Key']]
    if(!identical(class_key, integer(0)) ){
      human_genes = (mouse_human_genes %>% filter(DB.Class.Key == class_key & Common.Organism.Name=="human"))[,"Symbol"]
      for(human_gene in human_genes){
        output = append(output,human_gene)
      }
    }
  }

  return (output)
}

gsID <- convert_mouse_to_human(ap$V1)

```

## Combine IKZF1/IKZF3 KO Ex-vivo data
We also have Ex-vivo data from Ikzf3, generated in the lab, so we read in the data too, to add annotations for genes in regards to whether or not they are DEGs in this data as well. 
```{r}
ai <- read.csv("../data/DE_AiolosKO_vs_WT_annotated.csv", stringsAsFactors = F)
colnames(ai)[3] <- "logFC_KO3_Exvivo"
colnames(ai)[7] <- "adj.P.Val_KO3_Exvivo"

DEStatDataPrs_ev3 <- merge(DEStatDataPrs, ai[, c(2, 3, 7)], by = "Symbol", all.x = T)
DEStatDataPrs_ev3$DEG_IKZF3_Exvivo <- FALSE
DEStatDataPrs_ev3$DEG_IKZF3_Exvivo[complete.cases(DEStatDataPrs_ev3$logFC_KO3_Exvivo)] <-
  TRUE
```

```{r}
ko1_ex <- DE_KO1_Exvivo
colnames(ko1_ex)[3] <- "Symbol"
colnames(ko1_ex)[5] <- "logFC_KO1_Exvivo"
colnames(ko1_ex)[9] <- "adj.P.Val_KO1_Exvivo"

DEStatDataPrs_ev3_ev1 <-
  merge(DEStatDataPrs_ev3, ko1_ex[, c(3, 5, 9)], by = "Symbol", all.x = T)

DEStatDataPrs_ev3_ev1$DEG_IKZF1_Exvivo <- FALSE
DEStatDataPrs_ev3_ev1$DEG_IKZF1_Exvivo[DEStatDataPrs_ev3_ev1$adj.P.Val_KO1_Exvivo < 0.05] <-
  TRUE

DEStatDataPrs_ev3_ev1$AP1_Target <- FALSE
DEStatDataPrs_ev3_ev1$AP1_Target[DEStatDataPrs_ev3_ev1$Symbol%in% ap_ids$Gene.name] <- TRUE

DEStatDataPrs_ev3_ev1_export <- DEStatDataPrs_ev3_ev1[,
                                                      c(
                                                        "Symbol",
                                                        "DEGs_group",
                                                        "DEG_IKZF3_Exvivo",
                                                        "DEG_IKZF1_Exvivo",
                                                        "Target",
                                                        "AP1_Target",
                                                        grep("logFC", colnames(DEStatDataPrs_ev3_ev1), value = T),
                                                        grep("adj.P.Val", colnames(DEStatDataPrs_ev3_ev1), value = T),
                                                        "Target_details",
                                                        "Target_Ikzf1_details",
                                                        "Target_Ikzf3_details",
                                                        "Biotype",
                                                        "Symbol_fixed",
                                                        "ProteinClass"
                                                      )]


write.csv(
  DEStatDataPrs_ev3_ev1_export,
  paste0(
    outPath,
    # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
    "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_AP1Target_selColumns_ProteinClass.csv"
  ),
  row.names = F
)
# 
# DEStatDataPrs_ev3_ev1_export <- read.csv(paste0(
#     outPath,
#     # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
#     "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_AP1Target_selColumns_ProteinClass.csv"
#   ), stringsAsFactors = F)
```

Get AP targets that are DEGs, and export a table of AP1 targets.
```{r}
DEGs_AP1_InVitro <- DEStatDataPrs_ev3_ev1_export[DEStatDataPrs_ev3_ev1_export$DEGs_group != "Non-DEGs" &
                               DEStatDataPrs_ev3_ev1_export$AP1_Target, ]

# 499  22

write.csv(
  DEGs_AP1_InVitro,
  paste0(
    outPath,
    # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
    "DEGsInvitro_AP1Target.csv"
  ),
  row.names = F
)


DEGs_both <- DEGs_AP1_InVitro[DEGs_AP1_InVitro$DEG_IKZF3_Exvivo & DEGs_AP1_InVitro$DEG_IKZF1_Exvivo, ]

write.csv(
  DEGs_both,
  paste0(
    outPath,
    # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
    "DEGsInvitro.any_DEGsExvivo.both.1.3_AP1Target.csv"
  ),
  row.names = F
)

ko1_ex$AP1_Target <- FALSE
ko1_ex$AP1_Target[ko1_ex$Symbol%in% ap_ids$Gene.name] <- TRUE
  
DEGs_AP1_ExVivo1 <- ko1_ex[ko1_ex$adj.P.Val_KO1_Exvivo < 0.05 & ko1_ex$AP1_Target, ]
  # 337  11
  
write.csv(
  DEGs_AP1_ExVivo1,
  paste0(
    outPath,
    # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
    "DEGsExvivo.Ikzf1_AP1Target.csv"
  ),
  row.names = F
)

ai$AP1_Target <- FALSE
ai$AP1_Target[ai$Symbol%in% ap_ids$Gene.name] <- TRUE
  
DEGs_AP1_ExVivo3 <- ai[ ai$AP1_Target, ]
# 24  11
  
write.csv(
  DEGs_AP1_ExVivo3,
  paste0(
    outPath,
    # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
    "DEGsExvivo.Ikzf3_AP1Target.csv"
  ),
  row.names = F
)

## Add DEG status to API targets
ap_ids$DEGs_InVitro <- FALSE
ap_ids$DEGs_InVitro[ap_ids$Gene.name %in% DEGs_AP1_InVitro$Symbol] <- TRUE

## Add DEG status to API targets
ap_ids$DEGs_ExVivo_IKZF1 <- FALSE
ap_ids$DEGs_ExVivo_IKZF1[ap_ids$Gene.name %in% DEGs_AP1_ExVivo1$Symbol] <- TRUE

## Add DEG status to API targets
ap_ids$DEGs_ExVivo_IKZF3 <- FALSE
ap_ids$DEGs_ExVivo_IKZF3[ap_ids$Gene.name %in% DEGs_AP1_ExVivo3$Symbol] <- TRUE

## Add DEG status to API targets
ap_ids$DEGs_Both_InVitro_ExVivo <- FALSE
ap_ids$DEGs_Both_InVitro_ExVivo[ap_ids$Gene.name %in% DEGs_both$Symbol] <- TRUE

write.csv(
  ap_ids,
  paste0(
    outPath,
    # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
    "AP1Target_DEGsStatus_InVitro_ExVivo.csv"
  ),
  row.names = F
)
```

## Prepare Ex-vivo data
```{r}
DEStat_ev1_ev3 <-
  merge(ko1_ex[, c(3, 5, 9)], ai, by = "Symbol", all.x = T)

DEStat_ev1_ev3_target <-
  merge(DEStat_ev1_ev3, DEStatDataPrs_ev3_ev1_export[,-c(6, 10, 11, 15, 16)], all.x = T)


write.csv(DEStat_ev1_ev3_target,
          paste0(outPath, "DEStta_ExVivo_Ikzf1_Ikzf3.csv"),
          row.names = F)


DEGs_ev1_ev3 <-
  DEStat_ev1_ev3_target[complete.cases(DEStat_ev1_ev3_target$logFC_KO3_Exvivo) &
                          DEStat_ev1_ev3_target$adj.P.Val_KO1_Exvivo < 0.05,]
## 175 genes

# Add missing target info:
x <- DEGs_ev1_ev3[is.na(DEGs_ev1_ev3$Target), ]

intersect(x$Symbol, targets_Ikzf1)
# [1] "Gprc5a" "Xdh"   
intersect(x$Symbol, targets_Ikzf3)
# [1] "Cd47" "Xdh" 

DEGs_ev1_ev3$Target[is.na(DEGs_ev1_ev3$Target)] <- "Non-Target gene"
DEGs_ev1_ev3$Target[DEGs_ev1_ev3$Symbol == "Xdh"] <- "Ikzf1/3 target"
DEGs_ev1_ev3$Target[DEGs_ev1_ev3$Symbol == "Gprc5a"] <- "Ikzf1 target"
DEGs_ev1_ev3$Target[DEGs_ev1_ev3$Symbol == "Cd47"] <- "Ikzf3 target"
```

## Add protein classes (TF) to Maturity data
Also add these protein info to the immature and mature NK cell data:
```{r}
## 51 genes
m1_im <- read.csv(paste0(dataPath, "ImmM1M2/DE_NK_mature_vs_immature.csv"), 
                  stringsAsFactors = F)
## 1041
m2_im <- read.csv(paste0(dataPath, "ImmM1M2/DE_NK_terminal_vs_immature.csv"), 
                  stringsAsFactors = F)
## 37
m2_m1 <- read.csv(paste0(dataPath, "ImmM1M2/DE_NK_terminal_vs_mature.csv"), 
                  stringsAsFactors = F)

## also read the RPKM data for immM1M2 - 12482 genes
mm <-
  read.csv(
    paste0(
      dataPath,
      "ImmM1M2/IMM_M1_M2 normalized_intensities(log2 RPKM).csv"
    ),
    stringsAsFactors = F
  )
## remove 62 NA symbols:
mm <- mm[complete.cases(mm$Symbol), ]   ## 12420 genes


## 12420 genes
mm <-
  merge(mm,
        prsSelectedIDs,
        by.x = "Symbol",
        by.y = "Gene.name",
        all.x = T)
## 632 TFs + 21 Membrane and TF 


# which of these are DEGs?
mm$DEG_M1_Imm <- FALSE
mm$DEG_M1_Imm[mm$Symbol %in% m1_im$Symbol] <- TRUE

mm$DEG_M2_Imm <- FALSE
mm$DEG_M2_Imm[mm$Symbol %in% m2_im$Symbol] <- TRUE

mm$DEG_M2_M1 <- FALSE
mm$DEG_M2_M1[mm$Symbol %in% m2_m1$Symbol] <- TRUE

mm_TFs_DEGs <- mm[grepl("TF", mm$ProteinClass) &
     (mm$DEG_M2_M1  | mm$DEG_M2_Imm | mm$DEG_M1_Imm), ]  ## 47


# write.csv(mm,
#           paste0(outPath, "MatureImmature_NK_logRPKM_ProteinClass.csv"),
#           row.names = F)

# mm <-
#   read.csv(paste0(outPath, "MatureImmature_NK_logRPKM_ProteinClass.csv"),
#            stringsAsFactors = F)
```

Overlap of mm_TFs_DEGs with DEStatDataPrs_TF_DEGs_and_Targe:
```{r}
intersect(mm_TFs_DEGs$Symbol, DEStatDataPrs_TF_DEGs_and_Targe$Symbol)
 # [1] "Aebp2"   "Batf3"   "Bhlhe40" "Dnmt3a"  "E2f1"    "Etv3"    "Foxd2"  
 # [8] "Klf6"    "Maz"     "Nfkb2"   "Prdm1"   "Tox"     "Trp53"   "Zfp362" 
```

Overlap of mm_TFs_DEGs with DEStatDataPrs_TF_DEGs_or_Targe:
```{r}
intersect(mm_TFs_DEGs$Symbol, DEStatDataPrs_TF_DEGs_or_Targe$Symbol)
#  [1] "Aebp2"   "Batf3"   "Bhlhe40" "Dmrta1"  "Dnmt3a"  "E2f1"    "Eomes"  
#  [8] "Etv3"    "Foxd2"   "Hbp1"    "Id2"     "Klf10"   "Klf2"    "Klf4"   
# [15] "Klf6"    "Maz"     "Meis3"   "Myc"     "Ncor2"   "Nfatc2"  "Nfkb1"  
# [22] "Nfkb2"   "Nfkbil1" "Prdm1"   "Rere"    "Tbx21"   "Tcf12"   "Tcf7"   
# [29] "Tfdp1"   "Thra"    "Tox"     "Trp53"   "Tsc22d3" "Zbtb18"  "Zbtb42" 
# [36] "Zeb2"    "Zfp362"  "Zfp512" 
```

Overlap with the DEGs genes
```{r}
intersect(mm_TFs_DEGs$Symbol, DEStatDataPrs_TF_DEGs_or_Targe$Symbol[
  DEStatDataPrs_TF_DEGs_or_Targe$DEGs_group != "Non-DEGs"
])
## 31
# [1] "Aebp2"   "Batf3"   "Bhlhe40" "Dmrta1"  "Dnmt3a"  "E2f1"    "Eomes"  
#  [8] "Etv3"    "Foxd2"   "Hbp1"    "Id2"     "Klf2"    "Klf4"    "Klf6"   
# [15] "Maz"     "Meis3"   "Nfatc2"  "Nfkb2"   "Nfkbil1" "Prdm1"   "Tbx21"  
# [22] "Tcf7"    "Tfdp1"   "Thra"    "Tox"     "Trp53"   "Tsc22d3" "Zbtb18" 
# [29] "Zbtb42"  "Zeb2"    "Zfp362" 
```

Overlap with the Target genes
```{r}
intersect(mm_TFs_DEGs$Symbol, DEStatDataPrs_TF_DEGs_or_Targe$Symbol[
  DEStatDataPrs_TF_DEGs_or_Targe$Target != "Non-Target gene"
])
# 21
# [1] "Aebp2"   "Batf3"   "Bhlhe40" "Dnmt3a"  "E2f1"    "Etv3"    "Foxd2"  
#  [8] "Klf10"   "Klf6"    "Maz"     "Myc"     "Ncor2"   "Nfkb1"   "Nfkb2"  
# [15] "Prdm1"   "Rere"    "Tcf12"   "Tox"     "Trp53"   "Zfp362"  "Zfp512"
```

# Venn diagram comparisons
## Venn diagram: Ikzf1/3 targets

```{r}

library(VennDiagram)
myCol <- brewer.pal(3, "Pastel2")[1:2]

ven <- venn.diagram(
  x = list(
   DEStatDataPrs$Symbol[DEStatDataPrs$Target %in% c("Ikzf1 target", "Ikzf1/3 target")],
   DEStatDataPrs$Symbol[DEStatDataPrs$Target %in% c("Ikzf3 target", "Ikzf1/3 target")]
  ),
  category.names = c("Ikzf1\ntargets" , "Ikzf3\ntargets"),
  filename = NULL,
  # filename = paste0(figPath, 'VennDiagram_ComTarget_Ikzf1_OhWil.png'),
  output = TRUE,
  # Circles
  # fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3)),
  # col = c("#440154ff", '#21908dff'),
  fill = myCol,
  alpha = c(0.3, 0.3),
  lwd = 1,
  # lty = 'blank',
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 550 ,
  width = 550 ,
  resolution = 400,
  compression = "lzw",
   # Numbers
  cex = 1,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27),
  cat.dist = c(0.055, 0.055),
  cat.fontfamily = "sans"
  # ext.pos = c(20, -20)
  # rotation = 1
)

pdf(paste0(figPath, "VennDiagram_ComTarget_Ikzf1_Ikzf3.pdf"), height = 3, width = 3)
grid.draw(ven)
dev.off()

```

## Venn diagram: DEGs
```{r}
myCol <- brewer.pal(3, "Pastel2")

venn.diagram(
  x = list(
DEStatDataPrs$Symbol[DEStatDataPrs$DEGs_group %in% c("DEGs in all KOs", grep("Ikzf1", unique(DEStatDataPrs$DEGs_group), value = T))], 
DEStatDataPrs$Symbol[DEStatDataPrs$DEGs_group %in% c("DEGs in all KOs", grep("Ikzf3", unique(DEStatDataPrs$DEGs_group), value = T))],
DEStatDataPrs$Symbol[DEStatDataPrs$DEGs_group %in% c("DEGs in all KOs", grep("DKO", unique(DEStatDataPrs$DEGs_group), value = T))]
  ),
  category.names = c("Ikzf1 KO\nDEGs" , "Ikzf3 KO\nDEGs" , "DKO\nDEGs"),
  filename = paste0(figPath, 'VennDiagram_DEGs_IkzfData_InVitro.png'),
  output = TRUE,
  # Circles
  fill = myCol,
  lwd = 1,
  # lty = 'blank',
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 600 ,
  width = 600 ,
  resolution = 300,
  compression = "lzw",
   # Numbers
  cex = .6,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.095, 0.095, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)
```

## Venn diagram: Ikzf1 targets & Ikzf1 DEGs
```{r}

library(VennDiagram)
myCol <- brewer.pal(3, "Pastel2")[1:2]

ven <- venn.diagram(
  x = list(
   DEStatDataPrs$Symbol[DEStatDataPrs$Target %in% c("Ikzf1 target", "Ikzf1/3 target")],
   DEStatDataPrs$Symbol[DEStatDataPrs$DEGs_group %in% c("DEGs in all KOs", grep("Ikzf1", DEStatDataPrs$DEGs_group, value = T))]
  ),
  category.names = c("Ikzf1\ntargets" , "Ikzf1 KO\nDEGs"),
  filename = NULL,
  output = TRUE,
  main = "In-vitro",
  fill = myCol,
  alpha = c(0.3, 0.3),
  lwd = 1,
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 550 ,
  width = 550 ,
  resolution = 400,
  compression = "lzw",
   # Numbers
  cex = 1,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-110, 120),
  cat.dist = c(0.055, 0.055),
  cat.fontfamily = "sans"
)

pdf(paste0(figPath, "VennDiagram_Ikzf1Target_Ikzf1DEGs_InVitro2.pdf"), height = 3, width = 3)
grid.draw(ven)
dev.off()

```


## Venn diagram: Ex-vivo Ikzf1 targets & Ikzf1 DEGs

```{r}

DE_KO1_Exvivo_pr$DEGs_group <- "Non-DEG"
DE_KO1_Exvivo_pr$DEGs_group[DE_KO1_Exvivo_pr$adj.P.Val < 0.05] <- "Ikzf1 DEGs (ex-vivo)"
library(VennDiagram)
myCol <- brewer.pal(3, "Pastel2")[1:2]

ven <- venn.diagram(
  x = list(
   DE_KO1_Exvivo_pr$external_gene_name[DE_KO1_Exvivo_pr$Target %in% c("Ikzf1 target", "Ikzf1/3 target")],
   DE_KO1_Exvivo_pr$external_gene_name[DE_KO1_Exvivo_pr$DEGs_group %in% "Ikzf1 DEGs (ex-vivo)"]
  ),
  category.names = c("Ikzf1\ntargets" , "Ikzf1 KO\nDEGs"),
  filename = NULL,
  output = TRUE, 
  main = "Ex-vivo",
  fill = myCol,
  alpha = c(0.3, 0.3),
  lwd = 1,
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 550 ,
  width = 550 ,
  resolution = 400,
  compression = "lzw",
   # Numbers
  cex = 1,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-110, 120),
  cat.dist = c(0.055, 0.055),
  cat.fontfamily = "sans"
)

pdf(paste0(figPath, "VennDiagram_Ikzf1Target_Ikzf1DEGs_ExVivo.pdf"), height = 3, width = 3)
grid.draw(ven)
dev.off()

```


## Venn diagram: Ex-vivo Ikzf1 DEGs & Ikzf1 targets 
```{r}
myCol <- brewer.pal(3, "Pastel2")
library(VennDiagram)

venn.diagram(
  x = list(
    targets_Ikzf1, 
    DEGs_KO1_Exvivo$external_gene_name,
    ap_ids$Gene.name
  ),
  category.names = c("Ikzf1\ntarget" , "Ikzf1 DEGs\nex-vivo" , "Ap-1\ntarget"),
  filename = paste0(figPath, 'VennDiagram_DEGsIkzf1Exvivo_Ikzf1Target_Ap1Target.png'),
  output = TRUE,
  # Circles
  fill = myCol,
  lwd = 1,
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 600 ,
  width = 600 ,
  resolution = 300,
  compression = "lzw",
   # Numbers
  cex = .6,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.095, 0.095, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)
```


## Venn diagram: In-vitro Ikzf1 DEGs & Ikzf1 targets 
```{r}
myCol <- brewer.pal(3, "Pastel2")
library(VennDiagram)

venn.diagram(
  x = list(
    targets_Ikzf1, 
    DEGs_KO1$external_gene_name ,
    ap_ids$Gene.name
  ),
  category.names = c("Ikzf1\ntarget" , "Ikzf1 DEGs\nin-vitro" , "Ap-1\ntarget"),
  filename = paste0(figPath, 'VennDiagram_DEGsIkzf1InVitro_Ikzf1Target_Ap1Target.png'),
  output = TRUE,
  # Circles
  fill = myCol,
  lwd = 1,
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 600 ,
  width = 600 ,
  resolution = 300,
  compression = "lzw",
   # Numbers
  cex = .6,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.095, 0.095, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)
```

# Export subsets of the combined table:

- Ikzf1/3 and Ap1 targets and annotate them for DEG status
- Ap1 only targets that are ex-vivo DEGs
- Ikzf1 only targets that are ex-vivo DEGs
- Ikzf1 & Ap1 targets that are ex-vivo DEGs

```{r}
ikzfTargets_DEGsStats <- data.frame(Targets = unique(c(targets_Ikzf1, targets_Ikzf3, ap_ids$Gene.name)))

ikzfTargets_DEGsStats$Ikzf1_target <- FALSE
ikzfTargets_DEGsStats$Ikzf1_target[ikzfTargets_DEGsStats$Targets %in% targets_Ikzf1] <- TRUE

ikzfTargets_DEGsStats$Ikzf3_target <- FALSE
ikzfTargets_DEGsStats$Ikzf3_target[ikzfTargets_DEGsStats$Targets %in% targets_Ikzf3] <- TRUE

ikzfTargets_DEGsStats$Ap1_target <- FALSE
ikzfTargets_DEGsStats$Ap1_target[ikzfTargets_DEGsStats$Targets %in% ap_ids$Gene.name] <- TRUE

ikzfTargets_DEGsStats$DEG_KO1_InVitro <- FALSE
ikzfTargets_DEGsStats$DEG_KO1_InVitro[ikzfTargets_DEGsStats$Targets %in% DEGs_KO1$external_gene_name] <- TRUE

ikzfTargets_DEGsStats$DEG_KO3_InVitro <- FALSE
ikzfTargets_DEGsStats$DEG_KO3_InVitro[ikzfTargets_DEGsStats$Targets %in% DEGs_KO3$external_gene_name] <- TRUE

ikzfTargets_DEGsStats$DEG_DKO_InVitro <- FALSE
ikzfTargets_DEGsStats$DEG_DKO_InVitro[ikzfTargets_DEGsStats$Targets %in% DEGs_DKO$external_gene_name] <- TRUE

ikzfTargets_DEGsStats$DEG_KO1_ExVivo <- FALSE
ikzfTargets_DEGsStats$DEG_KO1_ExVivo[ikzfTargets_DEGsStats$Targets %in% DEGs_KO1_Exvivo$external_gene_name] <- TRUE

ikzfTargets_DEGsStats$DEG_KO3_ExVivo <- FALSE
ikzfTargets_DEGsStats$DEG_KO3_ExVivo[ikzfTargets_DEGsStats$Targets %in% ai$Symbol] <- TRUE

write.csv(
  ikzfTargets_DEGsStats,
  paste0(
    outPath,
    "Targets.Ikzf1.Ikzf3.Ap1_DEGStatus_InVitro_ExVivo.csv"
  ),
  row.names = T
)

## Ap1 only targets that are ex-vivo DEGs
x <- ikzfTargets_DEGsStats %>% 
  filter(!Ikzf1_target) %>%
  filter(DEG_KO1_ExVivo) %>% 
  filter(Ap1_target) %>% 
  data.frame()

write.csv(
  x,
  paste0(
    outPath,
    "TargetsAp1_DEGsIkzf1ExVivo_NonTargetIkzf1.csv"
  ),
  row.names = T
)


## Ikzf1 only targets that are ex-vivo DEGs
y <- ikzfTargets_DEGsStats %>% 
  filter(Ikzf1_target) %>%
  filter(DEG_KO1_ExVivo) %>% 
  filter(!Ap1_target) %>% 
  data.frame()

  
write.csv(
  y,
  paste0(
    outPath,
    "TargetsIkzf1_DEGsIkzf1ExVivo_NonTargetAp1.csv"
  ),
  row.names = T
)

## Ikzf1 & Ap1 targets that are ex-vivo DEGs
z <- ikzfTargets_DEGsStats %>% 
  filter(Ikzf1_target) %>%
  filter(DEG_KO1_ExVivo) %>% 
  filter(Ap1_target) %>% 
  data.frame()

  
write.csv(
  z,
  paste0(
    outPath,
    "TargetsIkzf1Ap1_DEGsIkzf1ExVivo.csv"
  ),
  row.names = T
)

```


# Ikzf1 targets with opposite RNA direction
```{r}
DEStatDataPrs_1 <- DEStatDataPrs %>% 
  filter(Target == "Ikzf1 target") %>% 
  filter((logFC_KO1 > 0 & logFC_KO3 < 0) | (logFC_KO1 < 0 & logFC_KO3 > 0)) %>% 
  data.frame()
```

## Waterfall plot: DEGs in Ikzf1/DKO
### Target of both Ikzf1/3
```{r}
DEStatDataPrs_sub13 <- DEStatDataPrs %>% 
  filter(DEGs_group %in% c("DEGs in all KOs", "DEGs in DKO & Ikzf1 KO")) %>% 
  filter(Target %in% c("Ikzf1/3 target")) %>% 
  data.frame()
## 91 genes
```

```{r}
currentDEStat <-  DEStatDataPrs_sub13
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "DEGs_group")) %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Consistent",
                               TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    x = Symbol,
    y = logFC,
    fill = DEGs_group,
    color = Direction
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 3) +
  scale_fill_manual(values = degCols[c("DEGs in DKO & Ikzf1 KO", "DEGs in all KOs")]) +
  scale_color_manual(values = dirCols) +
  coord_flip() +
  ggtitle("In-vitro data") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

ggsave(
  filename = "WaterFall_DEGs.Ikzf1.DKO_Targets.Ikzf1.Ikzf3_IkzfData_InVitro.pdf",
  device = "pdf",
  path = figPath,
  width = 4,
  height = 15
)
```

Subset to those that have the same direction in Ikzf1 and DKO
```{r}
DEStatDataPrs_sub13dir <- DEStatDataPrs_sub13 %>% 
  mutate(Direction_DEG = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 ) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 ) ~ "Consistent",
                               TRUE ~ "Different")) %>% 
  filter(Direction_DEG == "Consistent") %>% 
  data.frame()




currentDEStat <-  DEStatDataPrs_sub13dir
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol


currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "DEGs_group")) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    x = Symbol,
    y = logFC,
    fill = DEGs_group
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 3) +
  scale_fill_manual(values = degCols[c("DEGs in DKO & Ikzf1 KO", "DEGs in all KOs")]) +
  coord_flip() +
  ggtitle("In-vitro data") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

ggsave(
  filename = "WaterFall_DEGs.Ikzf1.DKO.SameDir_Targets.Ikzf1.Ikzf3_IkzfData_InVitro.pdf",
  device = "pdf",
  path = figPath,
  width = 4,
  height = 14
)

```


### Target of Ikzf1
```{r}
DEStatDataPrs_sub1 <- DEStatDataPrs %>%
  filter(DEGs_group %in% c("DEGs in all KOs", "DEGs in DKO & Ikzf1 KO")) %>%
  filter(Target %in% c("Ikzf1 target", "Ikzf1/3 target")) %>%
  data.frame()
# ## 427 genes

## Subset to those that have the same direction in Ikzf1 and DKO
DEStatDataPrs_sub1dir <- DEStatDataPrs_sub1 %>% 
  mutate(Direction_DEG = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 ) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 ) ~ "Consistent",
                               TRUE ~ "Different")) %>% 
  filter(Direction_DEG == "Consistent") %>% 
  data.frame()
## 405 genes

currentDEStat <-  DEStatDataPrs_sub1dir
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol


currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "DEGs_group")) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    x = Symbol,
    y = logFC,
    fill = DEGs_group
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 3) +
  scale_fill_manual(values = degCols[c("DEGs in DKO & Ikzf1 KO", "DEGs in all KOs")]) +
  coord_flip() +
  ggtitle("In-vitro data") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

ggsave(
  filename = "WaterFall_DEGs.Ikzf1.DKO.SameDir_Targets.Ikzf1_IkzfData_InVitro.pdf",
  device = "pdf",
  path = figPath,
  width = 4,
  height = 45
)

```

#### In-vitro-data: selected genes from above
Selected genes from all the list above.
```{r}

## selected list from 
gg <- c("Zswim6", "Zfp827", "Ypel2", "Wnt10a", "Usp53", "Tyrobp", "Ttc32", "Trim62", "Trak1", "Tnik", "Tmem53", "Tgfb1", "Tbkbp1", "Susd4", "Stk40", "Srgap3", "Slc11a1", "Slamf7", "Sema6d", "Rab20", "Plxdc2", "Pik3ap1", "Pdcd1lg2", "Osgin1", "Nr4a3", "Ninj1", "Nedd9", "Myo6", "Mycn", "Mmp9", "Mcl1", "Mast1", "Marveld1", "Klra5", "Klf3", "Khdc1a", "Jund", "Itpka", "Il12rb1", "Ikzf2", "Ifng", "Gem", "Fosl2", "Fam174b", "Etv6", "Dusp5", "Dock8", "Cryl1", "Cish", "Cldn12", "Cdh17", "Cass4", "Bhlhe40", "Bcl2", "Asb2", "Actg1", "Abr")


## Subset to those that have the same direction in Ikzf1 and DKO
DEStatDataPrs_sub1dir <- DEStatDataPrs_sub1 %>%
  filter(Symbol %in% gg) %>% 
  mutate(Direction_DEG = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 )  ~ "Up-regulated",
                               TRUE ~ "Down-regulated")) %>%
  arrange(Direction_DEG) %>% 
  # filter(Direction_DEG == "Consistent") %>%
  data.frame()
## 405 genes

currentDEStat <-  DEStatDataPrs_sub1dir
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol

ordGenes <- currentDEStat$Symbol[order(currentDEStat$logFC_DKO)]

currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target", "Direction_DEG")) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  mutate(Symbol_ord = factor(Symbol, levels = ordGenes)) %>% 
  ggplot(., aes(
    x = Symbol_ord,
    y = logFC,
    fill = Direction_DEG,
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 3) +
  scale_fill_manual(values = brewer.pal(3, "Set1")[2:1]) +
  coord_flip() +
  ggtitle("In-vitro data") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

ggsave(
  filename = "WaterFall_DEGs.Ikzf1.DKO.SameDir_Targets.Ikzf1_InVitro_SelectedGenes_ordDir.pdf",
  device = "pdf",
  path = figPath,
  width = 4,
  height = 10
)


```

#### Ex-vivo: selected genes from above
```{r}
gg[! gg %in% DE_KO1_Exvivo$external_gene_name]
# "Wnt10a" "Susd4"  "Mycn"   "Itpka"  "Cass4" 
currentDEStat <- DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name %in% gg, ]

currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

row.names(currentDEStat) <- currentDEStat$external_gene_name

missingGenes <- gg[! gg %in% currentDEStat$external_gene_name]
missingData <-
  data.frame(matrix(
    ncol = ncol(currentDEStat),
    nrow = length(missingGenes)
  ))

colnames(missingData) <- colnames(currentDEStat)
row.names(missingData) <- missingGenes
missingData$external_gene_name <- missingGenes

currentDEStat <- rbind(currentDEStat[, -1], missingData[, -1])

currentDEStat <- currentDEStat[ordGenes, ]

currentDEStat$logFC[is.na(currentDEStat$logFC)] <- 0


currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T)
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  mutate(Symbol_ord = factor(external_gene_name, levels = ordGenes)) %>% 
  ggplot(., aes(
    # x = reorder(Symbol, logFC),
    x = Symbol_ord,
    y = logFC
  )) +
    geom_bar(stat = "identity",  fill = "gray30") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      coord_flip() +
      ggtitle("Ex-vivo data") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")


ggsave(
  filename = "WaterFall_DEGs.Ikzf1.DKO.SameDir_Targets.Ikzf1_Exvivo_SelectedGenes_ordDir.pdf",
  device = "pdf",
  path = figPath,
  width = 2,
  height = 10
)
```

#### Maturity data heatmap: selected genes from above
```{r}
# mm <-
#   read.csv(
#     paste0(
#       dataPath,
#       "ImmM1M2/IMM_M1_M2 normalized_intensities(log2 RPKM).csv"
#     ),
#     stringsAsFactors = F
#   )
# ## remove 62 NA symbols:
# mm <- mm[complete.cases(mm$Symbol), ]   ## 12420 genes
mm <- mm[!duplicated(mm$Symbol), ]   ## 12420 genes

library("grid")
gg[! gg %in% mm$Symbol]
# "Wnt10a" "Susd4"  "Mycn"   "Itpka" 

rnames <- mm$Symbol
mm <- mm[, 3:5]
row.names(mm) <- rnames

missingGenes <- gg[! gg %in% row.names(mm)]
missingData <- matrix(ncol = ncol(mm), nrow = length(missingGenes))

colnames(missingData) <- colnames(mm)
row.names(missingData) <- missingGenes

mm2 <- rbind(mm, missingData)
mmgg <- mm2[ordGenes, ]


# minVal <- min(mmgg, na.rm = T)
# maxVal <- max(mmgg, na.rm = T)

source(paste0(scriptPath, "plotHeatmapExpr.R"))
library(viridis)
library(ggpubr)
library(grid)

x <- plotHeatmapExpr(
  exprData = mmgg,
  heatmapCol = viridis::viridis(100),
  genes = rev(ordGenes),
  subsetSample = NULL,
  annotData = NULL,
  annotColumns = NULL,
  annotColList = NULL,
  scaleData = T,
  geneAnnot = NULL,
  geneCol = NULL,
  clustRows = F,
  clustCols = F,
  showRowNames = T,
  showColNames = T,
  textSize = 9,
  heatmapLegendTitle = "scaled\nlogRPKM",
  plotTitle = "Maturity data"
)

## the four missing genes are also missing from Ikzf data:
pdf(
  paste0(
    figPath,
    "Heatmap_DEGs.Ikzf1.DKO.SameDir_Targets.Ikzf1_MatureNK_SelectedGenes_ordDir.pdf"
  ),
  height = 10,
  width = 2.4
)
x
dev.off()

```


# Check genes
## Oh et al targets
Ikzf targets
```{r}
oh <- read.csv(paste0(dataPath, "Oh_JI/Oh_ChIPseq.csv"), 
               stringsAsFactors = F)

oh_ikzf1 <- oh[oh$Ikaros_ChIP_0h > 0, ] ## 883

##---- all the Oh targets that are DEGs in our data
DEStatDataPrs_ohTargets_DEGs_invitro <- DEStatDataPrs[
 ! grepl("Non-DEGs", DEStatDataPrs$DEGs_group) 
 & DEStatDataPrs$Symbol %in% unique(oh_ikzf1$Gene.Symbol), ]  ## 341 genes

table(DEStatDataPrs_ohTargets_DEGs_invitro$DEGs_group, DEStatDataPrs_ohTargets_DEGs_invitro$Target)
  #                           Ikzf1 target Ikzf1/3 target Ikzf3 target Non-Target gene
  # DEGs in all KOs                     14              5            6              30
  # DEGs in DKO & Ikzf1 KO              15              7            7              44
  # DEGs in DKO & Ikzf3 KO               7              6            4              21
  # DEGs in Ikzf1 & Ikzf3 KOs            1              1            0               8
  # DEGs only in DKO                    21              6            8              55
  # DEGs only in Ikzf1 KO               10              3            5              27
  # DEGs only in Ikzf3 KO                8              1            2              19

table( DEStatDataPrs_ohTargets_DEGs_invitro$Target)
   # Ikzf1 target  Ikzf1/3 target    Ikzf3 target Non-Target gene 
   #           76              29              32             204 

write.csv(DEStatDataPrs_ohTargets_DEGs_invitro, paste0(outPath, "Ikzf1_OhTargets_WilDEGs_InVitro.csv"), row.names = F)

##----- all the Oh targets that are Ikzf1 target in our data

DEStatDataPrs_ikzf1 <- DEStatDataPrs[
  grepl("Ikzf1", DEStatDataPrs$Target), ]

ikzf1_comTargets <- intersect(unique(oh_ikzf1$Gene.Symbol), DEStatDataPrs_ikzf1$Symbol) ## 159 genes common

oh_ikzf1_com <- oh_ikzf1[oh_ikzf1$Gene.Symbol %in% ikzf1_comTargets, ]

DEStatDataPrs_ikzf1_com <- DEStatDataPrs_ikzf1[
  DEStatDataPrs_ikzf1$Symbol %in% ikzf1_comTargets, 
]

## Ikzf1 ex-vivo:
DEGs_KO1_Exvivo_ikzf1_com <- DE_KO1_Exvivo[
  DE_KO1_Exvivo$external_gene_name %in% unique(oh_ikzf1$Gene.Symbol) &
  DE_KO1_Exvivo$adj.P.Val < 0.05 , ]

##----- 62 genes are common across the three data sets: ex-vivo, in-vitro, and Oh data
intersect(ikzf1_comTargets, DEGs_KO1_Exvivo_ikzf1_com$external_gene_name)
#  [1] "Hivep3"   "Kctd12"   "Trib1"    "Cd69"     "Etv6"     "Slamf7"   "Pfkfb3"   "Ccnl1"    "Filip1l" 
# [10] "Junb"     "Ptpn12"   "Rbpj"     "Slc30a4"  "Acp2"     "Gphn"     "Mrpl45"   "Osbpl3"   "Tmbim4"  
# [19] "Zdhhc6"   "Ccl3"     "Arhgap18" "Armc9"    "Camk1d"   "Cdipt"    "Cltc"     "Csf2ra"   "Ctbp2"   
# [28] "Fam3c"    "Fignl1"   "Fos"      "Gadd45g"  "Gdi2"     "Gpatch8"  "Hdac4"    "Il21r"    "Jarid2"  
# [37] "Klf10"    "Malat1"   "Pdk1"     "Pik3cb"   "Plekha1"  "Plekhg3"  "Psma1"    "Ptprv"    "Rcbtb2"  
# [46] "Reep3"    "Rplp1"    "Rpn1"     "Rps19"    "Rps7"     "Sh2d3c"   "Snx5"     "Srpk2"    "Ssbp3"   
# [55] "Stx16"    "Tmtc1"    "Zfp217"   "Bcl2l11"  "Fosl2"    "Phlda1"   "Sipa1l2"  "Polr3d"  


selGs <- c(
  intersect(ikzf1_comTargets, DEGs_KO1_Exvivo_ikzf1_com$external_gene_name), 
  'Socs2', 'Cish', 'Ikzf2', 'Batf3', 'Foxo1', 'Bhlhe40', 'Il2ra'
)
##----- get the info for the 159 common targets between Wil and Oh, in the ex-vivo data; 
## 156 genes out of 159 genes present in the ex-viov data
## The missing genes are: "Sh2b2" "Ninj2" "Dusp4"
DEGs_Exvivo_ikzf1_com159 <- DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name %in% ikzf1_comTargets, ]

DEGs_Exvivo_ikzf1_62DEGs_selGs <- DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name %in%  selGs, ]


write.csv(DEStatDataPrs_ikzf1_com, 
          paste0(outPath, "Ikzf1_WilTaregts_ComTargets_OhTargets.csv"), 
          row.names = F)

write.csv(oh_ikzf1_com, 
          paste0(outPath, "Ikzf1_OhTargets_ComTargets_WilTaregts.csv"), 
          row.names = F)

write.csv(DEGs_KO1_Exvivo_ikzf1_com, 
          paste0(outPath, "Ikzf1_OhTargets_WilDEGs_ExVivo.csv"), 
          row.names = F)

# 
# DEStatDataPrs_ikzf1_com <-
#   read.csv(paste0(outPath, "Ikzf1_WilTaregts_ComTargets_OhTargets.csv"),
#            stringsAsFactors = F)
```

### Venn: common targets
```{r}

library(VennDiagram)
myCol <- brewer.pal(3, "Pastel2")[2]

ven <- venn.diagram(
  x = list(
   oh_ikzf1$Gene.Symbol,
   DEStatDataPrs$Symbol[DEStatDataPrs$Target %in% c("Ikzf1 target", "Ikzf1/3 target")]
  ),
  category.names = c("Oh et al \nIkzf1 targets" , "Ikzf1 targets"),
  filename = NULL,
  # filename = paste0(figPath, 'VennDiagram_ComTarget_Ikzf1_OhWil.png'),
  output = TRUE,
  # Circles
  fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3)),
  col = c("#440154ff", '#21908dff'),
  lwd = 1,
  # lty = 'blank',
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 550 ,
  width = 550 ,
  resolution = 400,
  compression = "lzw",
   # Numbers
  cex = 1,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-110, 115),
  cat.dist = c(0.055, 0.085),
  cat.fontfamily = "sans",
  ext.pos = c(20, -20)
  # rotation = 1
)

pdf(paste0(figPath, "VennDiagram_ComTarget_Ikzf1_OhWil.pdf"), height = 3, width = 3)
grid.draw(ven)
dev.off()

# library(nVennR)
# myV <- nVennR::plotVenn(list(
#    Oh_Ikzf1 = oh_ikzf1$Gene.Symbol,
#    Goh_Ikzf1 = DEStatDataPrs$Symbol[DEStatDataPrs$Target %in% c("Ikzf1 target", "Ikzf1/3 target")]
#   ), outFile = )
```

### Waterplot (logFC) in Ikzf1/3
for 159 genes
```{r}
## 156 genes - excluding "Sh2b2" "Ninj2" "Dusp4" missing from ex-vivo data
y <- intersect(DEGs_Exvivo_ikzf1_com159$external_gene_name, DEStatDataPrs_ikzf1_com$Symbol)


currentDEStat <-  DEStatDataPrs_ikzf1_com
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]
row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat <- currentDEStat[y, ]
# currentDEStat <- currentDEStat[gg0[ordGenes], ]

# currentDEStat$Symbol <- as.factor(currentDEStat$Symbol)
# currentDEStat$Symbol <- factor(currentDEStat$Symbol, levels = rev(gg0[ordGenes]))


currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "DEGs_group")) %>%
  # dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target")) %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Consistent",
                               TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    # x = reorder(Symbol, ordGenes),
    x = Symbol,
    y = logFC,
    # fill = Target,
    fill = DEGs_group,
    color = Direction
  )) +
  geom_bar(stat = "identity") +
  # facet_wrap(. ~ Group, ncol = 3, scales = "free_x") +
  facet_wrap(. ~ Group, ncol = 3) +
  # scale_fill_manual(values = targetCols) +
  scale_fill_manual(values = degCols) +
  scale_color_manual(values = dirCols) +
  coord_flip() +
  ggtitle("In-vitro data") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

ggsave(
  filename = "WaterFall_Oh_comTargetsIkzf1_156Genes_IkzfData_InVitro_DEGsCol.pdf",
  # filename = "WaterFall_Marke_5Genes_IkzfData_InVitro_scalesFree.pdf",
  device = "pdf",
  path = figPath,
  width = 5,
  height = 20
  # width = 6,
  # height = 3.2
)
```


### Waterplot (logFC) in Ikzf1 ex vivo
```{r}
currentDEStat <- DEGs_Exvivo_ikzf1_com159
currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

row.names(currentDEStat) <- currentDEStat$external_gene_name

currentDEStat <- currentDEStat[y, ]

# missingGenes <- gg[! gg %in% currentDEStat$external_gene_name]
# missingData <-
#   data.frame(matrix(
#     ncol = ncol(currentDEStat),
#     nrow = length(missingGenes)
#   ))
# 
# colnames(missingData) <- colnames(currentDEStat)
# row.names(missingData) <- missingGenes
# missingData$external_gene_name <- missingGenes
# 
# currentDEStat <- rbind(currentDEStat[, -1], missingData[, -1])

# currentDEStat <- currentDEStat[gg[ordGenes], ]

# currentDEStat$logFC[is.na(currentDEStat$logFC)] <- 0

currentDEStat <-
  merge(currentDEStat,
        DEStatDataPrs[, c("Target", "Symbol", "DEGs_group")],
        by.x = "external_gene_name" ,
        by.y = "Symbol", all.x = T)

# currentDEStat$external_gene_name <- as.factor(currentDEStat$external_gene_name)
# 
# currentDEStat$external_gene_name <- factor(currentDEStat$external_gene_name, levels = rev(gg0[ordGenes]))

# currentDEStat$Target[is.na(currentDEStat$Target)] <- "Unknown"
# targetCols2 <- c(targetCols, "gray90")
# names(targetCols2)[5] <- "Unknown"

currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T),
    "DEGs_group"
    # "Target"
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  ggplot(., aes(# x = reorder(Symbol, ordGenes),
    x = external_gene_name,
    y = logFC,
    fill = DEGs_group
    # fill = Target
    )) +
    # color = Direction)) +
    geom_bar(stat = "identity") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      scale_fill_manual(values = degCols) +
      # scale_fill_manual(values = targetCols) +
      # scale_fill_manual(values = targetCols2) +
      coord_flip() +
      ggtitle("Ex-vivo data") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

# ggsave(
#   filename = "WaterFall_Marke_5Genes_IkzfData_ExVivo.pdf",
#   device = "pdf",
#   path = figPath,
#   width = 3.5,
#   height = 5
# )

ggsave(
  filename = "WaterFall_Oh_comTargetsIkzf1_156Genes_IkzfData_ExVivo_DEGsCol.pdf",
  # filename = "WaterFall_Marke_5Genes_IkzfData_InVitro_scalesFree.pdf",
  device = "pdf",
  path = figPath,
  width = 2,
  height = 20
  # width = 6,
  # height = 3.2
)
```


PURPLE up and GREEN down and rank in order of magnitude Log fold change. Remove the other colours which relate to DE in in vitro IKZF1, 3, 1/3 data. 
```{r}
currentDEStat <- DEGs_Exvivo_ikzf1_62DEGs_selGs
currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

row.names(currentDEStat) <- currentDEStat$external_gene_name

currentDEStat$Direction[currentDEStat$logFC > 0 ] <- "Up-regulated"
currentDEStat$Direction[currentDEStat$logFC < 0 ] <- "Down-regulated"
currentDEStat <- currentDEStat[order(currentDEStat$logFC), ]
    
currentDEStat <-
  merge(currentDEStat,
        DEStatDataPrs[, c("Target", "Symbol")],
        by.x = "external_gene_name" ,
        by.y = "Symbol", all.x = T)


currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T),
    "Direction"
    # "Target"
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  ggplot(., aes(# x = reorder(Symbol, ordGenes),
    x = reorder(external_gene_name, logFC), 
    y = logFC,
    fill = Direction
    # fill = Target
    )) +
    # color = Direction)) +
    geom_bar(stat = "identity") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      scale_fill_manual(values = c(brewer.pal(4, "Dark2")[c(1, 3)])) +
      # scale_fill_manual(values = targetCols) +
      # scale_fill_manual(values = targetCols2) +
      coord_flip() +
      ggtitle("Ex-vivo data") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

# ggsave(
#   filename = "WaterFall_Marke_5Genes_IkzfData_ExVivo.pdf",
#   device = "pdf",
#   path = figPath,
#   width = 3.5,
#   height = 5
# )

ggsave(
  filename = "WaterFall_Oh_comTargetsIkzf1_69Genes_DEGsInIkzf1Exvivo_plusSelGenes_ExVivo.pdf",
  # filename = "WaterFall_Marke_5Genes_IkzfData_InVitro_scalesFree.pdf",
  device = "pdf",
  path = figPath,
  width = 2,
  height = 12
  # width = 6,
  # height = 3.2
)
```

## Genes in Marke et al paper
https://www.haematologica.org/article/view/8410
```{r}
# c-Kit, 
# Flt3 and ## did not find Flt3 but got Flt3l
# Il7r
# SH2B3
# IGLL1,  ## did not find this
# SYK, and 
# SLP65 (BLNK) ## did not find this
# CD34 ## did not find this
# CD43 (SPN)## 
# Ctnnd1 encoding p120-catenin
# 
# IK6, ## Ikzf1
# Fak (Ptk2?) ## Ptk2b
# HK2, HK3, and 
# G6PD ## G6pdx
# TXNIP and CNR2
# THY1 
# Cdkn1a, # Cdkn1b
# Cdkn2a, and Cdk6
# BCL6 and MYC 

# grep("Myc", DEStatDataPrs$Symbol, value = T, ignore.case = T)
gs <-
  c(
    "Kit",
    "Flt3l",
    "Il7r",
    "Sh2b3",
    "Syk",
    "Spn",
    "Ctnnd1",
    "Ptk2b",
    "Hk2",
    "Hk3",
    "G6pdx",
    "Txnip",
    "Cnr2",
    "Thy1",
    "Cdkn1b",
    "Cdkn2a",
    "Cdk6",
    "Bcl6",
    "Myc"
  )

DEStatDataPrs_gs <- DEStatDataPrs[DEStatDataPrs$Symbol %in% gs, ]

write.csv(
  DEStatDataPrs_gs,
  paste0(outPath, "DEstat_Targets_MarkeGenes_Ikzf1_Ikzf3.csv"),
  row.names = F
)

DEGsTarget_Invitro_gs <- DEStatDataPrs_gs %>% 
  filter(DEGs_group != "Non-DEGs" | Target != "Non-Target gene") %>% 
  data.frame()
## get the stat of these in ex-vivo data:
# grep("CD34", DE_KO1_Exvivo$external_gene_name, value = T, ignore.case = T)
DE_KO1_Exvivo_gs <- DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name %in% gs, ]

write.csv(
  DE_KO1_Exvivo_gs,
  paste0(outPath, "DEstat_MarkeGenes_Ikzf1_Exvivo.csv"),
  row.names = F
)

DEG_KO1_Exvivo_gs <-
  DE_KO1_Exvivo_gs[DE_KO1_Exvivo_gs$adj.P.Val < 0.05, ]

## take genes that are DE in any of the data or are target in teh in vitro data:

## 19 genes
gg <- unique(c(DEGsTarget_Invitro_gs$Symbol, DEG_KO1_Exvivo_gs$external_gene_name))

# get stats for these geens in two data sets:
DEstat_invitro_gg <-
  DEStatDataPrs_gs[DEStatDataPrs_gs$Symbol %in% gg,]

DEstat_exvivo_gg <-
  DE_KO1_Exvivo_gs[DE_KO1_Exvivo_gs$external_gene_name %in% gg, ]

## Cdkn2a does not exist in the ex-vivo data
## Syk and Thy1 are nonDEG and nonTarget in invitro data
```

### Waterplot (logFC) in Ikzf1/3
```{r}
##---- for the 5 common genes:
xx <-
  DEStatDataPrs_gs %>% 
  filter(DEGs_group != "Non-DEGs") %>% 
  data.frame()

y <- intersect(DEG_KO1_Exvivo_gs$external_gene_name, xx$Symbol)
# "Ptk2b"  "Cnr2"   "Ctnnd1" "Il7r"   "Flt3l" 




currentDEStat <-  DEstat_invitro_gg[]
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat <- currentDEStat[y, ]
# currentDEStat <- currentDEStat[gg0[ordGenes], ]

# currentDEStat$Symbol <- as.factor(currentDEStat$Symbol)
# currentDEStat$Symbol <- factor(currentDEStat$Symbol, levels = rev(gg0[ordGenes]))


currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "DEGs_group")) %>%
  # dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target")) %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Consistent",
                               TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    # x = reorder(Symbol, ordGenes),
    x = Symbol,
    y = logFC,
    # fill = Target,
    fill = DEGs_group,
    color = Direction
  )) +
  geom_bar(stat = "identity") +
  # facet_wrap(. ~ Group, ncol = 3, scales = "free_x") +
  facet_wrap(. ~ Group, ncol = 3) +
  # scale_fill_manual(values = targetCols) +
  scale_fill_manual(values = degCols) +
  scale_color_manual(values = dirCols) +
  coord_flip() +
  ggtitle("In-vitro data") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

ggsave(
  filename = "WaterFall_Marke_5Genes_IkzfData_InVitro.pdf",
  # filename = "WaterFall_Marke_5Genes_IkzfData_InVitro_scalesFree.pdf",
  device = "pdf",
  path = figPath,
  width = 4,
  height = 4.5
  # width = 6,
  # height = 3.2
)
```


### Waterplot (logFC) in Ikzf1 ex vivo
```{r}
currentDEStat <- DEstat_exvivo_gg
currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

row.names(currentDEStat) <- currentDEStat$external_gene_name

currentDEStat <- currentDEStat[y, ]

# missingGenes <- gg[! gg %in% currentDEStat$external_gene_name]
# missingData <-
#   data.frame(matrix(
#     ncol = ncol(currentDEStat),
#     nrow = length(missingGenes)
#   ))
# 
# colnames(missingData) <- colnames(currentDEStat)
# row.names(missingData) <- missingGenes
# missingData$external_gene_name <- missingGenes
# 
# currentDEStat <- rbind(currentDEStat[, -1], missingData[, -1])

# currentDEStat <- currentDEStat[gg[ordGenes], ]

# currentDEStat$logFC[is.na(currentDEStat$logFC)] <- 0

currentDEStat <-
  merge(currentDEStat,
        DEStatDataPrs[, c("Target", "Symbol", "DEGs_group")],
        by.x = "external_gene_name" ,
        by.y = "Symbol", all.x = T)

# currentDEStat$external_gene_name <- as.factor(currentDEStat$external_gene_name)
# 
# currentDEStat$external_gene_name <- factor(currentDEStat$external_gene_name, levels = rev(gg0[ordGenes]))

# currentDEStat$Target[is.na(currentDEStat$Target)] <- "Unknown"
# targetCols2 <- c(targetCols, "gray90")
# names(targetCols2)[5] <- "Unknown"

currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T),
    "DEGs_group"
    # "Target"
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  ggplot(., aes(# x = reorder(Symbol, ordGenes),
    x = external_gene_name,
    y = logFC,
    fill = DEGs_group
    # fill = Target
    )) +
    # color = Direction)) +
    geom_bar(stat = "identity") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      scale_fill_manual(values = degCols) +
      # scale_fill_manual(values = targetCols) +
      # scale_fill_manual(values = targetCols2) +
      coord_flip() +
      ggtitle("Ex-vivo data") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

# ggsave(
#   filename = "WaterFall_Marke_5Genes_IkzfData_ExVivo.pdf",
#   device = "pdf",
#   path = figPath,
#   width = 3.5,
#   height = 5
# )

ggsave(
  filename = "WaterFall_Marke_5Genes_IkzfData_ExVivo.pdf",
  # filename = "WaterFall_Marke_5Genes_IkzfData_InVitro_scalesFree.pdf",
  device = "pdf",
  path = figPath,
  width = 1.8,
  height = 4.5
  # width = 6,
  # height = 3.2
)
```

## Check selected genes - PAPER
### Waterplot (logFC) in Ikzf1/3
XIAP, cIAPs (Birc2/3/5), PMAIP1, PUMA/BBC3, BAX/BAK1, BCL2, BCL2l1, BCL2l11, MCL1, CISH, SOCS1, SOCS3
```{r}
# write.csv(
#   DEStatDataPrs_ev3_ev1_export,
#   paste0(
#     outPath,
#     # "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_selColumns_ProteinClass.csv"
#     "DEStat_DEGs_Targets_Ikzf1_Ikzf3_InVitro_Exvivo_AP1Target_selColumns_ProteinClass.csv"
#   ),
#   row.names = F
# )
# gg <- c("Mcl1", "Bcl2", "Cish", "Bcl2l11")
# gg <-
#   c(
#     "Mcl1",
#     "Cish",
#     "Xiap",
#     "Birc2",
#     "Birc3",
#     "Birc5",
#     "Pmaip1",
#     "Bbc3",
#     "Bax",
#     "Bak1",
#     "Bcl2",
#     "Bcl2l1",
#     "Bcl2l2",
#     "Bcl2a1b",
#     "Bcl2a1d",
#     "Bcl2l11",
#     "Bcl2l12",
#     "Bcl2l13",
#     "Socs2",
#     "Socs3",
#     # "Socs4",
#     # "Socs6",
#     # "Socs7",
#     "Itgam",
#     "Il2ra",
#     # "Il2rb",
#     "Il2rg",
#     "Mki67"
#   )

##-------- updated gg in July 2022 based on Nick's comments
gg <-
  as.character(quote(
    c(
      Asb2,
      B4galnt2,
      Bhlhe40,
      Casp8,
      Cass4,
      Cd69,
      Cdh17,
      Dclk2,
      Evt6,
      Hhex,
      Gpr171,
      Icos,
      Icosl,
      Ikzf1,
      Ikzf3,
      Ikzf2,
      Ikzf4,
      Itpka,
      Khdc1a,
      Klra5,
      Mmp9,
      Nedd9,
      N4ra3,
      N4ra2,
      Osgin1,
      Plxdc2,
      Serpinb9b,
      Slamf6,
      Slamf7,
      Susd4,
      Tshz3,
      Tyrobp,
      Wnt10a,
      Zfp608,
      Nrp1,
      Ctnnd1,
      Sox4,
      Ptpn3,
      Mcam,
      Hic1,
      Itgb1,
      Fcrl6,
      Kctd15,
      Ccl3,
      Ifitm1,
      Ifitm2,
      Lif,
      Tmtc1,
      Sh2b2,
      Vegfc,
      Gzmk,
      Klrb1,
      Klra9,
      Mcf21,
      Hes1,
      Itag2b
    )
  ))[-1]

##------ AP-1 TFs
# gg <-
#   c(
#     'Fos',
#     'Fosb',
#     'Fosl1',
#     'Fosl2',
#     'Jun',
#     'Junb',
#     'Jund',
#     'Batf',
#     'Batf2',
#     'Batf3',
#     'Atf7',
#     'Atf6',
#     'Atf7ip',
#     'Atf6b',
#     'Atf5',
#     'Mafg',
#     'Maf'
#   )

missingGenes <- gg[! gg %in% DEStatDataPrs_ev3_ev1_export$Symbol]
# "Evt6"   "N4ra3"  "N4ra2"  "Mcf21"  "Hes1"   "Itag2b"
# missingGenes[missingGenes %in% ai$Symbol]
# [1] "Hes1"
# missingGenes[missingGenes %in% ko1_ex$Symbol]
# [1] "Hes1"

# DEStatDataPrs_gg <- DEStatDataPrs[DEStatDataPrs$Symbol %in% gg, ]
DEStatDataPrs_gg <- DEStatDataPrs_ev3_ev1_export[DEStatDataPrs_ev3_ev1_export$Symbol %in% gg, ]

currentDEStat <-  DEStatDataPrs_gg
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]
currentDEStat <- currentDEStat[! duplicated(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol


targetCols_current <- c("#00bfff", "gray10")
names(targetCols_current) <- c("Ikzf1/3 target", "Non-target")

currentDEStat <- currentDEStat %>% arrange(logFC_DKO)
currentDEStat$Symbol <- factor(currentDEStat$Symbol, levels = currentDEStat$Symbol)

# currentDEStat %>%
#   dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target")) %>%
#   



currentDEStatLong <- currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target")) %>%
  mutate(Targets = case_when(Target %in% c('Ikzf1 target',  'Ikzf1/3 target',    'Ikzf3 target') ~ "Ikzf1/3 target", TRUE ~ "Non-target")) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:6
  )

currentDEStatLong$Group <-
  factor(
    currentDEStatLong$Group,
    levels = c(
      "logFC_DKO",
      "logFC_KO1" ,
      "logFC_KO3",
      "logFC_KO1_Exvivo",
      "logFC_KO3_Exvivo"
    )
  )


  ggplot(currentDEStatLong, aes(
    # x = reorder(Symbol, ordGenes),
    x = Symbol,
    y = logFC,
    fill = Targets,
    # color = Direction
  )) +
  geom_bar(stat = "identity") +
  # facet_wrap(. ~ Group, ncol = 3, scales = "free_x") +
  facet_wrap(. ~ Group, ncol = 5) +
  # scale_fill_manual(values = currentcol) +
  scale_fill_manual(values = targetCols_current) +
  # scale_color_manual(values = dirCols) +
  coord_flip() +
  ggtitle("") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank())





# currentDEStat %>%
#   # dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "DEGs_group")) %>%
#   dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target")) %>%
#   mutate(Direction = case_when((logFC_DKO > 0 &
#                                   logFC_KO1 > 0 & logFC_KO3 > 0) |
#                                  (logFC_DKO < 0 &
#                                     logFC_KO1 < 0 &
#                                     logFC_KO3 < 0) ~ "Consistent",
#                                TRUE ~ "Different"
#   )) %>%
#   pivot_longer(
#     data = . ,
#     names_to = "Group",
#     values_to = "logFC",
#     cols = 2:4
#   ) %>%
#   ggplot(., aes(
#     # x = reorder(Symbol, ordGenes),
#     x = Symbol,
#     y = logFC,
#     fill = Target,
#     # fill = DEGs_group,
#     color = Direction
#   )) +
#   geom_bar(stat = "identity") +
#   # facet_wrap(. ~ Group, ncol = 3, scales = "free_x") +
#   facet_wrap(. ~ Group, ncol = 3) +
#   scale_fill_manual(values = targetCols) +
#   # scale_fill_manual(values = degCols) +
#   scale_color_manual(values = dirCols) +
#   coord_flip() +
#   ggtitle("In-vitro data") +
#   geom_hline(yintercept = 0) +
#   theme_bw() +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1),
#         axis.title.y = element_blank(), 
#         legend.position = "bottom", 
#         legend.direction = "vertical")

ggsave(
  filename = "WaterFall_selGenesJuly2022_InVitro_ExVivo.pdf",
  # filename = "WaterFall_selAP1TFs_July2022_InVitro_ExVivo.pdf",
  # filename = "WaterFall_selGenes_apop_IkzfData_InVitro.pdf",
  # filename = "WaterFall_Mcl1_Cish_Bcl2_IkzfData_InVitro.pdf",
  device = "pdf",
  path = figPath,
  # width = 4,
  width = 8,
  height = 8
  # height = 4
)
```

### Waterplot (logFC) in Ikzf1 ex vivo - PAPER
```{r}
# gg <- c("Mcl1", "Bcl2", "Cish", "Bcl2l11")
# gg <-
#   c(
#     "Mcl1",
#     "Cish",
#     "Xiap",
#     "Birc2",
#     "Birc3",
#     "Birc5",
#     "Pmaip1",
#     "Bbc3",
#     "Bax",
#     "Bak1",
#     "Bcl2",
#     "Bcl2l1",
#     "Bcl2l2",
#     "Bcl2a1b",
#     "Bcl2a1d",
#     "Bcl2l11",
#     "Bcl2l12",
#     "Bcl2l13",
#     "Socs2",
#     "Socs3",
#     # "Socs4",
#     # "Socs6",
#     # "Socs7", 
#     "Itgam", 
#     "Il2ra",
#     # "Il2rb",
#     "Il2rg",
#     "Mki67" 
#   )

gfile <- read.csv(paste0(dataPath, "ex vivo ikzf1 waterfall and imm.m1.m2 subset plot.csv"), stringsAsFactors = F, header = F)

gfile$V1[gfile$V1 == "Idb2 "] <- "Idb2"
gg <- gfile$V1

gg[! gg %in% DE_KO1_Exvivo$external_gene_name]
# [1] "Erg1"    "Idb2"   "CD200r2"

# 6 genes are also missing from mature data: 
# "Fgd5"    "Fut10"   "Gstt3"   "Erg1"    "Idb2"   "CD200r2"

gg <- gg[! gg %in% c("Fgd5",    "Fut10",   "Gstt3",   "Erg1",    "Idb2",   "CD200r2")]

## there are two Itgam in the data, one is down, and the other one is up. The down regulated transcript is not significant, while the up-regulated one is significant. However, looking at the in-vitro data, Itgam is down in all three conditions. For now, I consider the transcript that is down to be consistent with the in-vitro data.

# DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name == "Itgam", c(3, 4, 5, 9)]

DEstat_exvivo_gg <- DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name %in% gg, ]

## also retain only DEGs: 77 genes
DEstat_exvivo_gg <- DEstat_exvivo_gg[DEstat_exvivo_gg$adj.P.Val < 0.05, ]

currentDEStat <- DEstat_exvivo_gg
currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

# currentDEStat <- currentDEStat[! (currentDEStat$external_gene_name == "Itgam" & currentDEStat$logFC < 0), ]
row.names(currentDEStat) <- currentDEStat$external_gene_name

currentDEStat <-
  merge(currentDEStat,
        DEStatDataPrs[, c("Target", "Symbol", "DEGs_group")],
        by.x = "external_gene_name" ,
        by.y = "Symbol", all.x = T)

## order genes based on logFCs
# currentDEStat <- currentDEStat %>% arrange(logFC)
# currentDEStat$external_gene_name <-
#   factor(currentDEStat$external_gene_name,
#          levels = currentDEStat$external_gene_name)

## order genes based on teh mature data heatmap (saved in x) - see a few chuncks below
# ordGenes <- ComplexHeatmap::row_order(x)
currentDEStat$external_gene_name <-
  factor(currentDEStat$external_gene_name,
         levels = gg[rev(ordGenes)])

currentDEStat$Target[currentDEStat$Target %in% c("Ikzf1 target",  "Ikzf1/3 target")] <- "Ikzf1 target"
currentDEStat$Target[currentDEStat$Target %in% c("Ikzf3 target", "Non-Target gene")] <- "Non-target"

targetCols_current <- c("#00bfff", "gray10")
names(targetCols_current) <- c("Ikzf1 target", "Non-target")


currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T),
    # "DEGs_group"
    "Target"
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  ggplot(., aes(# x = reorder(Symbol, ordGenes),
    x = external_gene_name,
    y = logFC,
    # fill = DEGs_group
    fill = Target
    )) +
    # color = Direction)) +
    geom_bar(stat = "identity") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      # scale_fill_manual(values = degCols) +
      scale_fill_manual(values = targetCols_current) +
      # scale_fill_manual(values = targetCols) +
      # scale_fill_manual(values = targetCols2) +
      coord_flip() +
      ggtitle("Ex-vivo data") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank(), 
        # legend.position = "left", 
        # legend.direction = "vertical"
        legend.position = "bottom",
        legend.direction = "horizontal"
        )


# ggsave(
#   filename = "WaterFall_Marke_5Genes_IkzfData_ExVivo.pdf",
#   device = "pdf",
#   path = figPath,
#   width = 3.5,
#   height = 5
# )

ggsave(
  # filename = "WaterFall_Mcl1_Cish_Bcl2_IkzfData_ExVivo.pdf",
  # filename = "WaterFall_selGenes_apop_IkzfData_ExVivo.pdf",
  filename = "WaterFall_selDEGs_IkzfData_ExVivo_Jul2022.pdf",
  device = "pdf",
  path = figPath,
  # width = 1.8,
  width = 3.1,
  height = 10
  # height = 4.5
  # width = 6,
  # height = 3.2
)

ggsave(
  # filename = "WaterFall_Mcl1_Cish_Bcl2_IkzfData_ExVivo.pdf",
  # filename = "WaterFall_selGenes_apop_IkzfData_ExVivo.pdf",
  filename = "WaterFall_selDEGs_IkzfData_ExVivo_Jul2022.png",
  device = "png",
  path = figPath,
  # width = 1.8,
  width = 3,
  height = 10, units = "in"
  # height = 4.5
  # width = 6,
  # height = 3.2
)
```


### Waterplot of TFs that are DEGs in Ikzf1 ex vivo
```{r}
# 226 genes

gg <- DEGs_KO1_Exvivo_prTF$external_gene_name

currentDEStat <- DEGs_KO1_Exvivo_prTF
currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

row.names(currentDEStat) <- currentDEStat$external_gene_name

currentDEStat$Target[currentDEStat$Target %in% c("Ikzf1 target",  "Ikzf1/3 target")] <- "Ikzf1 target"
currentDEStat$Target[currentDEStat$Target %in% c("Ikzf3 target", "Non-Target gene")] <- "Non-Target gene"


currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T),
    # "DEGs_group"
    "Target"
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  ggplot(., aes(
    x = external_gene_name,
    y = logFC,
    # fill = DEGs_group
    fill = Target
    )) +
    geom_bar(stat = "identity") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      # scale_fill_manual(values = degCols) +
      scale_fill_manual(values = targetCols[3:4]) +
      coord_flip() +
      ggtitle("Ex-vivo data") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank(), 
        legend.position = "bottom", 
        legend.direction = "vertical")

# ggsave(
#   filename = "WaterFall_Marke_5Genes_IkzfData_ExVivo.pdf",
#   device = "pdf",
#   path = figPath,
#   width = 3.5,
#   height = 5
# )

ggsave(
  # filename = "WaterFall_Mcl1_Cish_Bcl2_IkzfData_ExVivo.pdf",
  filename = "WaterFall_Ikzf1DEGsTFs_ExVivo.pdf",
  device = "pdf",
  path = figPath,
  width = 2.5,
  height = 35
  # width = 6,
  # height = 3.2
)
```




## TF visulaisations

### Choose TFs
Here I am putting together a list of genes from [Maayan Lab](https://maayanlab.cloud/Harmonizome/gene_set/IKZF3/Pathway+Commons+Protein-Protein+Interactions), as well as genes selected by Nick Huntington to make more visualisations. 
```{r}
# gg <- intersect(mm_TFs_DEGs$Symbol, DEStatDataPrs_TF_DEGs_and_Targe$Symbol)
```

```{r}
## read genes from https://maayanlab.cloud/Harmonizome/gene_set/IKZF3/Pathway+Commons+Protein-Protein+Interactions
maayan <- read.csv(paste0(dataPath, "IKZF_Genes_MaayanLab.csv"), stringsAsFactors = F)

## map these to gene symbols:
maayanIDs <- biomaRt::getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = maayan$Symbol,
  mart = martHuman,
  attributesL = c("external_gene_name"),
  martL = martMouse
)

## have no changes in ikzf data
badGenes <- c(
  'Ube2i',
  'Hdac1',
  'Hras',
  'Lck',
  'Hdac2',
  'Ctbp1',
  'Hdac4',
  'Stat5b',
  'Stat5a',
  'Shc1',
  'Atpaf2',
  'Crbn',
  'Sin3b',
  'Grb2'
  )
maayanIDs <- maayanIDs[! maayanIDs$Gene.name %in% badGenes, ]

gg <- c(
  'Nfil3',
  'Runx3',
  # 'Est1' # same as Smg6?
  'Smg6',
  'Tox',
  'Id2',
  # (Idb2)
  # 'Foxo1a',
  'Foxo1',
  'Zbtb32',
  # 'Irf8', # does not exist in the Ikzf data
  'Zeb2',
  'Prdm1',
  'Tbx21',
  'Eomes',
  'Gata3',
  'Stat3',
  # 'Stat5',
  'Stat5a',
  'Stat5b',
  'Stat4',
  # 'Zfp683',
  # 'Znf683',
  # 'Plzx',  ## can't find this
  'Zhx2',
  'Tcf7',
  'Bcl11b',
  'Irf2',
  'Mef2c',
  'Fosl2',
  # 'Erg2',
  # IKZF1 targets from a recent review https://pubmed.ncbi.nlm.nih.gov/33331000/
  'Cd8a',
  # 'Rag',
  # 'Rag1',
  # 'Rag2',
  'Myc',
  'Ahr',
  # family
  'Bcl6',
  # 'Il2', # does not exist in the Ikzf data
  'Il2ra',
  # Since Batf3 is key for other lineage development we could see if other TFs that drive alternate fate are deregulated too
  # 'Pax5', # does not exist in the Ikzf data
  # 'Foxp3', 
  
  'Ikzf1', 
  'Ikzf2',
  'Ikzf3',
  # 'Ikzf4',
  'Ikzf5'
  
)

maay <- maayanIDs$Gene.name[! maayanIDs$Gene.name %in% gg]
# 40 genes
#  [1] "Hdac1"   "Grb2"    "Hdac2"   "Tcaf1"   "Ctbp1"   "Lck"    
#  [7] "Il2"     "Chd4"    "Rasa1"   "Bcl2l1"  "Hdac7"   "Dok2"   
# [13] "Dgcr6"   "Arnt2"   "Sin3a"   "Atpaf2"  "Socs3"   "Gab2"   
# [19] "Sin3b"   "Prkab2"  "Chd3"    "Rcor3"   "Rbbp8"   "Il2rb"  
# [25] "Gm20489" "Il2rg"   "Ldoc1"   "Hdac5"   "Magohb"  "Jak1"   
# [31] "Crbn"    "Ubc"     "Sos1"    "Cdkn2d"  "Shc1"    "Jak3"   
# [37] "Ube2i"   "Hdac4"   "Hras"    "Notch1" 

## remove those that do not exist in the Ikzf data:
missingInIkzf <- maay[! maay %in% DEStatDataPrs$Symbol]
# [1] "Tcaf1"   "Il2"     "Dgcr6"   "Rbbp8"   "Il2rb"   "Gm20489" "Ldoc1" 
# [8] "Sos1"
missingInMature <- maay[! maay %in% mm$Symbol]
# [1] "Tcaf1"   "Gm20489" "Ldoc1"   "Ubc"  

maay <- maay[! maay %in% unique(c(missingInIkzf, missingInMature))]
## 31 genes



gg <- unique(
  c(
    gg,
    maay,
    'Klf2',
    'Mki67',
    # Survival genes:
    'Bcl2',
    'Mcl1',
    'Bcl2l11',
    'Klrg1',
    'Cd226',
    'Cish',
    'Mtor',
    
    ## most significant ones from the DE_exvivo:
    # 'Anxa8',
    'Camk2b',
    'Socs2', 
    'Tln1'
  )
)   ## 41 genes

## remove these genes based on Nick's request in Jan 2022:

gg <- gg[! gg %in% c("Mki67", "Bcl6", "Bhlhe40", "Erg1")]

## subset gg to only those that have significant p-values in ex-vivo experiment
# gg <- gg[gg%in% DEGs_KO1_Exvivo$external_gene_name] ## 29 genes

gg[! gg %in% mm$Symbol]
gg[! gg %in% DEStatDataPrs$Symbol]
gg[! gg %in% DE_KO1_Exvivo$external_gene_name] 
# "Zbtb32" "Cd8a"   "Arnt2" 

gg0 <- gg
```

### Heatmap of genes in Mature data - PAPER
gg is the list of Ikzf1 ex vivo DEGs shared by Nick in July 2022
```{r}
library("grid")
minVal <- min(mm[mm$Symbol %in% gg, 3:5])
maxVal <- max(mm[mm$Symbol %in% gg, 3:5])

mm <- mm[! duplicated(mm$Symbol), ] ## this removes duplicated "2-Mar"
row.names(mm) <- mm$Symbol

gg0 <- gg

x <- plotHeatmapExpr(
  exprData = mm[gg0, 3:5],
  heatmapCol = viridis::viridis(100),
  genes = gg0,
  subsetSample = NULL,
  annotData = NULL,
  annotColumns = NULL,
  annotColList = NULL,
  scaleData = T,
  geneAnnot = NULL,
  geneCol = NULL,
  clustRows = T,
  clustCols = F,
  showRowNames = T,
  showColNames = T,
  textSize = 9,
  heatmapLegendTitle = "Scaled\nlogRPKM",
  plotTitle = ""
)

## the four missing genes are also missing from Ikzf data:
pdf(
  paste0(
    figPath,
    "Heatmap_selDEGsIkzf1Exvivo_MatureNK_July2022.pdf"
  ),
  height = 11.5,
  # width = 3
  width = 3
)
x
dev.off()

png(
  paste0(
    figPath,
    "Heatmap_selDEGsIkzf1Exvivo_MatureNK_July2022.png"
  ),
  height = 11.5,
  # width = 3
  width = 3, 
  unit = "in", 
  res = 300
)
x
dev.off()


ordGenes <- ComplexHeatmap::row_order(x)

```

### Waterplot (logFC) in Ikzf1/3
```{r}

currentDEStat <-  DEStatDataPrs[DEStatDataPrs$Symbol %in% gg0,]
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat <- currentDEStat[gg0[ordGenes], ]
currentDEStat$Symbol <- as.factor(currentDEStat$Symbol)
currentDEStat$Symbol <- factor(currentDEStat$Symbol, levels = rev(gg0[ordGenes]))

currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target")) %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Consistent",
                               TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    x = Symbol,
    y = logFC,
    fill = Target,
    color = Direction
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 3, scales = "free_x") +
  scale_fill_manual(values = targetCols) +
  scale_color_manual(values = dirCols) +
  coord_flip() +
  ggtitle("logFC of selected genes") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank())

ggsave(
  filename = "WaterFall_selectedTFs_IkzfData_scalesFree.pdf",
  device = "pdf",
  path = figPath,
  width = 6,
  height = 9
)
```

### Waterplot (logFC) in Ikzf1/3 - PAPER
```{r}
currentDEStat <-  DEStatDataPrs[DEStatDataPrs$Symbol %in% gg0,]
currentDEStat <- currentDEStat[complete.cases(currentDEStat$Symbol), ]

row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat <- currentDEStat[gg0[ordGenes], ]

currentDEStat$Symbol <- as.factor(currentDEStat$Symbol)
currentDEStat$Symbol <- factor(currentDEStat$Symbol, levels = rev(gg0[ordGenes]))

targetCols_current <- c("#00bfff", "gray10")
names(targetCols_current) <- c("Ikzf1/3 target", "Non-target")


currentDEStat %>%
  dplyr::select(c("Symbol", grep("logFC", colnames(currentDEStat), value = T), "Target")) %>%
  mutate(Targets = case_when(Target %in% c('Ikzf1 target',  'Ikzf1/3 target',    'Ikzf3 target') ~ "Ikzf1/3 target", TRUE ~ "Non-target")) %>% 
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    x = Symbol,
    y = logFC,
    fill = Targets,
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 3) +
  scale_fill_manual(values = targetCols_current) +
  coord_flip() +
  ggtitle("") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank())

ggsave(
  filename = "WaterFall_selectedTFs_IkzfData_InVitro_2022Jan25.pdf",
  device = "pdf",
  path = figPath,
  width = 6,
  height = 8
  # height = 5.4
)
```

### Waterplot (logFC) in Ikzf1 ex vivo
```{r}
currentDEStat <-
  DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name %in% gg0, ]
currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

row.names(currentDEStat) <- currentDEStat$external_gene_name


missingGenes <- gg0[! gg0 %in% currentDEStat$external_gene_name]
missingData <-
  data.frame(matrix(
    ncol = ncol(currentDEStat),
    nrow = length(missingGenes)
  ))

colnames(missingData) <- colnames(currentDEStat)
row.names(missingData) <- missingGenes
missingData$external_gene_name <- missingGenes

currentDEStat <- rbind(currentDEStat[, -1], missingData[, -1])

currentDEStat <- currentDEStat[gg0[ordGenes], ]

currentDEStat$logFC[is.na(currentDEStat$logFC)] <- 0

currentDEStat <-
  merge(currentDEStat,
        DEStatDataPrs[, c("Target", "Symbol")],
        by.x = "external_gene_name" ,
        by.y = "Symbol", all.x =T)

currentDEStat$external_gene_name <- as.factor(currentDEStat$external_gene_name)
currentDEStat$external_gene_name <- factor(currentDEStat$external_gene_name, levels = rev(gg0[ordGenes]))

# currentDEStat$Target[is.na(currentDEStat$Target)] <- "Unknown"
# 
# targetCols2 <- c(targetCols, "gray90")
# names(targetCols2)[5] <- "Unknown"

currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T),
    "Target"
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  mutate(Targets = case_when(Target %in% c('Ikzf1 target',  'Ikzf1/3 target',    'Ikzf3 target') ~ "Ikzf1/3 target", TRUE ~ "Non-target")) %>% 
  ggplot(., aes(# x = reorder(Symbol, ordGenes),
    x = external_gene_name,
    y = logFC,
    fill = Targets)) +
    # color = Direction)) +
    geom_bar(stat = "identity") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      scale_fill_manual(values = targetCols_current) +
      # scale_fill_manual(values =targetCols2) +
      # scale_color_manual(values = dirCols) +
      coord_flip() +
      ggtitle("") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank())

ggsave(
  # filename = "WaterFall_TFs_MatureNK_DEGs_IkzfData_DEGs_and_Target_scalesFree.pdf",
  filename = "WaterFall_SelectedTFs_PlusIkzfGenes_Ikzf1Data_Exvivo_2022Jan25.pdf",
  # filename = "WaterFall_SelectedTFs_PlusIkzfGenes_Ikzf1Data_Exvivo_2022Jan.pdf",
  device = "pdf",
  path = figPath,
  width = 3.5,
  height = 8
  # height = 5.4
)
```

## Klrg1 promoter 
### Heatmap in Maturity data
```{r}
klrg_prom <-
  read.csv(paste0(dataPath, "KLRG1_Promoter_Enhancer_GeneCard.csv"),
           stringsAsFactors = F)

klrg_promIDs <- getLDS(
  attributes = c("hgnc_symbol"),
  filters = "hgnc_symbol",
  values = klrg_prom$Genes,
  mart = martHuman,
  attributesL = c("external_gene_name"),
  martL = martMouse
)

klrg_promIDs <- klrg_promIDs[! duplicated(klrg_promIDs$Gene.name), ]

klrg1_DEStatDataPrs <- DEStatDataPrs[DEStatDataPrs$Symbol %in% klrg_promIDs$Gene.name, ] ## 188
klrg1_DEStat_Ikzf1 <- klrg1_DEStatDataPrs[grepl("Ikzf1", klrg1_DEStatDataPrs$Target), ]

klrg1_DEStat_Ikzf1_TF <- klrg1_DEStat_Ikzf1[complete.cases(klrg1_DEStat_Ikzf1$ProteinClass),] ## 46 TFs


pdf(
  paste0(
    figPath,
    # "Heatmap_IntegrinsFromIkzfData_InMatureNK.pdf"
    "Heatmap_Klrg1_promoter_TFs_InMatureNK_legend.pdf"
  ),
  height = 8,
  width = 2.5
  # width = 2.2
)
y <- plotHeatmapExpr(
  # exprData = mm[mm$Symbol %in% gg , 3:5],
  exprData = mm[klrg1_DEStat_Ikzf1_TF$Symbol, 3:5],
  heatmapCol = viridis(100),
  genes = klrg1_DEStat_Ikzf1_TF$Symbol ,
  subsetSample = NULL,
  annotData = NULL,
  annotColumns = NULL,
  annotColList = NULL,
  scaleData = T,
  geneAnnot = NULL,
  geneCol = NULL,
  clustRows = T,
  clustCols = F,
  showRowNames = T,
  showColNames = T,
  textSize = 9,
  heatmapLegendTitle = "scaled\nlogRPKM",
  plotTitle = paste0(
    "Expression of genes \nthat are TFs and \ncan bind to Klrg1\n promoter region"
  )
)
  
y
dev.off()

ordGenes_klrg <- ComplexHeatmap::row_order(y)
```


### Waterplot (logFC) in Ikzf1 ex vivo
```{r}
gKlrg <- klrg1_DEStat_Ikzf1_TF$Symbol

currentDEStat <-
  DE_KO1_Exvivo[DE_KO1_Exvivo$external_gene_name %in% gKlrg, ]
currentDEStat <-
  currentDEStat[complete.cases(currentDEStat$external_gene_name),]

row.names(currentDEStat) <- currentDEStat$external_gene_name


currentDEStat <- currentDEStat[gKlrg[ordGenes_klrg], ]

currentDEStat <-
  merge(currentDEStat,
        DEStatDataPrs[, c("Target", "Symbol")],
        by.x = "external_gene_name" ,
        by.y = "Symbol", all.x =T)

currentDEStat$external_gene_name <- as.factor(currentDEStat$external_gene_name)
currentDEStat$external_gene_name <- factor(currentDEStat$external_gene_name, levels = rev(gKlrg[ordGenes_klrg]))

currentDEStat %>%
  dplyr::select(c(
    "external_gene_name",
    grep("logFC", colnames(currentDEStat), value = T),
    "Target"
  )) %>%
  mutate(Group = "logFC_KO1_exvivo") %>%
  ggplot(., aes(
    x = external_gene_name,
    y = logFC,
    fill = Target)) +
    geom_bar(stat = "identity") +
      facet_wrap(. ~ Group, ncol = 1, scales = "free_x") +
      scale_fill_manual(values = targetCols) +
      coord_flip() +
      ggtitle("") +
      geom_hline(yintercept = 0) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            axis.title.y = element_blank())

ggsave(
  filename = "WaterFall_Klrg1_promoter_TFs_Ikzf1Data_Exvivo.pdf",
  device = "pdf",
  path = figPath,
  width = 3.5,
  height = 8
)
```


## Integrin visulaisations
### Heatmap in Maturity data
```{r}
gg <- grep("Itg", DEStatDataPrs$Symbol, value = T)

gg[! gg %in% grep("Itg", mm$Symbol, value = T)]
# Itga7

minVal <- min(mm[mm$Symbol %in% gg, 3:5])
maxVal <- max(mm[mm$Symbol %in% gg, 3:5])

missingData <- matrix(ncol = ncol(mm), nrow = 1)
colnames(missingData) <- colnames(mm)
row.names(missingData) <- "Itga7"
mm2 <- rbind(mm, missingData)

source(paste0(scriptPath, "plotHeatmapExpr.R"))
library(viridis)

pdf(
  paste0(
    figPath,
    "Heatmap_IntegrinsFromIkzfData_InMatureNK_legend.pdf"
  ),
  height = 6,
  width = 3
)
plotHeatmapExpr(
  exprData = mm2[gg, 3:5],
  heatmapCol = viridis(100),
  genes = gg ,
  subsetSample = NULL,
  annotData = NULL,
  annotColumns = NULL,
  annotColList = NULL,
  scaleData = T,
  geneAnnot = NULL,
  geneCol = NULL,
  clustRows = F,
  clustCols = F,
  showRowNames = T,
  showColNames = T,
  textSize = 9,
  heatmapLegendTitle = "scaled\nlogRPKM",
  plotTitle = paste0(
    "Expression of Integrin genes in\nthe Mature/immature NK data\nDEGs are:\nItgam and Itgb3 in mat vs imm\nItgam and Itgb2/3 in term vs imm"
  )
)
dev.off()
```

### Set colours
```{r}
targetCols <- brewer.pal(8, "Set2")[c(1, 2, 3, 8)]
names(targetCols) <- c("Ikzf1/3 target", "Ikzf3 target", "Ikzf1 target", "Non-Target gene")

dirCols <- brewer.pal(9, "Set1")[c(9, 1)]
names(dirCols) <- c("Consistent", "Different")

# degCols <- currentcol[1:7]
# names(degCols) <- unique(plotData$DEGs_group)

degCols <- c(currentcol[1:7], "gray")
names(degCols) <- c(
  "DEGs only in Ikzf1 KO",
  "DEGs in DKO & Ikzf1 KO",
  "DEGs in all KOs",
  "DEGs in DKO & Ikzf3 KO",
  "DEGs in Ikzf1 & Ikzf3 KOs",
  "DEGs only in DKO",
  "DEGs only in Ikzf3 KO"
)
```

### Waterplot (logFC) in Ikzf1/3
```{r}

currentDEStat <-  DEStatDataPrs[DEStatDataPrs$Symbol %in% gg,]

currentDEStat$Symbol <- factor(currentDEStat$Symbol, levels = rev(gg[order(gg)]))

currentDEStat %>% 
  dplyr::select(c("Symbol", grep("logFC", colnames(DEstats), value = T), "Target")) %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Consistent",
                               TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    x = Symbol, 
    y = logFC,
    fill = Target, 
    color = Direction 
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 3, scales = "free_x") +
  scale_fill_manual(values = targetCols) +
  scale_color_manual(values = dirCols) +
  coord_flip()+
  ggtitle("logFC of Integrin genes") +
  geom_hline(yintercept = 0) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank())

ggsave(
  filename = "WaterFall_Integrins_IkzfData_DEGs_scalesFree.pdf",
  device = "pdf",
  path = figPath,
  width = 6,
  height = 5.4
)
```

## DEGs that are Target genes - logFC plots
```{r}
library(ggrepel)
plotData <- DEStatDataPrs %>%
  filter(DEGs_group != "Non-DEGs") %>%
  filter(Target != "Non-Target gene") %>%
  arrange(Target)

pTarget <-
  ggplot(plotData, aes(logFC_KO3, logFC_DKO, color = Target, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1,
             show.legend = F) +
  geom_vline(xintercept = 0,
             color = "gray40",
             lty = 2.5) +
  geom_hline(yintercept = 0,
             color = "gray40",
             lty = 2.5) +
  geom_label_repel(
    data = plotData[
      # abs(plotData$logFC_KO1) > 2 |
        abs(plotData$logFC_KO3) > 2 |
      abs(plotData$logFC_DKO) > 2 , ],
    aes(label = Symbol,
        fill = factor(Target)),
    color = 'white',
    segment.colour = "black",
    size = 2.5,
    segment.size = 0.1
  ) +
  ggtitle("DEGs that are among target genes (N = 1689)") +
  scale_color_manual(values = targetCols) +
  scale_fill_manual(values = targetCols) +
  labs(fill = "Target") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction = "vertical")

ggsave(
  plot = pTarget,
  filename = "logFC_IKZF3_DKO_allDEGs_colorTargets_Macs1Union_moreDEGs.pdf",
  device = "pdf",
  path = figPath,
  width = 5,
  height = 5.8,
  dpi = 300
)
```

1689 genes that are DEGs and are target genes

## Ikzf1/3 DEGs that are Target genes 
### In-vitro logFC plots 
305 genes that are Ikzf1/3 DEGs and are target genes
```{r}
library(ggrepel)
plotData <- DEStatDataPrs_ev3_ev1_export %>%
  filter(DEGs_group == "DEGs in all KOs" | DEGs_group == "DEGs in Ikzf1 & Ikzf3 KOs") %>%
  filter(Target != "Non-Target gene") %>%
  arrange(as.factor(as.character(Target)))
 
pTarget <-
  pTarget <-  ggplot(plotData, aes(logFC_KO1, logFC_KO3, color = Target, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1,
             show.legend = F) +
  geom_vline(xintercept = 0,
             color = "gray40",
             lty = 2.5) +
  geom_hline(yintercept = 0,
             color = "gray40",
             lty = 2.5) +
  geom_label_repel(
    data = plotData[
        abs(plotData$logFC_KO3) > 2 |
      abs(plotData$logFC_KO1) > 2 , ],
    aes(label = Symbol,
        fill = factor(Target)),
    color = 'white',
    segment.colour = "black",
    size = 2.5,
    segment.size = 0.1
  ) +
  ggtitle("In-vitro experiment\nIkzf1/3 KO DEGs that are among target genes (N = 305)") +
  scale_color_manual(values = targetCols[-4]) +
  scale_fill_manual(values = targetCols[-4]) +
  labs(fill = "Target") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction = "vertical")

ggsave(
  plot = pTarget,
  filename = "logFC_Ikzf1_Ikzf3_InVitro_DEGsIkzf1.3_colorTargets.pdf",
  device = "pdf",
  path = figPath,
  width = 5,
  height = 6,
  dpi = 300
)

p <- plotly::ggplotly(pTarget)
htmlwidgets::saveWidget(as_widget(p), paste0(figPath, "DEGsIkzf1.3_InVitro_TargetGenesIkzf1.3.html"))

write.csv(plotData, paste0(outPath, "DEGsIkzf1.3_InVitro_TargetGenesIkzf1.3.csv"), row.names = F)
```




### Ex-vivo logFC plots 
43 genes that are Ikzf1/3 DEGs and are target genes
```{r}
library(ggrepel)

# plotData <- DEStatDataPrs %>%
plotData <- DEGs_ev1_ev3 %>%
  filter(Target != "Non-Target gene") %>%
  arrange(as.factor(as.character(Target)))
 
pTarget <-
  pTarget <-  ggplot(plotData, aes(logFC_KO1_Exvivo, logFC_KO3_Exvivo, color = Target, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1,
             show.legend = F) +
  geom_vline(xintercept = 0,
             color = "gray40",
             lty = 2.5) +
  geom_hline(yintercept = 0,
             color = "gray40",
             lty = 2.5) +
  geom_label_repel(
    data = plotData[
      # abs(plotData$logFC_KO1) > 2 |
        abs(plotData$logFC_KO3_Exvivo) > 1 |
      abs(plotData$logFC_KO1_Exvivo) > 1, ],
    aes(label = Symbol,
        fill = factor(Target)),
    color = 'white',
    segment.colour = "black",
    size = 2.5,
    segment.size = 0.1
  ) +
  ggtitle("Ex-vivo experiment\nIkzf1/3 KO DEGs that are among target genes (N = 43)") +
  scale_color_manual(values = targetCols[-4]) +
  scale_fill_manual(values = targetCols[-4]) +
  labs(fill = "Target") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction = "vertical")

ggsave(
  plot = pTarget,
  filename = "logFC_Ikzf1_Ikzf3_ExVivo_DEGsIkzf1.3_colorTargets.pdf",
  device = "pdf",
  path = figPath,
  width = 5,
  height = 6,
  dpi = 300
)

write.csv(plotData, paste0(outPath, "DEGsIkzf1.3_ExVivo_TargetGenesIkzf1.3.csv"), row.names = F)
```

```{r}
# degTar <-
#   DEStatDataPrs[!DEStatDataPrs$DEGs_group == "Non-DEGs" &
#                    !DEStatDataPrs$Target == "Non-Target gene",]

for(d in unique(plotData$DEGs_group)){
  d0 <- plotData[plotData$DEGs_group == d, ]
  
  p0 <- d0 %>%
  arrange(Target) %>%
  ggplot(., aes(logFC_KO1, logFC_KO3, color = Target, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1.2,
             show.legend = F) +
  geom_vline(xintercept = 0, color = "gray40", lty = 2.5) +
  geom_hline(yintercept = 0, color = "gray40", lty = 2.5) +
  geom_label_repel(
    data = d0[
      abs(d0$logFC_KO1) > 1 |
        abs(d0$logFC_KO3) > 1, ],
    aes(label = Symbol,
        fill = factor(Target)),
    color = 'white',
    segment.colour = "black",
    size = 2.5, segment.size = 0.1
  ) +
  ggtitle(paste0(d, " that are among target genes")) +
  scale_color_manual(values = targetCols) +
  scale_fill_manual(values = targetCols) +
  labs(fill = "Target") +
  theme_bw() 
  # theme(legend.position = "bottom",
  #       legend.direction = "vertical")
  
  pdf(paste0(figPath, "logFC_Ikzf1_Ikzf3_colTargets_", d, ".pdf"), height = 4.8, width = 6)  
  print(p0)
  
  dev.off()
}

```

## TFs that are targets/DEGs
```{r}
DEStatDataPrs %>% 
  filter(ProteinClass == "TF" & DEGs_group != "Non-DEGs" & Target != "Non-Target gene") %>% 
  dplyr::select(c("Symbol", grep("logFC", colnames(DEstats_export), value = T))) %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) ~ "Up",
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Down",
                               TRUE ~ "Different"
  # mutate(Direction = case_when((logFC_DKO > 0 &
  #                                 logFC_KO1 > 0 & logFC_KO3 > 0) |
  #                                (logFC_DKO < 0 &
  #                                   logFC_KO1 < 0 &
  #                                   logFC_KO3 < 0) ~ "Consistent",
  #                              TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  # group_by(Group, Direction) %>% 
  # arrange(logFC) %>% 
  # arrange(Group, Direction, Symbol) %>% 
  ggplot(., aes(
    # x = reorder(Symbol, logFC),
    x = reorder(Symbol, as.numeric(as.factor(Direction))),
    # x = Symbol,
    y = logFC,
    # fill = DEGs_group,
    color = Direction
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, nrow = 1) +
  # facet_wrap(. ~ Group, ncol = 1) +
  # scale_fill_manual(values = brewer.pal(9, "Paired")[c(1, 2, 3, 4, 7, 8, 9, 10)]) +
  scale_color_manual(values = brewer.pal(9, "Set1")[c(1, 9, 9 )]) +
  ggtitle("TFs that are DEGs and target genes in at least one Ikzf experiment") +
  coord_flip() +
  # ggtitle("DEGs only in Ikzf1 KO that are amongst the Ikzf1 target genes\nbased on MACS2 in at least one sample") +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank()
    )

ggsave(
  filename = "WaterFall_TFsInIkzfData_DEGs_Target_inAtLeastOneExperiment.pdf",
  device = "pdf",
  path = figPath,
  width = 8,
  height = 20
  # width = 24,
  # height = 7
)

```


## Massive changes unique to Ikzf1 - WaterFall
```{r}
##----- only Ikzf1 DEGs
DEstatsOnlyIkzf1 <- DEStatDataPrs[DEStatDataPrs$DEGs_group == "DEGs only in Ikzf1 KO", ]

##---- Ikzf1 DEGs: 3680 genes
DEstatsIkzf1 <- DEStatDataPrs %>% 
  filter(grepl("Ikzf1", DEGs_group) |
         grepl("DEGs in all KOs", DEGs_group))


DEstatsOnlyIkzf1 <- DEstatsOnlyIkzf1[order(DEstatsOnlyIkzf1$logFC_KO1), ]

##--- which of these are also a target gene in ChIP-seq
intersect(DEstatsOnlyIkzf1$Symbol, Ikzf1Only) 
 # [1] "1810037I17Rik" "Ahi1"          "Car2"          "Cep192"        "Cep290"        "Far1"         
 # [7] "Hint2"         "Map2k1"        "Sepsecs"       "Slx1b"         "Ybx3"          "Zfp282"   

intersect(DEstatsOnlyIkzf1$Symbol, comCalls) 
#  [1] "1810037I17Rik" "Ahi1"          "Car2"          "Cep192"        "Cep290"        "Clasp1"       
#  [7] "Far1"          "Hint2"         "Map2k1"        "Rdm1"          "Ripk2"         "Sepsecs"      
# [13] "Slx1b"         "Ttc9c"         "Ybx3"          "Zcchc7"        "Zfp282"  

intersect(DEstatsOnlyIkzf1$Symbol, csawOnly) ##  "Intu"  "Ube2f"
intersect(DEstatsOnlyIkzf1$Symbol, macsUnion) ## Cep290
#  [1] "1110002L01Rik" "1700028E10Rik" "1810030O07Rik" "1810037I17Rik" "2010111I01Rik" "4930524J08Rik"
#   [7] "Abi2"          "Acp2"          "Acsl4"         "Actr3"         "Ahcyl2"        "Ahi1"         
#  [13] "Arl4a"         "Arl5b"         "Atp2c1"        "B3galnt2"      "Batf"          "Bmf"          
#  [19] "Bnip3"         "Brip1os"       "Car2"          "Ccdc61"        "Ccdc69"        "Ccno"         
#  [25] "Cdk19"         "Cdk5rap2"      "Cdkn2c"        "Celf2"         "Cep192"        "Cep290"       
#  [31] "Cep295"        "Chordc1"       "Cit"           "Clasp1"        "Clasp2"        "Col4a3bp"     
#  [37] "Commd8"        "Coro2a"        "Cox15"         "Cpped1"        "Creb3"         "Crtc3"        
#  [43] "Csf2ra"        "Ctdsp2"        "Cux1"          "Dbndd2"        "Ddhd1"         "Desi1"        
#  [49] "Dhcr7"         "Diaph1"        "Dmap1"         "Dnaja1"        "Dnajb5"        "E2f1"         
#  [55] "Fam107b"       "Fancm"         "Far1"          "Foxn2"         "Gm10524"       "Gm13994"      
#  [61] "Gm15564"       "Gm17399"       "Gm31718"       "Gm37305"       "Gphn"          "Gtpbp3"       
#  [67] "Hint2"         "Hmgxb4"        "Hsd17b12"      "Hspa8"         "Irf2bp2"       "Isoc2b"       
#  [73] "Krcc1"         "Lama3"         "Lysmd2"        "Map2k1"        "Maz"           "Mettl21a"     
#  [79] "Mfsd14a"       "Micu1"         "Mir6236"       "Naa40"         "Nav2"          "Nek3"         
#  [85] "Nelfe"         "Nemp1"         "Nmt2"          "Nr2c2"         "Nubp1"         "Nup93"        
#  [91] "Ocrl"          "P4ha2"         "Patj"          "Paxip1"        "Pdk1"          "Phip"         
#  [97] "Plscr1"        "Polr3d"        "Ppfibp1"       "Ppm1d"         "Ptpn12"        "Ptprc"        
# [103] "Rab3ip"        "Ranbp9"        "Rdm1"          "Rest"          "Rgs9bp"        "Ripk2"        
# [109] "Rit1"          "Rnf145"        "Rtn4"          "Rtp4"          "Sccpdh"        "Sdhc"         
# [115] "Sepsecs"       "Slc25a10"      "Slc25a28"      "Slc37a3"       "Slco4a1"       "Slx1b"        
# [121] "Smim13"        "Sp4"           "Spcs3"         "Spred1"        "Stk3"          "Sypl"         
# [127] "Tanc2"         "Tbc1d14"       "Thrap3"        "Tiparp"        "Tle4"          "Tmem129"      
# [133] "Tmem131"       "Tmem154"       "Tmem170"       "Tpm4"          "Tsen34"        "Ttc9c"        
# [139] "Txlng"         "Ubb"           "Ube3a"         "Wdr48"         "Wipi2"         "Ybx3"         
# [145] "Zc3h4"         "Zcchc7"        "Zfp239"        "Zfp282"        "Zfp473"        "Zfp64"        
# [151] "Zfyve19"       "Zswim7"      

DEstatsOnlyIkzf1 %>%
# DEstatsIkzf1 %>%
  # filter(Symbol %in% intersect(DEstatsOnlyIkzf1$Symbol, macsUnion)) %>% 
  filter(grepl("Ikzf1", Target)) %>%   ## 154 genes
  dplyr::select(c("Symbol", grep("logFC", colnames(DEstats_export), value = T))) %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) ~ "Up",
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Down",
                               TRUE ~ "Different"
  # mutate(Direction = case_when((logFC_DKO > 0 &
  #                                 logFC_KO1 > 0 & logFC_KO3 > 0) |
  #                                (logFC_DKO < 0 &
  #                                   logFC_KO1 < 0 &
  #                                   logFC_KO3 < 0) ~ "Consistent",
  #                              TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  # group_by(Group, Direction) %>% 
  # arrange(logFC) %>% 
  # arrange(Group, Direction, Symbol) %>% 
  ggplot(., aes(
    # x = reorder(Symbol, logFC),
    x = reorder(Symbol, as.numeric(as.factor(Direction))),
    # x = Symbol,
    y = logFC,
    # fill = DEGs_group,
    color = Direction
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, nrow = 1) +
  # facet_wrap(. ~ Group, ncol = 1) +
  # scale_fill_manual(values = brewer.pal(9, "Paired")[c(1, 2, 3, 4, 7, 8, 9, 10)]) +
  scale_color_manual(values = brewer.pal(9, "Set1")[c(1, 9, 9 )]) +
  ggtitle("DEGs only in Ikzf1 KO that are also targets of Ikzf1") +
  coord_flip() +
  # ggtitle("DEGs only in Ikzf1 KO that are amongst the Ikzf1 target genes\nbased on MACS2 in at least one sample") +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank()
    )

ggsave(
  filename = "WaterFall_IKZF1OnlyDEGs_TargetIkzf1_moreGenes.pdf",
  device = "pdf",
  path = figPath,
  width = 8,
  height = 20
)
```


Extreme Ikzf1 DEGs
```{r}
library(ggrepel)
pTarget <- DEstats %>%
  filter(grepl("IKZF1", DEGs_group) |
         grepl("DEGs in all KOs", DEGs_group)) %>%
  filter(abs(logFC_KO1) > 3 & adj.P.Val_KO1 < 0.01) %>% 
  ## Also filter for those that have small changes in Ikzf3 and DKO
  filter(abs(logFC_KO3) < 0.5 & adj.P.Val_KO3 > 0.05) %>% 
  # filter(abs(logFC_DKO) < 1 & adj.P.Val_DKO > 0.05) %>%
  ggplot(., aes(logFC_KO1, logFC_KO3, color = DEGs_group, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1.2,
             show.legend = F) +
  geom_vline(xintercept = 0, color = "gray40", lty = 2.5) +
  geom_hline(yintercept = 0, color = "gray40", lty = 2.5) +
  geom_label_repel(
    aes(label = Symbol,
        fill = factor(DEGs_group)),
    color = 'white',
    segment.colour = "black",
    size = 2.5, segment.size = 0.2
  ) +
  ggtitle("Ikzf1 DEGs with large logFC (> 3 or < -3) and adj.pval < 0.01") +
  scale_color_manual(values = cols) +
  scale_fill_manual(values = cols) +
  labs(fill = "DEGs_group") +
  theme_bw() 
  # theme(legend.position = "bottom",
  #       legend.direction = "vertical")

ggsave(
  plot = pTarget,
  filename = "logFC_Ikzf1-ExtremeDEGs_Ikzf3-NonDEGs_colDEGs.pdf",
  device = "pdf",
  path = figPath,
  width = 8,
  height = 5.5,
  dpi = 300
)


pTarget <- DEstats %>%
  filter(grepl("IKZF1", DEGs_group) |
         grepl("DEGs in all KOs", DEGs_group)) %>%
  filter(abs(logFC_KO1) > 3 & adj.P.Val_KO1 < 0.01) %>% 
  ## Also filter for those that have small changes in Ikzf3 and DKO
  filter(abs(logFC_KO3) < 0.5 & adj.P.Val_KO3 > 0.05) %>% 
  # filter(abs(logFC_DKO) < 1 & adj.P.Val_DKO > 0.05) %>%
  ggplot(., aes(logFC_KO1, logFC_KO3, color = Target, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1.2,
             show.legend = F) +
  geom_vline(xintercept = 0, color = "gray40", lty = 2.5) +
  geom_hline(yintercept = 0, color = "gray40", lty = 2.5) +
  geom_label_repel(
    aes(label = Symbol,
        fill = factor(Target)),
    color = 'white',
    segment.colour = "black",
    size = 2.5, segment.size = 0.2
  ) +
  ggtitle("Ikzf1 DEGs with large logFC (> 3 or < -3) and adj.pval < 0.01") +
  scale_color_manual(values = currentcol) +
  scale_fill_manual(values = currentcol) +
  labs(fill = "Target") +
  theme_bw() 
  # theme(legend.position = "bottom",
  #       legend.direction = "vertical")

ggsave(
  plot = pTarget,
  filename = "logFC_Ikzf1-ExtremeDEGs_Ikzf3-NonDEGs_colTargets.pdf",
  device = "pdf",
  path = figPath,
  width = 8,
  height = 5.5,
  dpi = 300
)

```


## Ikzf genes in Maturity data - Heatmap
```{r}
gg <- grep("Ikzf", mm$Symbol, value = T)

mm2 <- mm[mm$Symbol %in% gg, ]
row.names(mm2) <- mm2$Symbol
mm2 <- mm2[, 3:5]

minVal <- min(mm2)
maxVal <- max(mm2)

source(paste0(scriptPath, "plotHeatmapExpr.R"))
library(viridis)

pdf(
  paste0(
    figPath,
    "Heatmap_IkzfGenes_InMatureNK.pdf"
  ),
  height = 2.8,
  width = 2.6
  # width = 2.2
)
plotHeatmapExpr(
  # exprData = mm[mm$Symbol %in% gg , 3:5],
  exprData = mm2,
  heatmapCol = viridis(100),
  genes = gg ,
  subsetSample = NULL,
  annotData = NULL,
  annotColumns = NULL,
  annotColList = NULL,
  scaleData = T,
  geneAnnot = NULL,
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = T,
  showColNames = T,
  textSize = 9,
  heatmapLegendTitle = "scaled\nlogRPKM",
  plotTitle = paste0(
    "Ikzf genes"
  )
)
dev.off()
```

barplot of Ikzf genes in mature data
```{r}
mm2$Gene <- row.names(mm2)
mm2Long <- pivot_longer(mm2, cols = 1:3, values_to = "logRPKM", names_to = "Sample")

pdf(
  paste0(figPath, "Barplot_IkzfGenes_InMatureNK.pdf"),
  height = 1.5,
  width = 4
)
ggplot(mm2Long, aes(Sample, logRPKM, fill = Sample)) +
  geom_bar(stat = "identity", alpha = 0.7) +
  facet_wrap( ~ Gene, ncol = 4) +
  scale_fill_manual(values = brewer.pal(8, "Blues")[c(4, 6, 8)]) +
  theme_bw() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        strip.background = element_blank(), 
        panel.grid.major = element_blank())
dev.off()
```

# Volcano plot
```{r}

ggplot(DEStatDataPrs, aes(x = logFC_KO1, y = -log10(adj.P.Val_KO1))) +
  geom_point(size = 0.2) 

currentDEStat <- DEStatDataPrs
currentDEStat <- currentDEStat[!duplicated(currentDEStat$Symbol), ]
row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat$Threshold <- FALSE

currentDEStat$Threshold[currentDEStat$Symbol %in% gg] <-
  TRUE

currentDEStat$Direction <- FALSE
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_KO1 > 0] <- "Up-regulated"
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_KO1 < 0] <- "Down-regulated"

currentDEStat$genelabels <- "Other"
currentDEStat$genelabels[currentDEStat$Threshold] <- currentDEStat$Symbol[currentDEStat$Threshold]

currentDEStat$logPval <- -log10(currentDEStat$adj.P.Val_KO1)



pdf(
  paste0(figPath, "Volcano_Ikzf1KO_labelSelectedGenes.pdf"),
  height = 8,
  width = 8
)
currentDEStatArrange <- plyr::arrange(currentDEStat, Threshold)
ggplot(currentDEStatArrange,
       aes(
         x = logFC_KO1,
         y = logPval,
         colour = Direction,
         # colour = Threshold,
         label = genelabels
       )) +
  geom_point(alpha = 1) +
  geom_hline(yintercept = 1.30, lty = 2) +
  # geom_text_repel(max.overlaps = 50) +
  geom_label_repel(
    data         = subset(currentDEStatArrange, Threshold),
    aes(label = genelabels),
    size          = 4.5,
    seed = 1234,
    box.padding   = 1,
    point.padding = 0.1,
    force         = 0.2,
    segment.size  = 0.5,
    segment.color = "black",
    max.overlaps = 55
    # direction     = "x"
  ) +
  # scale_color_manual(values = c("gray60", "purple4")) +
  scale_color_manual(values = brewer.pal(9, "Set1")[c(3, 9, 4)]) +
  
  ggtitle("Vocano plot; Ikzf1 KO vs WT (in-vitro)") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  theme_bw() +
  theme(
    legend.position = "none",
    plot.title = element_text(size = rel(2), hjust = 0.5),
    axis.title = element_text(size = rel(2.5)),
    axis.text = element_text(size = rel(2.5)),
  )
dev.off()

```


## Volcano plot: Ikzf1 DEGs - in vitro
```{r}
# gg <- head(DEStatDataPrs$Symbol[order(DEStatDataPrs$adj.P.Val_KO1)], 30)

## selected top 10 DEGs up/down from Nick's list
gg <- c(
  'Procr',
'Ctnnd1',
'Nrp1',
'Cma1',
'B4galt6',
'Sox4',
'Dock1',
'Cx3cl1',
'Cd177',
'Plxdc2',
'Cd27',
'Ccnd1',
'Ltb',
'Cntn1',
'Atp7b',
'Mmp9',
'Camk2b',
'Plcxd1',
'Kcnc1',
'Tmem119'
)

ggplot(DEStatDataPrs, aes(x = logFC_KO1, y = -log10(adj.P.Val_KO1))) +
  geom_point(size = 0.2) 

currentDEStat <- DEStatDataPrs
currentDEStat <- currentDEStat[!duplicated(currentDEStat$Symbol), ]
row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat$Threshold <- FALSE

currentDEStat$Threshold[currentDEStat$Symbol %in% gg] <-
  TRUE

currentDEStat$Direction <- FALSE
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_KO1 > 0] <- "Up-regulated"
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_KO1 < 0] <- "Down-regulated"

currentDEStat$genelabels <- "Other"
currentDEStat$genelabels[currentDEStat$Threshold] <- currentDEStat$Symbol[currentDEStat$Threshold]

currentDEStat$logPval <- -log10(currentDEStat$adj.P.Val_KO1)



pdf(
  # paste0(figPath, "Volcano_Ikzf1KO_labelTop30Genes.pdf"),
  paste0(figPath, "Volcano_Ikzf1KO_labelTop20Genes.pdf"),
  height = 8,
  width = 8
)
currentDEStatArrange <- plyr::arrange(currentDEStat, Threshold)
ggplot(currentDEStatArrange,
       aes(
         x = logFC_KO1,
         y = logPval,
         colour = Direction,
         # colour = Threshold,
         label = genelabels
       )) +
  geom_point(alpha = 1) +
  geom_hline(yintercept = 1.30, lty = 2) +
  # geom_text_repel(max.overlaps = 50) +
  geom_label_repel(
    data         = subset(currentDEStatArrange, Threshold),
    aes(label = genelabels),
    size          = 8,
    seed = 1234,
    box.padding   = 1,
    point.padding = 0.1,
    force         = 0.2,
    segment.size  = 0.5,
    segment.color = "black",
    max.overlaps = 55
    # direction     = "x"
  ) +
  # scale_color_manual(values = c("gray60", "purple4")) +
  scale_color_manual(values = brewer.pal(9, "Set1")[c(3, 9, 4)]) +
  
  ggtitle("Ikzf1 KO vs WT (in-vitro)") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  theme_bw() +
  theme(
    legend.position = "none",
    plot.title = element_text(size = rel(2), hjust = 0.5),
    axis.title = element_text(size = rel(2.5)),
    axis.text = element_text(size = rel(2.5)),
  )
dev.off()

```


## Volcano plot: Ikzf3 DEGs - in vitro
```{r}
# gg <- head(DEStatDataPrs$Symbol[order(DEStatDataPrs$adj.P.Val_KO3)], 30)
gg <- c(
  'B4galt6',
'Cd177',
'Tmem40',
'Gzmg',
'Cma1',
'Gpr87',
'Nrp1',
'Fut7',
'Plxdc2',
'Zbtb32',

'Mcam',
# 'P2rx7',
'Ccl9',
'Bcl11b',
'Tlr7',
'Itgad',
'Cd74',
'Cmklr1',
'S1pr5',
'Slfn8',
'Ccl5'
)

ggplot(DEStatDataPrs, aes(x = logFC_KO3, y = -log10(adj.P.Val_KO3))) +
  geom_point(size = 0.2) 

currentDEStat <- DEStatDataPrs
currentDEStat <- currentDEStat[!duplicated(currentDEStat$Symbol), ]
row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat$Threshold <- FALSE

currentDEStat$Threshold[currentDEStat$Symbol %in% gg] <-
  TRUE

currentDEStat$Direction <- FALSE
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_KO3 > 0] <- "Up-regulated"
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_KO3 < 0] <- "Down-regulated"

currentDEStat$genelabels <- "Other"
currentDEStat$genelabels[currentDEStat$Threshold] <- currentDEStat$Symbol[currentDEStat$Threshold]

currentDEStat$logPval <- -log10(currentDEStat$adj.P.Val_KO3)



pdf(
  paste0(figPath, "Volcano_Ikzf3KO_labelTop20Genes.pdf"),
  height = 8,
  width = 8
)
currentDEStatArrange <- plyr::arrange(currentDEStat, Threshold)
ggplot(currentDEStatArrange,
       aes(
         x = logFC_KO3,
         y = logPval,
         colour = Direction,
         # colour = Threshold,
         label = genelabels
       )) +
  geom_point(alpha = 1) +
  geom_hline(yintercept = 1.30, lty = 2) +
  # geom_text_repel(max.overlaps = 50) +
  geom_label_repel(
    data         = subset(currentDEStatArrange, Threshold),
    aes(label = genelabels),
    size          = 8,
    seed = 1234,
    box.padding   = 1,
    point.padding = 0.1,
    force         = 0.2,
    segment.size  = 0.5,
    segment.color = "black",
    max.overlaps = 55
    # direction     = "x"
  ) +
  # scale_color_manual(values = c("gray60", "purple4")) +
  scale_color_manual(values = brewer.pal(9, "Set1")[c(3, 9, 4)]) +
  
  ggtitle("Ikzf3 KO vs WT (in-vitro)") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  theme_bw() +
  theme(
    legend.position = "none",
    plot.title = element_text(size = rel(2), hjust = 0.5),
    axis.title = element_text(size = rel(2.5)),
    axis.text = element_text(size = rel(2.5)),
  )
dev.off()

```


## Volcano plot: DKO DEGs - in vitro
```{r}

# gg <- head(DEStatDataPrs$Symbol[order(DEStatDataPrs$adj.P.Val_DKO)], 30)
gg <- c(
  'Procr',
  'Dock1',
  'Ccl17',
  'Sox4',
  'Slamf6',
  'Nrp1',
  'Tgfbr3',
  # 'Batf3',
  'C1qtnf1',
  'Cma1',
  'Cd40',
  'Kndc1',
  'Fcrl6',
  'Ltb4r1',
  'Prf1',
  'Mcam',
  'Susd4',
  'Itgax',
  'Sh2d1b2',
  'Entpd1',
  'Itgam'
)
ggplot(DEStatDataPrs, aes(x = logFC_DKO, y = -log10(adj.P.Val_DKO))) +
  geom_point(size = 0.2) 

currentDEStat <- DEStatDataPrs
currentDEStat <- currentDEStat[!duplicated(currentDEStat$Symbol), ]
row.names(currentDEStat) <- currentDEStat$Symbol

currentDEStat$Threshold <- FALSE

currentDEStat$Threshold[currentDEStat$Symbol %in% gg] <-
  TRUE

currentDEStat$Direction <- FALSE
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_DKO > 0] <- "Up-regulated"
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC_DKO < 0] <- "Down-regulated"

currentDEStat$genelabels <- "Other"
currentDEStat$genelabels[currentDEStat$Threshold] <- currentDEStat$Symbol[currentDEStat$Threshold]

currentDEStat$logPval <- -log10(currentDEStat$adj.P.Val_DKO)



pdf(
  paste0(figPath, "Volcano_DKO_labelTop20Genes.pdf"),
  height = 8,
  width = 8
)
currentDEStatArrange <- plyr::arrange(currentDEStat, Threshold)
ggplot(currentDEStatArrange,
       aes(
         x = logFC_DKO,
         y = logPval,
         colour = Direction,
         # colour = Threshold,
         label = genelabels
       )) +
  geom_point(alpha = 1) +
  geom_hline(yintercept = 1.30, lty = 2) +
  # geom_text_repel(max.overlaps = 50) +
  geom_label_repel(
    data         = subset(currentDEStatArrange, Threshold),
    aes(label = genelabels),
    size          = 8,
    seed = 1234,
    box.padding   = 1,
    point.padding = 0.1,
    force         = 0.2,
    segment.size  = 0.5,
    segment.color = "black",
    max.overlaps = 55
    # direction     = "x"
  ) +
  # scale_color_manual(values = c("gray60", "purple4")) +
  scale_color_manual(values = brewer.pal(9, "Set1")[c(3, 9, 4)]) +
  
  ggtitle("DKO vs WT (in-vitro)") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  theme_bw() +
  theme(
    legend.position = "none",
    plot.title = element_text(size = rel(2), hjust = 0.5),
    axis.title = element_text(size = rel(2.5)),
    axis.text = element_text(size = rel(2.5)),
  )
dev.off()

```

## Volcano plot: Ex-vivo
```{r}
ggplot(DE_KO1_Exvivo, aes(x = logFC, y = -log10(adj.P.Val))) +
  geom_point(size = 0.2) 

currentDEStat <- DE_KO1_Exvivo
currentDEStat <- currentDEStat[!duplicated(currentDEStat$external_gene_name), ]
row.names(currentDEStat) <- currentDEStat$external_gene_name

# currentDEStat$Threshold <- FALSE
# 
# currentDEStat$Threshold[abs(currentDEStat$logFC_KO1) > 2 &
#                           (-log10(currentDEStat$adj.P.Val_KO1) > 5)] <-
#   TRUE


currentDEStat$Threshold <- FALSE

currentDEStat$Threshold[currentDEStat$external_gene_name %in% gg] <-
  TRUE

currentDEStat$Direction <- FALSE
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC > 0] <- "Up-regulated"
currentDEStat$Direction[currentDEStat$Threshold &
                          currentDEStat$logFC < 0] <- "Down-regulated"

currentDEStat$genelabels <- "Other"
currentDEStat$genelabels[currentDEStat$Threshold] <- currentDEStat$external_gene_name[currentDEStat$Threshold]

currentDEStat$logPval <- -log10(currentDEStat$adj.P.Val)


pdf(
  paste0(figPath, "Volcano_Ikzf1KO_ExVivo_labelSelectedGenes_onlySignif.pdf"),
  height = 8,
  width = 8
)
currentDEStatArrange <- plyr::arrange(currentDEStat, Threshold)
ggplot(currentDEStatArrange,
       aes(
         x = logFC,
         y = logPval,
         colour = Direction,
         # colour = Threshold,
         label = genelabels
       )) +
  geom_point(alpha = 1, size = 2) +
  geom_hline(yintercept = 1.30, lty = 2) +
  # geom_text_repel(max.overlaps = 50) +
  geom_label_repel(
    data         = subset(currentDEStatArrange, Threshold),
    aes(label = genelabels), 
    fontface = "bold",
    size          = 6,
    seed = 1234,
    box.padding   = 1,
    point.padding = 0.1,
    force         = 0.2,
    segment.size  = 0.8,
    segment.color = "black",
    max.overlaps = 55
    # direction     = "x"
  ) +
  # scale_color_manual(values = c("gray60", "purple4")) +
  scale_color_manual(values = brewer.pal(9, "Set1")[c(3, 9, 4)]) +
  
  ggtitle("Vocano plot; Ikzf1 KO vs WT (ex-vivo)") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  theme_bw() +
  theme(
    legend.position = "none",
    plot.title = element_text(size = rel(2), hjust = 0.5),
    axis.title = element_text(size = rel(2.5)),
    axis.text = element_text(size = rel(2.5)),
  )
dev.off()


# 
# ggplot(currentDEStat, aes(
#     x = logFC,
#     y = logPval,
#     colour = Threshold, label = genelabels
#   )) + 
#   geom_point(alpha = 0.5) +
#   geom_label_repel(
#     data = currentDEStat[currentDEStat$Threshold, ],
#     aes(label = external_gene_name,
#         fill = factor(Threshold)
#         ),
#     color = 'white',
#     segment.colour = "black",
#     size = 2.5,
#     segment.size = 0.1
#   ) 
# 
# ggplot(data=z, aes(AveExpr, logFC)) +
# + geom_point(aes(color=DE)) + scale_color_manual(values = c('0'='black', '1'='red', '-1' = 'blue')) +
# + geom_label_repel(data=subset(z, (z$logFC > 2) & ( z$AveExpr > 3)), aes(label=external_gene_name)) +
# + geom_label_repel(data=subset(z, (z$logFC > 1) & ( z$AveExpr > 8)), aes(label=external_gene_name)) +
# + geom_label_repel(data=subset(z, (z$logFC < -2) & ( z$AveExpr > 3)), aes(label=external_gene_name)) +
# + geom_label_repel(data=subset(z, grepl("Ikzf1", external_gene_name)), aes(label=external_gene_name), color="red") + ggtitle("Ikzf1 KO vs Cre Control") +
# + theme_Publication()
```

# GO/Pathway analysis 
## Using Ikzf1 target genes
```{r}
gPath <- "/Users/mfor0011/Documents/data/geneAnnotation/NCBI_GeneInfo/"
annot <-
  read.delim(paste0(gPath, "/Mus_musculus.gene_info"))[, -1]
annot <- annot[! duplicated(annot$Symbol) & ! is.na(annot$Symbol), ]

Ikzf1Targets <- DEStatDataPrs[DEStatDataPrs$Target %in% c("Ikzf1 target" , "Ikzf1/3 target"), ]

Ikzf1Targets <-
  merge(Ikzf1Targets, annot[, 1:2], by.x = "Symbol",
        by.y = "Symbol")

```

```{r}
compr <- "Ikzf1Targets"
for(i in compr) {
  # currentDEGs <- get(paste0("DEGsID_", i))
  # currentDEGs <- currentDEGs[order(currentDEGs$logFC),]
 currentDEGs <- Ikzf1Targets
  ##----- KEGG and GO analysis
  IDs_DEGs <-
    as.character(currentDEGs$GeneID)[complete.cases(as.character(currentDEGs$GeneID))]
  
  universeGenes <- NULL

  kg <- kegga(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes, 
    convert = T)

  write.table(
    topKEGG(kg, number = Inf),
    paste0(outPath, "KEGG_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )

  go <- goana(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes)

  go <- go[go$N >=20 & go$N <= 1500, ]
  
  write.table(
    topGO(go, ontology = "BP", number = Inf),
    paste0(outPath, "GO_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )
}
```

## Using Ikzf1/3 target genes
```{r}
Ikzf13Targets <- DEStatDataPrs[DEStatDataPrs$Target %in% c("Ikzf1/3 target"), ]

Ikzf13Targets <-
  merge(Ikzf13Targets, annot[, 1:2], by.x = "Symbol",
        by.y = "Symbol")

```

```{r}
compr <- "Ikzf13Targets"
for(i in compr) {
  # currentDEGs <- get(paste0("DEGsID_", i))
  # currentDEGs <- currentDEGs[order(currentDEGs$logFC),]
 currentDEGs <- Ikzf13Targets
  ##----- KEGG and GO analysis
  IDs_DEGs <-
    as.character(currentDEGs$GeneID)[complete.cases(as.character(currentDEGs$GeneID))]
  
  universeGenes <- NULL

  kg <- kegga(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes, 
    convert = T)

  write.table(
    topKEGG(kg, number = Inf),
    paste0(outPath, "KEGG_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )

  go <- goana(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes)

  go <- go[go$N >=20 & go$N <= 1500, ]
  
  write.table(
    topGO(go, ontology = "BP", number = Inf),
    paste0(outPath, "GO_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )
}
```

# GO/pathway analysis in Ex-vivo data
## kegga & goana
```{r}
currentData <- ex_ex
compr <- c("Exvivo")

load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c2_v5p2.rdata"
))
load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c3_v5p2.rdata"
))
load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_H_v5p2.rdata"
))

load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c5_v5p2.rdata"
))
load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c7_v5p2.rdata"
))
```

## Get Entrez Gene IDs 
```{r}
gPath <- "/Users/mfor0011/Documents/data/geneAnnotation/NCBI_GeneInfo/"
annot <-
  read.delim(paste0(gPath, "/Mus_musculus.gene_info"))[, -1]
annot <- annot[! duplicated(annot$Symbol) & ! is.na(annot$Symbol), ]

DEGs_KO1_Exvivo$external_gene_name[! DEGs_KO1_Exvivo$external_gene_name %in% annot$Symbol]
## 176 genes

DEGsID_Exvivo <-
  merge(DEGs_KO1_Exvivo, annot[, 1:2], by.x = "external_gene_name",
        by.y = "Symbol")

head(currentData)
## missing ~ 400 genes
exID <- merge(annot[, 1:2], currentData, by.x = "Symbol", by.y = "row.names")
exID_rnames <- exID$GeneID
exID <- as.matrix(exID[, 3:ncol(exID)])
row.names(exID) <- exID_rnames
```


```{r}
for(i in compr) {
  currentDEGs <- get(paste0("DEGsID_", i))
  currentDEGs <- currentDEGs[order(currentDEGs$logFC),]

  ##----- KEGG and GO analysis
  IDs_DEGs <-
    as.character(currentDEGs$GeneID)[complete.cases(as.character(currentDEGs$GeneID))]
  
  universeGenes <-
    as.character(row.names(exID))

  kg <- kegga(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes, convert = T)

  write.table(
    topKEGG(kg, number = Inf),
    paste0(outPath, "KEGG_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )

  go <- goana(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes)

  go <- go[go$N >=20 & go$N <= 1500, ]
  
  write.table(
    topGO(go, ontology = "BP", number = Inf),
    paste0(outPath, "GO_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )
}
```


## Gene set testing
### FRY 
KEGG pathways and hallmarks
```{r}

for(i in compr) {
  ##----- C2
  idx <- ids2indices(Mm.c2, row.names(exID))
  
  kegg_pathways <- grep("KEGG", names(idx))
  kegg_idx <- idx[kegg_pathways]
  
  kegg_fry <-
    fry(exID,
        index = kegg_idx,
        design = design_exvivo,
        contrast = cont.mat_exvivo)
  
  write.table(
    kegg_fry,
    paste0(outPath, "FRY_KEGG_MSigDB_C2_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
 ###----- Hallmark 
idx <- ids2indices(Mm.H, row.names(exID))

h_fry <-
  fry(
    exID,
    index = idx,
    design = design_exvivo,
    contrast = cont.mat_exvivo
  )

  write.table(
    h_fry,
    paste0(outPath, "FRY_Hallmarks_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
}

```


### CAMERA 
GO, Hallmark and C7  
```{r}

for (i in compr) {
  idx <- ids2indices(Mm.c5, row.names(exID))
  go_camera <-
    camera(
      exID,
      index = idx,
      design = design_exvivo,
      contrast = cont.mat_exvivo,
      inter.gene.cor = 0.01
    )
  
  write.table(
    go_camera,
    paste0(outPath, "CAMERA_GO_MSigDB_C5_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##--- Hallmark
  idx <- ids2indices(Mm.H, row.names(exID))
  h_camera <-
    camera(
      exID,
      index = idx,
      design = design_exvivo,
      contrast = cont.mat_exvivo,
      inter.gene.cor = 0.01
    )
  
  write.table(
    h_camera,
    paste0(outPath, "CAMERA_Hallmark_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##----- C7
  idx <- ids2indices(Mm.c7, row.names(exID))
  c7_camera <-
    camera(
      exID,
      index = idx,
      design = design_exvivo,
      contrast = cont.mat_exvivo,
      inter.gene.cor = 0.01
    )
  
  write.table(
    c7_camera,
    paste0(outPath, "CAMERA_MSigDB_C7_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##----- C3
  idx <- ids2indices(Mm.c3, row.names(exID))
  c3_camera <-
    camera(
      exID,
      index = idx,
      design = design_exvivo,
      contrast = cont.mat_exvivo,
      inter.gene.cor = 0.01
    )
  
  write.table(
    c3_camera,
    paste0(outPath, "CAMERA_MSigDB_C3_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##----- C2
  idx <- ids2indices(Mm.c2, row.names(exID))
  c2_camera <-
    camera(
      exID,
      index = idx,
      design = design_exvivo,
      contrast = cont.mat_exvivo,
      inter.gene.cor = 0.01
    )
  
  write.table(
    c2_camera,
    paste0(outPath, "CAMERA_MSigDB_C2_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
}

```

# GO/pathway analysis in vitro
## kegga & goana
```{r}
currentData <- ex
compr <- c("DKO", "KO1", "KO3")

load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c2_v5p2.rdata"
))
load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c3_v5p2.rdata"
))
load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_H_v5p2.rdata"
))

load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c5_v5p2.rdata"
))
load(url(
  "http://bioinf.wehi.edu.au/software/MSigDB/mouse_c7_v5p2.rdata"
))
```

## Get Entrez Gene IDs 
```{r}
gPath <- "/Users/mfor0011/Documents/data/geneAnnotation/NCBI_GeneInfo/"
annot <-
  read.delim(paste0(gPath, "/Mus_musculus.gene_info"))[, -1]
annot <- annot[! duplicated(annot$Symbol) & ! is.na(annot$Symbol), ]

DEGs_DKO$external_gene_name[! DEGs_DKO$external_gene_name %in% annot$Symbol]

DEGsID_DKO <-
  merge(DEGs_DKO, annot[, 1:2], by.x = "external_gene_name",
        by.y = "Symbol")

DEGsID_KO1 <-
  merge(DEGs_KO1, annot[, 1:2], by.x = "external_gene_name",
        by.y = "Symbol")

DEGsID_KO3 <-
  merge(DEGs_KO3, annot[, 1:2], by.x = "external_gene_name",
        by.y = "Symbol")


head(ex)
## missing ~ 400 genes
exID <- merge(annot[, 1:2], ex, by.x = "Symbol", by.y = "row.names")
exID_rnames <- exID$GeneID
exID <- as.matrix(exID[, 3:ncol(exID)])
row.names(exID) <- exID_rnames
```

```{r}
for(i in compr) {
  currentDEGs <- get(paste0("DEGsID_", i))
  currentDEGs <- currentDEGs[order(currentDEGs$logFC),]

  ##----- KEGG and GO analysis
  IDs_DEGs <-
    as.character(currentDEGs$GeneID)[complete.cases(as.character(currentDEGs$GeneID))]
  
  universeGenes <-
    as.character(row.names(exID))

  kg <- kegga(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes, convert = T)

  write.table(
    topKEGG(kg, number = Inf),
    paste0(outPath, "KEGG_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )

  go <- goana(de = IDs_DEGs,
    species = "Mm",
    universe = universeGenes)

  go <- go[go$N >=20 & go$N <= 1500, ]
  
  write.table(
    topGO(go, ontology = "BP", number = Inf),
    paste0(outPath, "GO_", i, ".txt"),
    row.names = T,
    sep = "\t"
  )
}
```


## Gene set testing
### FRY 
KEGG pathways and hallmarks
```{r}

for(i in compr) {
  ##----- C2
  idx <- ids2indices(Mm.c2, row.names(exID))
  
  kegg_pathways <- grep("KEGG", names(idx))
  kegg_idx <- idx[kegg_pathways]
  
  kegg_fry <-
    fry(exID,
        index = kegg_idx,
        design = design,
        contrast = cont.mat[, i])
  
  write.table(
    kegg_fry,
    paste0(outPath, "FRY_KEGG_MSigDB_C2_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
 ###----- Hallmark 
idx <- ids2indices(Mm.H, row.names(exID))

h_fry <-
  fry(
    exID,
    index = idx,
    design = design,
    contrast = cont.mat[, i]
  )

  write.table(
    h_fry,
    paste0(outPath, "FRY_Hallmarks_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
}

```


### CAMERA 
GO, Hallmark and C7  
```{r}

for (i in compr) {
  idx <- ids2indices(Mm.c5, row.names(exID))
  go_camera <-
    camera(
      exID,
      index = idx,
      design = design,
      contrast = cont.mat[, i],
      inter.gene.cor = 0.01
    )
  
  write.table(
    go_camera,
    paste0(outPath, "CAMERA_GO_MSigDB_C5_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##--- Hallmark
  idx <- ids2indices(Mm.H, row.names(exID))
  h_camera <-
    camera(
      exID,
      index = idx,
      design = design,
      contrast = cont.mat[, i],
      inter.gene.cor = 0.01
    )
  
  write.table(
    h_camera,
    paste0(outPath, "CAMERA_Hallmark_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##----- C7
  idx <- ids2indices(Mm.c7, row.names(exID))
  c7_camera <-
    camera(
      exID,
      index = idx,
      design = design,
      contrast = cont.mat[, i],
      inter.gene.cor = 0.01
    )
  
  write.table(
    c7_camera,
    paste0(outPath, "CAMERA_MSigDB_C7_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##----- C3
  idx <- ids2indices(Mm.c3, row.names(exID))
  c3_camera <-
    camera(
      exID,
      index = idx,
      design = design,
      contrast = cont.mat[, i],
      inter.gene.cor = 0.01
    )
  
  write.table(
    c3_camera,
    paste0(outPath, "CAMERA_MSigDB_C3_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
  ##----- C2
  idx <- ids2indices(Mm.c2, row.names(exID))
  c2_camera <-
    camera(
      exID,
      index = idx,
      design = design,
      contrast = cont.mat[, i],
      inter.gene.cor = 0.01
    )
  
  write.table(
    c2_camera,
    paste0(outPath, "CAMERA_MSigDB_C2_", i, ".txt"),
    sep = "\t",
    row.names = T
  )
  
}

```


# Compare pathway analysis results from different group comparisons
Here we look at the top significant pathways in different comparisons performed above and make a heatmap to see the level of similarities and differences between different comparisons at the pathway level.

## Hallmark (camera)
Read the results.
```{r}

fileList <- list.files(outPath)
hh <- fileList[grepl("CAMERA_Hallmark", fileList)]
hPath <- paste0(outPath, hh)

hList <- lapply(hPath, read.table, header = T, sep = "\t")

names(hList) <- gsub("CAMERA_Hallmark_", "", hh)
names(hList) <- gsub(".txt", "", names(hList) )

hList<- plyr::llply(hList, function(x){
  x$GeneSet <- rownames(x)
  x$Directions[x$Direction == "Up"] <- 1
  x$Directions[x$Direction == "Down"] <- -1
  return(x)
})

hListSignif <- plyr::llply(hList, function(x){
  x <- x[x$FDR < 0.05, ]
  return(x)
})

hData <- plyr::ldply(hListSignif)


hFDR <- plyr::ldply(hList, function(x){ 
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "FDR"), drop = F])
  })

hWideFDR <-
  tidyr::pivot_wider(hFDR, names_from = .id, values_from = FDR)
hMatFDR <-  as.matrix(hWideFDR[, 2:ncol(hWideFDR)])
rownames(hMatFDR) <- hWideFDR$GeneSet

rownames(hMatFDR) <- gsub("HALLMARK_", "", rownames(hMatFDR) )


hDir <- plyr::ldply(hList, function(x){
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "Directions"), drop = F])
  })

hWideDir <- tidyr::pivot_wider(hDir, names_from = .id, values_from = Directions)
hMatDir <-  as.matrix(hWideDir[, 2:ncol(hWideDir)])
rownames(hMatDir) <- hWideDir$GeneSet

rownames(hMatDir) <- gsub("HALLMARK_", "", rownames(hMatDir) )

hMatDir[hMatFDR > 0.05] <- 0

```

```{r hallmark, fig.height = 10, fig.width = 8}
textSize <- 8
sc <- 1.4
library(grid)
library(ComplexHeatmap)

htmp <- ComplexHeatmap::Heatmap(
  hMatDir,
  col = c(brewer.pal(11, "PRGn")[10], "gray80", brewer.pal(11, "PRGn")[2]),
  column_title = "Significant hallmark gene-sets\n(Camera gene set test)",
  cluster_rows = T,
  cluster_columns = T,
  # cluster_columns = F,
  show_row_names = T,
  show_column_names = T,
  row_names_gp = gpar(fontsize = textSize),
  clustering_distance_rows = "manhattan",
  clustering_method_rows = "ward.D2",
  clustering_distance_columns = "manhattan",
  clustering_method_columns = "ward.D2",
  heatmap_legend_param = list(
    at = c(1, 0, -1),
    labels = c("Up-regulated", "Non-significant", "Down-regulated"),
    title = "Direction", grid_height = unit(0.7, "cm"),
    title_gp = gpar(fontsize = textSize*sc, fontface = "italic"),
    labels_gp = gpar(fontsize = textSize*sc)
  )
) 

pdf(
  paste0(figPath, "Heatmap_Hallmark_Camera_Comparisons.pdf"),
  height = 3.5,
  width = 5.2
)
draw(htmp, heatmap_legend_side = "left")
dev.off()
```


## Hallmark (fry)
Read the results.
```{r}

fileList <- list.files(outPath)
hh <- fileList[grepl("FRY_Hallmark", fileList)]
hPath <- paste0(outPath, hh)

hList <- lapply(hPath, read.table, header = T, sep = "\t")

names(hList) <- gsub("FRY_Hallmark_", "", hh)
names(hList) <- gsub(".txt", "", names(hList) )

hList<- plyr::llply(hList, function(x){
  x$GeneSet <- rownames(x)
  x$Directions[x$Direction == "Up"] <- 1
  x$Directions[x$Direction == "Down"] <- -1
  return(x)
})

hListSignif <- plyr::llply(hList, function(x){
  x <- x[x$FDR < 0.05, ]
  return(x)
})

hData <- plyr::ldply(hListSignif)


hFDR <- plyr::ldply(hList, function(x){ 
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "FDR"), drop = F])
  })

hWideFDR <-
  tidyr::pivot_wider(hFDR, names_from = .id, values_from = FDR)
hMatFDR <-  as.matrix(hWideFDR[, 2:ncol(hWideFDR)])
rownames(hMatFDR) <- hWideFDR$GeneSet

rownames(hMatFDR) <- gsub("HALLMARK_", "", rownames(hMatFDR) )


hDir <- plyr::ldply(hList, function(x){
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "Directions"), drop = F])
  })

hWideDir <- tidyr::pivot_wider(hDir, names_from = .id, values_from = Directions)
hMatDir <-  as.matrix(hWideDir[, 2:ncol(hWideDir)])
rownames(hMatDir) <- hWideDir$GeneSet

rownames(hMatDir) <- gsub("HALLMARK_", "", rownames(hMatDir) )
colnames(hMatDir) <- gsub("FRY_Hallmarks_", "", colnames(hMatDir) )

hMatDir[hMatFDR > 0.05] <- 0

```

```{r hallmark, fig.height = 10, fig.width = 8}
textSize <- 8
sc <- 1.4
library(grid)
library(ComplexHeatmap)

htmp <- ComplexHeatmap::Heatmap(
  hMatDir,
  col = c(brewer.pal(11, "PRGn")[10], "gray80", brewer.pal(11, "PRGn")[2]),
  column_title = "Significant hallmark gene-sets\n(Fry gene set test)",
  cluster_rows = T,
  cluster_columns = T,
  # cluster_columns = F,
  show_row_names = T,
  show_column_names = T,
  row_names_gp = gpar(fontsize = textSize),
  clustering_distance_rows = "manhattan",
  clustering_method_rows = "ward.D2",
  clustering_distance_columns = "manhattan",
  clustering_method_columns = "ward.D2",
  heatmap_legend_param = list(
    at = c(1, 0, -1),
    labels = c("Up-regulated", "Non-significant", "Down-regulated"),
    title = "Direction", grid_height = unit(0.7, "cm"),
    title_gp = gpar(fontsize = textSize*sc, fontface = "italic"),
    labels_gp = gpar(fontsize = textSize*sc)
  )
) 

pdf(
  paste0(figPath, "Heatmap_Hallmark_Fry_Comparisons.pdf"),
  height = 5,
  width = 5.2
)
draw(htmp, heatmap_legend_side = "left")
dev.off()
```


## KEGG (fry)
Read the results.
```{r}

fileList <- list.files(outPath)
hh <- fileList[grepl("FRY_KEGG", fileList)]
hPath <- paste0(outPath, hh)

hList <- lapply(hPath, read.table, header = T, sep = "\t")

names(hList) <- gsub("FRY_KEGG_MSigDB_C2_", "", hh)
names(hList) <- gsub(".txt", "", names(hList) )

hList<- plyr::llply(hList, function(x){
  x$GeneSet <- rownames(x)
  x$Directions[x$Direction == "Up"] <- 1
  x$Directions[x$Direction == "Down"] <- -1
  return(x)
})

hListSignif <- plyr::llply(hList, function(x){
  # x <- x[x$FDR < 0.01, ]
  x <- x[order(x$FDR, decreasing = F), ]
  x <- head(x, 10)
  return(x)
})

hData <- plyr::ldply(hListSignif)


hFDR <- plyr::ldply(hList, function(x){ 
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "FDR"), drop = F])
  })

hWideFDR <- tidyr::pivot_wider(hFDR, names_from = .id, values_from = FDR)
hMatFDR <-  as.matrix(hWideFDR[, 2:ncol(hWideFDR)])
rownames(hMatFDR) <- hWideFDR$GeneSet

rownames(hMatFDR) <- gsub("KEGG_", "", rownames(hMatFDR) )


hDir <- plyr::ldply(hList, function(x){
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "Directions"), drop = F])
  })

hWideDir <- tidyr::pivot_wider(hDir, names_from = .id, values_from = Directions)
hMatDir <-  as.matrix(hWideDir[, 2:ncol(hWideDir)])
rownames(hMatDir) <- hWideDir$GeneSet

rownames(hMatDir) <- gsub("KEGG_", "", rownames(hMatDir) )

hMatDir[hMatFDR > 0.05] <- 0


```

```{r}

hMatDir <- hMatDir[rowSums(hMatDir == 0) != ncol(hMatDir), ]

```


```{r hallmark, fig.height = 10, fig.width = 8}
textSize <- 8
sc <- 1.4
library(grid)
library(ComplexHeatmap)

htmp <- ComplexHeatmap::Heatmap(
  hMatDir,
  col = c(brewer.pal(11, "PRGn")[10], "gray80", brewer.pal(11, "PRGn")[2]),
  column_title = "Top 10 significant KEGG pathways\n(Fry gene set test)",
  cluster_rows = T,
  # cluster_columns = F,
  cluster_columns = T,
  show_row_names = T,
  show_column_names = T,
  row_names_gp = gpar(fontsize = textSize),
  clustering_distance_rows = "manhattan",
  clustering_method_rows = "ward.D2",
  clustering_distance_columns = "manhattan",
  clustering_method_columns = "ward.D2",
  heatmap_legend_param = list(
    at = c(1, 0, -1),
    labels = c("Up-regulated", "Non-significant", "Down-regulated"),
    title = "Direction", grid_height = unit(0.7, "cm"),
    title_gp = gpar(fontsize = textSize*sc, fontface = "italic"),
    labels_gp = gpar(fontsize = textSize*sc)
  )
) 

pdf(
  paste0(figPath, "Heatmap_KEGG_Fry_Comparisons_top10.pdf"),
  height = 6,
  width = 5.2
)
draw(htmp, heatmap_legend_side = "left")
dev.off()
```


## MSigDB C7 (camera)
Read the results.
```{r}

fileList <- list.files(outPath)
hh <- fileList[grepl("CAMERA_MSigDB_C7_", fileList)]
hPath <- paste0(outPath, hh)

hList <- lapply(hPath, read.table, header = T, sep = "\t")

names(hList) <- gsub("CAMERA_MSigDB_C7_", "", hh)
names(hList) <- gsub(".txt", "", names(hList) )

hList<- plyr::llply(hList, function(x){
  x$GeneSet <- rownames(x)
  x$Directions[x$Direction == "Up"] <- 1
  x$Directions[x$Direction == "Down"] <- -1
  
  x <- x[grepl("_UP", row.names(x)), ]
  return(x)
})

# grep("NK_", rownames(hList$CIS_24h_4h), value = T)

hListSignif <- plyr::llply(hList, function(x){
  # x <- x[grepl("NK_", rownames(x)), ]
  # x <- x[x$FDR < 0.05, ]
  x <- x[x$FDR < 0.01, ]
  
  # x <- head(x, 15)
  return(x)
})

hData <- plyr::ldply(hListSignif)


hFDR <- plyr::ldply(hList, function(x){ 
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "FDR"), drop = F])
  })

hWideFDR <- tidyr::pivot_wider(hFDR, names_from = .id, values_from = FDR)
hMatFDR <-  as.matrix(hWideFDR[, 2:ncol(hWideFDR)])
rownames(hMatFDR) <- hWideFDR$GeneSet


hDir <- plyr::ldply(hList, function(x){
  currentStats <- x[unique(hData$GeneSet), ]
  return(currentStats[, c("GeneSet", "Directions"), drop = F])
  })

hWideDir <- tidyr::pivot_wider(hDir, names_from = .id, values_from = Directions)
hMatDir <-  as.matrix(hWideDir[, 2:ncol(hWideDir)])
rownames(hMatDir) <- hWideDir$GeneSet

# rownames(hMatDir) <- gsub("KEGG_", "", rownames(hMatDir) )

hMatDir[hMatFDR > 0.05] <- 0

```


Summarize names
```{r}
rownames(hMatDir) <- gsub("CELL", "", rownames(hMatDir) )
rownames(hMatDir) <- gsub("__", "_", rownames(hMatDir) )
library(stringr)
rownames(hMatDir) <- str_sub(rownames(hMatDir), start = 9)
rownames(hMatDir) <- gsub("^_", "", rownames(hMatDir) )
rownames(hMatDir) <- gsub("VS", "v", rownames(hMatDir) )
rownames(hMatDir) <- gsub("H_", "h_", rownames(hMatDir) )
rownames(hMatDir) <- gsub("3h_POST", "3hPOST", rownames(hMatDir) )
rownames(hMatDir) <- gsub("UNTREATED", "UNTREAT", rownames(hMatDir) )
rownames(hMatDir) <- gsub("LYMPHOID", "LYMPH", rownames(hMatDir) )
rownames(hMatDir) <- gsub("MACROPHAGE", "MACRO", rownames(hMatDir) )
rownames(hMatDir) <- gsub("DIFFERENTIATED", "DIFF", rownames(hMatDir) )
rownames(hMatDir) <- gsub("_AND_", "_&_", rownames(hMatDir) )
```


```{r hallmark, fig.height = 10, fig.width = 8}
textSize <- 7
sc <- 1.4
library(grid)
library(ComplexHeatmap)

htmp <- ComplexHeatmap::Heatmap(
  hMatDir,
  col = c(brewer.pal(11, "PRGn")[10], "gray80", brewer.pal(11, "PRGn")[2]),
  column_title = "Significant immune signatures (MSigDB C7)",
  cluster_rows = T,
  cluster_columns = T,
  # cluster_columns = F,
  show_row_names = T,
  show_column_names = T,
  row_names_gp = gpar(fontsize = textSize),
  clustering_distance_rows = "manhattan",
  clustering_method_rows = "ward.D2",
  clustering_distance_columns = "manhattan",
  clustering_method_columns = "ward.D2",
  heatmap_legend_param = list(
    at = c(1, 0, -1),
    labels = c("Up-regulated", "Non-significant", "Down-regulated"),
    title = "Direction", grid_height = unit(0.7, "cm"),
    title_gp = gpar(fontsize = textSize*sc, fontface = "italic"),
    labels_gp = gpar(fontsize = textSize*sc)
  )
) 

pdf(
  paste0(figPath, "Heatmap_C7_Allsignatures_Camera_Comparisons.pdf"),
  height = 6,
  width = 5.2
)
draw(htmp, heatmap_legend_side = "left")
dev.off()
```


## Venn diagrams
### GO terms
```{r}

for(i in compr) {
  assign(paste0("GO_", i),
         read.table(
           paste0(outPath, "GO_", i, ".txt"),
           header = T,
           sep = "\t"
         ))
}

library(VennDiagram)
myCol <- brewer.pal(3, "Pastel2")

venn.diagram(
  x = list(
    head(GO_DKO$Term, 50),
    head(GO_KO1$Term, 50),
    head(GO_KO3$Term, 50)
  ),
  category.names = c("DKO" , "Ikzf1 KO" , "Ikzf3 KO"),
  filename = paste0(figPath, 'VennDiagram_GO_top50.png'),
  output = TRUE,
  # Circles
  fill = myCol,
  lwd = 1,
  # lty = 'blank',
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 480 ,
  width = 480 ,
  resolution = 300,
  compression = "lzw",
   # Numbers
  cex = .6,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)

```

### KEGG pathways

```{r}

for(i in compr) {
  assign(paste0("KEGG_", i),
         read.table(
           paste0(outPath, "KEGG_", i, ".txt"),
           header = T,
           sep = "\t"
         ))
}


library(VennDiagram)
myCol <- brewer.pal(3, "Pastel2")

venn.diagram(
  x = list(
    head(KEGG_DKO$Pathway, 50),
    head(KEGG_KO1$Pathway, 50),
    head(KEGG_KO3$Pathway, 50)
  ),
  category.names = c("DKO" , "Ikzf1 KO" , "Ikzf3 KO"),
  filename = paste0(figPath, 'VennDiagram_KEGG_top50.png'),
  output = TRUE,
  # Circles
  fill = myCol,
  lwd = 1,
  # lty = 'blank',
  lty = 1,
  # Output features
  imagetype = "png" ,
  height = 480 ,
  width = 480 ,
  resolution = 300,
  compression = "lzw",
   # Numbers
  cex = .6,
  fontface = "bold",
  fontfamily = "sans",
  # Set names
  cat.cex = 0.6,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055, 0.085),
  cat.fontfamily = "sans",
  rotation = 1
)

```



### GO for Ikzf1 genes - OLD
```{r}
library(GO.db)
library(rrvgo)
library(org.Hs.eg.db)
library(annotate)

extremeDEGs1 <- DEstats %>%
  filter(grepl("IKZF1", DEGs_group) |
         grepl("DEGs in all KOs", DEGs_group)) %>%
  filter(abs(logFC_KO1) > 3 & adj.P.Val_KO1 < 0.01) %>% 
  data.frame()

extremeDEGs1_only1 <- extremeDEGs1 %>%
  filter(
    grepl("DEGs only in IKZF1 KO", DEGs_group) |
    grepl("DEGs in DKO & IKZF1 KO", DEGs_group) 
     ) %>% 
           data.frame()

currentBackground <- unique(DEstats$Symbol)

currentIDs_bc <-
  mapIds(org.Mm.eg.db,
         keys = currentBackground,
         'ENTREZID',
         'SYMBOL')

currentIDs_bc <- currentIDs_bc[complete.cases(currentIDs_bc)]


# currentSymbols <-  extremeDEGs1_only1$Symbol
currentSymbols <-  DEstats$Symbol[DEstats$DEGs_group =="DEGs only in IKZF1 KO"] ## 33 genes

currentIDs <-
  mapIds(org.Mm.eg.db,
         keys = currentSymbols,
         'ENTREZID',
         'SYMBOL')

currentIDs <- currentIDs[complete.cases(currentIDs)]



kegg <-
  limma::kegga(currentIDs,
               FDR = 0.01,
               species ="Mm",
               universe = currentIDs_bc)

kegg_analysis <- limma::topKEGG(kegg, number = Inf)
## only Linoleic acid metabolism

go <-
  limma::goana(currentIDs,
               FDR = 0.01,
               species = "Mm",
               universe = currentIDs_bc)

go_analysis <- limma::topGO(go, ontology = "BP", number = Inf)

go_analysis <- go_analysis[go_analysis$P.DE < 0.01, ]
# go_analysis <- go_analysis[go_analysis$P.DE < 0.0001, ]
 

go_analysis <- go_analysis[go_analysis$N > 10 & go_analysis$N < 1000, ]
# bp <- names(Ontology(GOTERM))[Ontology(GOTERM) == "BP"]
# # 29211 BP terms

simMatrix <- calculateSimMatrix(rownames(go_analysis),
                                orgdb = "org.Mm.eg.db",
                                ont = "BP",
                                method = "Rel")

scores <- setNames(-log10(go_analysis$P.DE), rownames(go_analysis))

reducedTerms <- reduceSimMatrix(simMatrix,
                                scores,
                                threshold = 0.9,
                                orgdb = "org.Mm.eg.db")


pdf(
  paste0(figPath, "GO_DEGs_onlyIKZF1.pdf"),
  height = 7,
  width = 14
)
heatmapPlot(
  simMatrix,
  reducedTerms,
  annotateParent = TRUE,
  annotationLabel = "parentTerm",
  fontsize = 14,
  fontsize_row = 3,
  fontsize_col = 3
)
dev.off()
```



# Heatmaps
Look at the expression of these genes in the RNA-seq data.
```{r}
annotData <- annot_rna
annotData <- annot_rna[! row.names(annotData) %in% badSamples, ]
```

## Heatmap of common target genes
```{r}
cols <- brewer.pal(8, "Dark2")[1:4]
GroupCols <- cols
names(GroupCols) <- levels(annotData$Group)
colList <- list(Group = GroupCols)

library(ComplexHeatmap)
source("../script/plotHeatmapExpr.R")

htmp <- plotHeatmapExpr(
  genes = gg,
  exprData = ex,
  subsetSample = colnames(ex),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = colList,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = T,
  showColNames = T,
  textSize = 6,
  plotTitle = "IKZF1 target genes obtained from\n two ChIP-seq analysis methods"
)

pdf(paste0(figPath, "Heatmap_Target_ComTwoMethods_rowNames.pdf"), width = 6, height = 12)
draw(htmp)
dev.off()
```


## Heatmap of selected genes 
```{r}
g <- DEstats_comIkzf2_DEGsKO$Symbol
  
cols <- brewer.pal(8, "Dark2")[1:4]
GroupCols <- cols
names(GroupCols) <- levels(annotData$Group)
colList <- list(Group = GroupCols)

library(ComplexHeatmap)
source("../script/plotHeatmapExpr.R")

htmp <- plotHeatmapExpr(
  genes = g,
  exprData = ex[, row.names(annotData)],
  subsetSample = row.names(annotData),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = colList,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = T,
  showColNames = T,
  textSize = 6,
  plotTitle = "Common genes between DKO DEGs and Ikzf2 targets\n(Ikzf2 targets were common across biological reps)"
)

pdf(paste0(figPath, "Heatmap_Com_Ikzf2TargetCom_DKO.pdf"), width = 5.3, height = 5.5)
draw(htmp)
dev.off()
```

## Heatmaps of DEGs
### DEGs in DKO vs Cre
```{r}
gg <- DEGs_DKO$external_gene_name

currentColList <- list(Group = colList$Group[grepl("IKZF3_IKZF1_KO", names(colList$Group)) |
                                grepl("Cre", names(colList$Group))])

htmp <- plotHeatmapExpr(
  genes = gg,
  exprData = ex,
  subsetSample = colnames(ex)[grepl("IKZF3_IKZF1_KO", colnames(ex)) |
                                grepl("Cre", colnames(ex))],
  heatmapCol = viridis::viridis(100),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = currentColList,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = F,
  showColNames = F,
  textSize = 12,
  plotTitle = "DEGs: DKO vs Cre (n = 1768)"
)

pdf(paste0(figPath, "Heatmap_DEGs_DKO_Cre.pdf"), width = 6, height = 8)
draw(htmp)
dev.off()
```

### DEGs in IKZF1 vs Cre
```{r}
gg <- DEGs_KO1$external_gene_name

currentColList <- list(Group = colList$Group[grepl("^IKZF1_KO", names(colList$Group)) |
                                grepl("Cre", names(colList$Group))])

htmp <- plotHeatmapExpr(
  genes = gg,
  exprData = ex,
  subsetSample = colnames(ex)[grepl("RPKM_IKZF1_KO", colnames(ex)) |
                                grepl("Cre", colnames(ex))],
  heatmapCol = viridis::viridis(100),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = currentColList,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = F,
  showColNames = F,
  textSize = 12,
  plotTitle = "DEGs: Ikzf1 KO vs Cre (n = 1057)"
)

pdf(paste0(figPath, "Heatmap_DEGs_IKZF1_Cre.pdf"), width = 6, height = 8)
draw(htmp)
dev.off()
```

### DEGs in IKZF3 vs Cre
```{r}
gg <- DEGs_KO3$external_gene_name

currentColList <- list(Group = colList$Group[grepl("^IKZF3_KO", names(colList$Group)) |
                                grepl("Cre", names(colList$Group))])

htmp <- plotHeatmapExpr(
  genes = gg,
  exprData = ex,
  subsetSample = colnames(ex)[grepl("RPKM_IKZF3_KO", colnames(ex)) |
                                grepl("Cre", colnames(ex))],
  heatmapCol = viridis::viridis(100),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = currentColList,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = F,
  showColNames = F,
  textSize = 12,
  plotTitle = "DEGs: Ikzf3 KO vs Cre (n = 763)"
)

pdf(paste0(figPath, "Heatmap_DEGs_IKZF3_Cre.pdf"), width = 6, height = 8)
draw(htmp)
dev.off()
```


### DEGs Common between IKZF1 vs DKO
```{r}
gg <- intersect(DEGs_DKO$external_gene_name, DEGs_KO1$external_gene_name)

currentColList <- list(Group = colList$Group[names(colList$Group) != "IKZF3_KO"])

htmp <- plotHeatmapExpr(
  genes = gg,
  exprData = ex,
  subsetSample = colnames(ex)[ ! grepl("RPKM_IKZF3_KO", colnames(ex))],
  heatmapCol = viridis::viridis(100),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = currentColList,
  annotLegened_ncol = 1,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = F,
  showColNames = F,
  heatmapLegendTitle = "Scaled logRPKM",
  textSize = 12,
  plotTitle = "DEGs common between \nIkzf1 KO vs Cre and DKO vs Cre (n = 636)"
)

pdf(paste0(figPath, "Heatmap_DEGsCommon_IKZF1_DKO.pdf"), width = 6, height = 7)
draw(htmp)
dev.off()
```

### DEGs Common between IKZF3 vs DKO
```{r}
gg <- intersect(DEGs_DKO$external_gene_name, DEGs_KO3$external_gene_name)

currentColList <- list(Group = colList$Group[names(colList$Group) != "IKZF1_KO"])

htmp <- plotHeatmapExpr(
  genes = gg,
  exprData = ex,
  subsetSample = colnames(ex)[ ! grepl("RPKM_IKZF1_KO", colnames(ex))],
  heatmapCol = viridis::viridis(100),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = currentColList,
  annotLegened_ncol = 1,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = F,
  showColNames = F,
  heatmapLegendTitle = "Scaled logRPKM",
  textSize = 12,
  plotTitle = "DEGs common between \nIkzf3 KO vs Cre and DKO vs Cre (n = 369)"
)

pdf(paste0(figPath, "Heatmap_DEGsCommon_IKZF3_DKO.pdf"), width = 6, height = 7)
draw(htmp)
dev.off()
```

### DEGs Common across all KOs
```{r}
gg <- Reduce(intersect, list(DEGs_DKO$external_gene_name,
                             DEGs_KO3$external_gene_name,
                             DEGs_KO1$external_gene_name))

currentColList <- colList

htmp <- plotHeatmapExpr(
  genes = gg,
  exprData = ex,
  subsetSample = colnames(ex),
  heatmapCol = viridis::viridis(100),
  annotData = annotData,
  annotColumns = c("Group"),
  annotColList = currentColList,
  annotLegened_ncol = 1,
  geneAnnot = NULL,
  ### one column data frame with "Genes" as column name
  geneCol = NULL,
  clustRows = T,
  clustCols = T,
  showRowNames = F,
  showColNames = F,
  heatmapLegendTitle = "Scaled logRPKM",
  textSize = 12,
  plotTitle = "DEGs common between all comparisons (n = 165)"
)

pdf(paste0(figPath, "Heatmap_DEGsCommon_AllKOs.pdf"), width = 6, height = 7)
draw(htmp)
dev.off()
```

# logFC plots
## IKZF1 vs DKO logFCs
```{r}
currentcol <- c(brewer.pal(9, "Set1")[-6])

pDEGs <- DEstats %>%
  filter(DEGs_group != "Non-DEGs") %>%
  ggplot(., aes(logFC_KO1, logFC_DKO, color = DEGs_group, text = Symbol)) +
  geom_point(alpha = 0.5, size = 0.9) +
  scale_color_manual(values = currentcol) +
  theme_bw()

pDEGsIntr <- plotly::ggplotly(pDEGs, tooltips = "text")

htmlwidgets::saveWidget(
  widget = pDEGsIntr, 
  file = paste0(figPath, "logFC_IKZF1_DKO_allDEGs_colorDEGsGroups.html"))


pTarget <- DEstats %>%
  filter(DEGs_group != "Non-DEGs") %>%
  arrange(Target) %>%
  ggplot(., aes(logFC_KO1, logFC_DKO, color = Target, text = Symbol)) +
  geom_point(alpha = 0.5, size = 0.9) +
  scale_color_manual(values = currentcol[c(8, 1, 2, 3, 4, 5)]) +
  theme_bw()

pTargetIntr <- plotly::ggplotly(pTarget, tooltips = "text")

htmlwidgets::saveWidget(
  widget = pTargetIntr, 
  file = paste0(figPath, "logFC_IKZF1_DKO_allDEGs_colorTargets_IKZF1_IKZF3.html"))
```

DEGs that are amongst target genes
```{r}
library(ggrepel)
pTarget <- DEstats %>%
  filter(DEGs_group != "Non-DEGs") %>%
  filter(Target != "Non-Target gene") %>%
  arrange(Target) %>%
  # arrange(TargetGene) %>%
  ggplot(., aes(logFC_KO1, logFC_DKO, color = Target, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1.2,
             show.legend = F) +
  geom_vline(xintercept = 0, color = "gray40", lty = 2.5) +
  geom_hline(yintercept = 0, color = "gray40", lty = 2.5) +
  geom_label_repel(
    aes(label = Symbol,
        fill = factor(Target)),
    color = 'white',
    segment.colour = "black",
    size = 1.8, segment.size = 0.2
  ) +
  ggtitle("DEGs that are among target genes") +
  scale_color_manual(values = currentcol[c(1, 2, 3, 4, 5)]) +
  scale_fill_manual(values = currentcol[c(1, 2, 3, 4, 5)]) +
  labs(fill = "Target") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction = "vertical")

ggsave(
  plot = pTarget,
  filename = "logFC_IKZF1_DKO_allDEGs_colorTargets_MACS1Union.pdf",
  device = "pdf",
  path = figPath,
  width = 5,
  height = 6,
  dpi = 300
)
```

## IKZF3 vs DKO logFCs
```{r}
currentcol <- c(brewer.pal(9, "Set1")[-6])

pDEGs <- DEstats %>%
  filter(DEGs_group != "Non-DEGs") %>%
  ggplot(., aes(logFC_KO3, logFC_DKO, color = DEGs_group, text = Symbol)) +
  geom_point(alpha = 0.5, size = 0.9) +
  scale_color_manual(values = currentcol) +
  theme_bw()

pDEGsIntr <- plotly::ggplotly(pDEGs, tooltips = "text")

htmlwidgets::saveWidget(
  widget = pDEGsIntr, 
  file = paste0(figPath, "logFC_IKZF3_DKO_allDEGs_colorDEGsGroups.html"))


pTarget <- DEstats %>%
  filter(DEGs_group != "Non-DEGs") %>%
  arrange(Target) %>%
  ggplot(., aes(logFC_KO3, logFC_DKO, color = Target, text = Symbol)) +
  geom_point(alpha = 0.5, size = 0.9) +
  scale_color_manual(values = currentcol[c(8, 1, 2, 3, 4, 5)]) +
  theme_bw()

pTargetIntr <- plotly::ggplotly(pTarget, tooltips = "text")

htmlwidgets::saveWidget(
  widget = pTargetIntr, 
  file = paste0(figPath, "logFC_IKZF3_DKO_allDEGs_colorTarget_IKZF1_IKZF3.html"))
```


DEGs that are amongst target genes
```{r}
library(ggrepel)
pTarget <- DEstats %>%
  filter(DEGs_group != "Non-DEGs") %>%
  filter(Target != "Non-Target gene") %>%
  arrange(Target) %>%
  ggplot(., aes(logFC_KO3, logFC_DKO, color = Target, text = Symbol)) +
  geom_point(alpha = 1,
             size = 1.2,
             show.legend = F) +
  geom_vline(xintercept = 0, color = "gray40", lty = 2.5) +
  geom_hline(yintercept = 0, color = "gray40", lty = 2.5) +
  geom_label_repel(
    aes(label = Symbol,
        fill = factor(Target)),
    color = 'white',
    segment.colour = "black",
    size = 1.8, segment.size = 0.2
  ) +
  ggtitle("DEGs that are among target genes") +
  scale_color_manual(values = currentcol[c(1, 2, 3, 4, 5)]) +
  scale_fill_manual(values = currentcol[c(1, 2, 3, 4, 5)]) +
  labs(fill = "Target") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction = "vertical")

ggsave(
  plot = pTarget,
  filename = "logFC_IKZF3_DKO_allDEGs_colorTargets_MACS1Union.pdf",
  device = "pdf",
  path = figPath,
  width = 5,
  height = 6,
  dpi = 300
)
```

## Target genes that are among DEGs
There are 93 target genes that are among DEGs, including the four IKZF1-4. We remove Ikzf5 because it is not target genes and it is not DEGs.
```{r}
DEGs_target <- DEstats %>% 
  filter((DEGs_group != "Non-DEGs" & 
           Target != "Non-Target gene") | 
           grepl("Ikzf", Symbol)) %>% 
  filter(Symbol != "Ikzf5") %>% 
  data.frame()

rownames(DEGs_target) <- DEGs_target$Symbol
```

### Heatmap of logFCs
```{r}
library(tidyverse)
library(viridis)
source(paste0(scriptPath, "plotHeatmapExpr.R"))

cc <- DEGs_target %>%
  rownames_to_column('geneRows') %>%
  select(logFC_DKO, logFC_KO1, logFC_KO3, DEGs_group, TargetGene, geneRows, Symbol) %>%
  rename(Genes = DEGs_group) %>%
  column_to_rownames('geneRows')

geneAnnotData <- cc[, "Genes", drop = F]
geneColours <- brewer.pal(10, "Paired")[c(6, 1, 3, 5, 8, 2, 4 )]
names(geneColours) <- unique(geneAnnotData$Genes)[order(unique(geneAnnotData$Genes))]
  
  
for(t in unique(cc$TargetGene[cc$TargetGene != "Non-Target gene"])){
  dd <- cc[cc$TargetGene == t | cc$TargetGene == "Non-Target gene", ]
  currentT <- stringr::str_remove(t, "^Target gene: ")
  
  pdf(paste0(figPath, "Heatmap_TargetGenes_", currentT, "_DEGs_noClust.pdf"), height = 9, width = 5)

  currentGenes <- dd$Symbol
  currentGeneAnnotData <- dd[, "Genes", drop = F]
  currentGeneColours <- geneColours[names(geneColours) %in% currentGeneAnnotData$Genes]
  
  plotData <- dd[, c('logFC_DKO', 'logFC_KO1', 'logFC_KO3')]
  minVal <- min(plotData)
  maxVal <- max(plotData)

  plotHeatmapExpr(
    exprData = plotData,
    heatmapCol = circlize::colorRamp2(
      c(minVal, 0, maxVal), 
      c(rev(viridis(100))[1] , "white" , rev(viridis(100))[100])),
    # heatmapCol = rev(viridis(100)),
    genes = currentGenes,
    subsetSample = NULL,
    annotData = NULL, 
    annotColumns = NULL,
    annotColList = NULL, 
    scaleData = F,
    geneAnnot = geneAnnotData,
    geneCol = geneColours,
    clustRows = T,
    clustCols = F,
    showRowNames = T,
    showColNames = T, 
    textSize = 9,
    heatmapLegendTitle = "logFC",
    plotTitle = paste0("logFCs of Target genes: ", currentT)
  )
  dev.off()
}

```

### Heatmap of common DEGs
There are 165 DEGs in all comparisons, of which 5 are amongs the taget genes.

```{r}
DEGsAll <- DEstats[DEstats$DEGs_group == "DEGs in all KOs", ]

table(DEGsAll$TargetGene)
DEGsAll[! DEGsAll$TargetGene == "Non-Target gene", ]

# DEGsAllLogFC <- DEGsAll[, c("Symbol", grep("logFC", colnames(DEGsAll), value = T))]

DEGsAllLogFCLong <- 
  DEGsAll %>% 
  select(c("Symbol", grep("logFC", colnames(DEGsAll), value = T))) %>% 
  # DEGsAllLogFC %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Consistent",
                               TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  data.frame()



pdf(paste0(figPath, "WaterFall_logFCs_DEGs_allKOs.pdf"), height = 7, width = 25)
DEGsAllLogFCLong %>%
  # filter(Direction == "Different") %>%
  ggplot(., aes(
    x = reorder(Symbol, logFC),
    y = logFC,
    fill = Direction
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 1) +
  # scale_fill_manual(values = brewer.pal(9, "Set1")[c(4)]) +
  scale_fill_manual(values = brewer.pal(9, "Set1")[c(9, 4)]) +
  theme_bw() +
  ggtitle("Genes that are DE in all three comparisons") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank())
dev.off()
```


# Find genes correlated with IKZF (Optional)
## Correlation of target genes with IKZF1
```{r}
textSize <- 1.2
ikzfGenes <- c("Ikzf1", 
               "Ikzf2", 
               "Ikzf3", 
               "Ikzf4")

## 471 genes
targetGenes <- DEstats$Symbol[DEstats$TargetGene != "Non-Target gene"]
exprCre <- t(ex[, grepl("Cre", colnames(ex))])

exprCreTarget <- exprCre[, colnames(exprCre) %in% targetGenes |
                           colnames(exprCre) %in% ikzfGenes]


corVals <- sapply(colnames(exprCreTarget)[2:ncol(exprCreTarget)], function(x){
  # currentData <- exprCreTarget[, c(x, "Ikzf")]
  cc <- cor.test(exprCreTarget[, x], exprCreTarget[,"Ikzf1"], method = "spearman" )[[4]]
  return(cc)
})

names(corVals) <- colnames(exprCreTarget)[2:ncol(exprCreTarget)]
posCor1 <- names(corVals[corVals >= 0.8])
negCor1 <- names(corVals[corVals <= -0.8])

posData1 <- exprCreTarget[, c("Ikzf1", posCor1)]
negData1 <- exprCreTarget[, c("Ikzf1", negCor1)]

pPos1 <- list()

for(i in colnames(posData1)) {
  xcol <- paste0("`", i, "`")
  ycol <- "`Ikzf1`"
  pPos1[[i]] <- ggplot(data = data.frame(posData1, check.names = F),
                       aes_string(x = xcol, y = ycol)) +
    geom_point() +
    geom_smooth(method = "lm") +
    DEGreport::geom_cor(size = 3) +
    theme_bw() +
    scale_x_continuous(
      breaks = equal_breaks(n = nBreak, s = scalingFactor),
      expand = c(scalingFactor, 0)
    ) +
    scale_y_continuous(
      breaks = equal_breaks(n = nBreak, s = scalingFactor),
      expand = c(scalingFactor, 0)
    ) +
    theme(axis.title = element_text(size = rel(textSize)),
          axis.text = element_text(angle = 0, size = rel(textSize)))
}

pdf(
  paste0(
    figPath,
    "Scatterplot_IKZF1Targets_IKZF1Expr_Cre_PosCor.pdf"
  ),
  height = 18,
  width = 23
)
do.call(gridExtra::grid.arrange, pPos1[1:62])
do.call(gridExtra::grid.arrange, pPos1[63:124])
dev.off()

```

## Correlation of target genes with IKZF3
```{r}

corVals <- sapply(colnames(exprCreTarget)[colnames(exprCreTarget) != "Ikzf3"], function(x){
  # currentData <- exprCreTarget[, c(x, "Ikzf")]
  cc <- cor.test(exprCreTarget[, x], exprCreTarget[,"Ikzf3"], method = "spearman" )[[4]]
  return(cc)
})

names(corVals) <- colnames(exprCreTarget)[colnames(exprCreTarget) != "Ikzf3"]
posCor3 <- names(corVals[corVals >= 0.8])
negCor3 <- names(corVals[corVals <= -0.8])

posData3 <- exprCreTarget[, c("Ikzf3", posCor3)]
negData3 <- exprCreTarget[, c("Ikzf3", negCor3)]

textSize <- 1.3


pPos3 <- list()
for(i in colnames(posData3)){
  xcol <- paste0("`", i, "`")
  ycol <- "`Ikzf3`"
  pPos3[[i]] <- ggplot(data = data.frame(posData3, check.names = F),
                   aes_string(x = xcol, y = ycol)) +
    geom_point() + 
    geom_smooth(method = "lm") +
    DEGreport::geom_cor(size = 3) +
    theme_bw() +
    scale_x_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                       expand = c(scalingFactor, 0)) +
    scale_y_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                       expand = c(scalingFactor, 0))  +
    theme(
    axis.title = element_text(size = rel(textSize)),
    axis.text = element_text(angle = 0, size = rel(textSize)))
}

pdf(
  paste0(figPath, "Scatterplot_IKZF1Targets_IKZF3Expr_Cre_PosCor.pdf"),
  height = 18,
  width = 22
)
do.call(gridExtra::grid.arrange, pPos3[1:53])
do.call(gridExtra::grid.arrange, pPos3[54:107])
dev.off()

```


## Correlation of genes with IKZF1 in Cre and IKZF3 KO
```{r}
textSize <- 1.5
currentTheme <- theme_bw() +
    theme(
    axis.title = element_text(size = rel(textSize)),
    axis.text = element_text(angle = 0, size = rel(textSize)),
    # strip.background = element_rect(colour = "#f0f0f0", fill = "#f0f0f0"),
    strip.text = element_text(size = rel(textSize)),
    axis.line = element_line(colour = "black", size = 0.5),
        legend.title = element_text(size = rel(textSize), face = "italic"),
    legend.text=element_text(size = rel(textSize)),
    legend.key.size = unit(1, 'lines'),  ## increase the line space
    plot.title = element_text(
      face = "bold",
      size = rel(textSize),
      hjust = 0.5
    ))
```

### Genes positively correlated with IKZF1 in Cre and IKZF3 KO
```{r}

exprCreKO3 <- t(ex[, grepl("Cre", colnames(ex)) | grepl("IKZF3_KO", colnames(ex))])

corVals <- sapply(colnames(exprCreKO3)[colnames(exprCreKO3) != "Ikzf1"], function(x){
  # currentData <- exprCreTarget[, c(x, "Ikzf")]
  cc <- cor.test(exprCreKO3[, x], exprCreKO3[,"Ikzf1"], method = "spearman" )[[4]]
  return(cc)
})

names(corVals) <- colnames(exprCreKO3)[colnames(exprCreKO3) != "Ikzf1"]
corVals <- corVals[complete.cases(corVals)]

posCor <- names(corVals[corVals >= 0.9])
negCor <- names(corVals[corVals <= -0.9])

posData <- data.frame(exprCreKO3[, c("Ikzf1", posCor)], 
                      Group = c(rep("IKZF3_KO", 4), rep("Cre", 4)), 
                      check.names = F) %>% 
  rename(Mar5 = `Mar-05`)



pPos <- posData %>%
  pivot_longer(
    data = . ,
    names_to = "Genes",
    values_to = "logRPKM",
    cols = 2:(ncol(.) - 1)
  ) %>%
  ggplot(data = .,
         aes(x = logRPKM, y = Ikzf1, color = Group)) +
  geom_point(size = 2, alpha = 0.8) +
  facet_wrap( ~ Genes, scale = "free") +
  geom_smooth(aes(group = 1),
              method = "lm",
              fullrange = TRUE) +
  DEGreport::geom_cor(aes(group = 1), size = 4) +
  scale_color_manual(values = brewer.pal(4, "Set1")[3:4]) +
  ggtitle("Genes positively correlated with Ikzf1 expression in Cre and Ikzf3 KO subset") +
  scale_x_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                     expand = c(scalingFactor, 0)) +
  scale_y_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                     expand = c(scalingFactor, 0))  +
  currentTheme
    



pdf(
  paste0(figPath, "Scatterplot_Cor_IKZF1_OtherGenes_CreKO3_PosCor2.pdf"),
  height = 22,
  width = 30
)
pPos
dev.off()

```

### Genes negatively correlated with IKZF1 in Cre and IKZF3 KO
```{r}

negData <- data.frame(exprCreKO3[, c("Ikzf1", negCor)], 
                      Group = c(rep("IKZF3_KO", 4), rep("Cre", 4)), 
                      check.names = F) 



pNeg <- negData %>%
  pivot_longer(
    data = . ,
    names_to = "Genes",
    values_to = "logRPKM",
    cols = 2:(ncol(.) - 1)
  ) %>%
  ggplot(data = .,
         aes(x = logRPKM, y = Ikzf1, color = Group)) +
  geom_point(size = 2, alpha = 0.8) +
  facet_wrap( ~ Genes, scale = "free") +
  geom_smooth(aes(group = 1),
              method = "lm",
              fullrange = TRUE) +
  DEGreport::geom_cor(aes(group = 1), size = 4) +
  scale_color_manual(values = brewer.pal(4, "Set1")[3:4]) +
  ggtitle("Genes negatively correlated with Ikzf1 expression in Cre and Ikzf3 KO subset") +
  scale_x_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                     expand = c(scalingFactor, 0)) +
  scale_y_continuous(breaks = equal_breaks(n = nBreak, s = scalingFactor),
                     expand = c(scalingFactor, 0))  +
  currentTheme


pdf(
  paste0(figPath, "Scatterplot_Cor_IKZF1_OtherGenes_CreKO3_NegCor.pdf"),
  height = 22,
  width = 30
)
pNeg
dev.off()

```

### Common genes between these and DEGs only up in KO3

"Zbtb32"  "Slc1a4"  "Aldh1l2" "Dmpk"    "Gpt2"    "Tmem40" 
```{r}
ko3_pos <- DEstats %>%
  filter(logFC_KO3 > 0 & logFC_KO1 < 0 & logFC_DKO < 0 & grepl("IKZF3", DEGs_group))
  
intersect(ko3_pos$Symbol, colnames(posData))
# intersect(ko3_pos$Symbol, colnames(negData))
```


```{r}
cor_mat <- cor(exprCreTarget)

library(Hmisc)
flattenCorrMatrix <- function(cormat
                              # pmat
                              ) {
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut]
    # ,
    # p = pmat[ut]
    )
}

cor_mat <- rcorr(exprCreTarget)
corStat <- flattenCorrMatrix(cor_mat)
```

Look at the waterfall plots of the correlation values:

```{r}
ikzfGenes <- c("Ikzf1", 
               "Ikzf2", 
               "Ikzf3", 
               "Ikzf4")

DEstats %>% 
  filter((grepl("IKZF1", DEGs_group) | 
            DEGs_group == "DEGs in all KOs") &
           TargetGene != "Non-Target gene" |
           Symbol %in% ikzfGenes) %>% 
  select(c("Symbol", grep("logFC", colnames(DEGsAll), value = T))) %>%
  # DEGsAllLogFC %>%
  mutate(Direction = case_when((logFC_DKO > 0 &
                                  logFC_KO1 > 0 & logFC_KO3 > 0) |
                                 (logFC_DKO < 0 &
                                    logFC_KO1 < 0 &
                                    logFC_KO3 < 0) ~ "Consistent",
                               TRUE ~ "Different"
  )) %>%
  pivot_longer(
    data = . ,
    names_to = "Group",
    values_to = "logFC",
    cols = 2:4
  ) %>%
  ggplot(., aes(
    x = reorder(Symbol, logFC),
    y = logFC,
    fill = Direction
  )) +
  geom_bar(stat = "identity") +
  facet_wrap(. ~ Group, ncol = 1) +
  # scale_fill_manual(values = brewer.pal(9, "Set1")[c(4)]) +
  scale_fill_manual(values = brewer.pal(9, "Set1")[c(9, 4)]) +
  ggtitle("DEGs from Ikzf1 KO that are target genes of the Ikzf1") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.x = element_blank())
# 
# ggsave(filename = "WaterFall_DEGsIKZF1_TargetGeneIKZF1.pdf", device = "pdf", path = figPath, width = 8, height = 7)
```

## Genes correlated with IKZF1

Using the four Cre samples in the data, I calculated correlations between IKZF1 and all the other genes. Then subset these to keep those with spearman's rho correlation > 0.5 and expression variance > 1. These resulted in 7 genes: **Cnnm2, Cd300lb, Hist1h1c, Mir5125, Pmepa1, Rpl3-ps1, and Spsb1**. None of these genes are amongst our target genes, however, some of them are DEGs, for example, Hist1h1c (down-regulated in KDO and up-regulated in IKZF3 KO), Cd300lb (down in both DKO and IKZF3 KO), and Pmepa1 (down-regulated only in DKO).

This list increases to 13 gene if we focus on the genes with rho > 0.5 and variance > 0.8

```{r, warning = F, message = F}
exprCre <- t(ex[, grepl("Cre", colnames(ex))])

corVales <- apply(exprCre, 2, function(x){
  round(cor.test(x, exprCre[, "Ikzf1"], method = "spearman")[[4]], 3)
})

pVals <- apply(exprCre, 2, function(x){
  round(cor.test(x, exprCre[, "Ikzf1"], method = "spearman")[[3]], 3)
})

varianceVals <- apply(exprCre, 2, function(x){
  round(var(x), 3)
})


Ikzf1_corData <- data.frame(Genes = colnames(exprCre), 
                            Rho_Cor = corVales, 
                            p_Value = pVals, 
                            Variance = varianceVals)


Ikzf1_corData %>%
  arrange(-Rho_Cor,-Variance) %>%
  head()
  

plot(Ikzf1_corData$Rho_Cor, Ikzf1_corData$Variance)

Ikzf1_corData_Targets <- Ikzf1_corData %>%
  filter(Rho_Cor > 0.5 & Variance > 1)
  
Ikzf1_corData_Targets
```


```{r}
Ikzf1_corData_Targets <- Ikzf1_corData %>%
  filter(Rho_Cor > 0.5 & Variance > 0.8)
  
Ikzf1_corData_Targets
```

```{r}
as.character(Ikzf1_corData_Targets$Genes)
```

```{r}
for(i in unique(as.character(Ikzf1_corData_Targets$Genes))){
  plot(exprCre[, "Ikzf1"], exprCre[, i ], ylab = i)
}
```

```{r}
for(i in unique(as.character(Ikzf1_corData_Targets$Genes))){
  plot(ex["Ikzf1", ], ex[i ,], ylab = i, col = annot_rna$Group)
}
```





